Search.setIndex({"docnames": ["api/genomicranges", "api/genomicranges.io", "api/modules", "authors", "changelog", "contributing", "index", "license", "readme", "tutorial"], "filenames": ["api/genomicranges.rst", "api/genomicranges.io.rst", "api/modules.rst", "authors.md", "changelog.md", "contributing.md", "index.md", "license.md", "readme.md", "tutorial.md"], "titles": ["genomicranges package", "genomicranges.io package", "genomicranges", "Contributors", "Changelog", "Contributing", "GenomicRanges", "License", "GenomicRanges", "Tutorial"], "terms": {"io": [0, 2], "gtf": [0, 2, 4, 6], "parse_gtf": [0, 1], "read_gtf": [0, 1, 8, 9], "ucsc": [0, 2, 4, 6], "access_gtf_ucsc": [0, 1], "read_ucsc": [0, 1, 8, 9], "class": [0, 4, 5, 8, 9], "seqnam": [0, 2, 8, 9], "rang": [0, 2, 6, 8], "strand": [0, 2, 8, 9], "none": [0, 1, 9], "name": [0, 2, 5, 8, 9], "mcol": [0, 2, 8, 9], "metadata": [0, 2, 8, 9], "valid": [0, 1, 5], "true": [0, 5, 8, 9], "sourc": [0, 1, 5], "base": [0, 4, 5, 6, 8], "object": [0, 4, 6, 8], "provid": [0, 5, 7, 8, 9], "contain": [0, 8, 9], "repres": [0, 6, 8, 9], "oper": [0, 4, 5, 6], "over": [0, 6, 8], "genom": [0, 1, 2, 6, 8], "region": [0, 8, 9], "annot": [0, 1, 8, 9], "note": [0, 6, 9], "The": [0, 1, 5, 7, 9], "document": [0, 4, 6, 7, 8, 9], "some": [0, 4], "method": [0, 4, 5, 6], "ar": [0, 4, 5, 8, 9], "deriv": 0, "from": [0, 1, 4, 5, 6, 7], "r": [0, 5, 9], "bioconductor": [0, 6, 8, 9], "__copy__": [0, 2, 4], "return": [0, 1, 5, 9], "A": [0, 4, 7, 8, 9], "shallow": 0, "copi": [0, 2, 5, 7], "current": [0, 5, 8, 9], "__deepcopy__": [0, 2], "memo": 0, "_nil": 0, "deep": 0, "__getitem__": [0, 2, 4], "subset": [0, 9], "alia": 0, "get_subset": [0, 2], "type": [0, 1, 4, 9], "__iter__": [0, 2], "iter": [0, 6], "row": [0, 1, 6], "genomicrangesit": [0, 2], "__len__": [0, 2], "int": [0, 1], "number": [0, 9], "__repr__": [0, 2], "str": [0, 1, 8], "string": 0, "represent": [0, 8, 9], "thi": [0, 4, 5, 7, 8, 9], "__setitem__": [0, 2], "arg": 0, "valu": [0, 9], "set_subset": [0, 2], "modifi": [0, 5, 7, 9], "place": 0, "binned_averag": [0, 2], "scorenam": [0, 9], "bin": [0, 5, 6], "outnam": [0, 9], "in_plac": 0, "fals": [0, 9], "calcul": [0, 9], "averag": [0, 9], "column": [0, 8], "across": [0, 9], "all": [0, 4, 5, 7, 8, 9], "set": [0, 6, 8], "specifi": [0, 8, 9], "those": 0, "paramet": [0, 1, 4, 9], "score": [0, 8, 9], "comput": [0, 6], "you": [0, 5, 8, 9], "want": [0, 5], "us": [0, 4, 5, 6, 7, 8, 9], "new": [0, 5, 9], "add": [0, 5], "bool": [0, 1], "whether": [0, 1, 7], "rais": [0, 1], "valueerror": [0, 1], "If": [0, 5, 8, 9], "doe": [0, 1, 9], "exist": 0, "i": [0, 1, 4, 5, 6, 7, 8, 9], "float": 0, "typeerror": 0, "either": [0, 5, 9], "origin": [0, 5], "refer": [0, 4, 5, 6, 9], "count_overlap": [0, 2, 9], "queri": [0, 8, 9], "query_typ": 0, "ani": [0, 5, 7], "max_gap": 0, "1": [0, 5, 6, 8, 9], "min_overlap": 0, "ignore_strand": [0, 8, 9], "count": [0, 9], "overlap": [0, 6], "between": [0, 9], "subject": [0, 5, 7, 8, 9], "self": 0, "liter": [0, 1], "start": [0, 2, 5, 8, 9], "end": [0, 2, 8, 9], "within": [0, 9], "must": [0, 8, 9], "one": [0, 4, 5, 9], "good": [0, 5], "begin": 0, "interv": [0, 4, 6], "fulli": 0, "default": [0, 1, 9], "maximum": [0, 9], "gap": [0, 2, 9], "allow": [0, 8, 9], "minimum": 0, "ignor": [0, 4], "list": [0, 1, 5, 8, 9], "same": [0, 5], "length": [0, 9], "indic": [0, 1, 9], "coverag": [0, 2], "shift": [0, 2, 9], "0": [0, 5, 6, 8, 9], "width": [0, 2, 8, 9], "weight": 0, "each": [0, 9], "chromosom": 0, "For": [0, 5, 8, 9], "posit": [0, 9], "cover": 0, "option": [0, 5, 9], "restrict": [0, 2, 7, 9], "dict": [0, 9], "ndarrai": [0, 8], "dictionari": [0, 9], "kei": 0, "vector": [0, 9], "disjoin": [0, 2, 9], "with_reverse_map": [0, 9], "disjoint": [0, 9], "distinct": [0, 9], "pair": [0, 9], "map": 0, "back": [0, 9], "classmethod": 0, "empti": [0, 2], "creat": [0, 4, 8, 9], "an": [0, 7, 8, 9], "zero": 0, "caller": 0, "case": [0, 4, 5], "properti": [0, 6], "get_end": [0, 2], "find_overlap": [0, 2, 9], "select": 0, "find": [0, 5, 9], "first": [0, 5, 9], "last": 0, "arbitrari": 0, "determin": 0, "what": [0, 5], "hit": [0, 8], "choos": [0, 9], "when": [0, 1, 5], "multipl": [0, 8, 9], "flank": [0, 2, 9], "both": [0, 8, 9], "logic": 0, "come": [0, 5], "irang": [0, 6], "given": 0, "occur": [0, 5], "otherwis": [0, 7], "can": [0, 4, 5, 8, 9], "neg": 0, "which": [0, 5], "revers": 0, "so": [0, 7], "prefix": 0, "suffix": 0, "usag": [0, 8], "gr": [0, 8, 9], "3": [0, 5, 6, 8, 9], "where": [0, 8, 9], "x": [0, 8, 9], "result": [0, 9], "xxxxxxx": 0, "were": 0, "becom": 0, "e": [0, 1, 4, 5, 8, 9], "xxxx": 0, "extend": 0, "out": [0, 5, 7, 9], "straddl": 0, "endpoint": 0, "ha": [0, 1, 4, 8], "twice": 0, "illustr": [0, 5], "below": 0, "mai": [0, 4, 5], "onli": [0, 9], "follow": [0, 2, 5, 7, 9], "search": [0, 5, 6], "nearest": [0, 2, 8, 9], "upstream": [0, 5, 9], "from_panda": [0, 2, 8, 9], "input": [0, 1, 9], "datafram": [0, 1, 6], "panda": [0, 1, 4, 6], "data": 0, "identifi": [0, 5, 9], "complement": 0, "union": [0, 1, 2, 9], "extract": 0, "sequenc": [0, 6, 8], "inform": [0, 5, 6], "avail": [0, 5, 6, 9], "get": [0, 5, 6, 8], "numpi": [0, 6], "arrai": [0, 9], "32": 0, "bit": 0, "sign": 0, "integ": 0, "get_mcol": [0, 2], "biocfram": [0, 4, 8, 9], "py": 0, "per": 0, "get_metadata": [0, 2], "get_nam": [0, 2], "get_rang": [0, 2], "get_seqinfo": [0, 2], "get_seqnam": [0, 2], "as_typ": 0, "get_start": [0, 2], "get_strand": [0, 2], "forward": 0, "boolean": [0, 9], "thereof": 0, "support": [0, 6, 8], "normalize_subscript": 0, "scalar": 0, "treat": 0, "attr": 0, "see": [0, 5, 8], "occurr": 0, "interest": [0, 5], "get_width": [0, 2], "intersect": [0, 2, 9], "other": [0, 5, 7, 9], "invert_strand": [0, 2, 9], "invert": [0, 9], "convers": 0, "stai": 0, "trim": [0, 2, 9], "match": [0, 1, 2, 9], "element": [0, 8, 9], "wise": [0, 9], "comparison": [0, 6], "exact": [0, 9], "narrow": [0, 2, 9], "import": [0, 4, 5, 8, 9], "rel": [0, 9], "downstream": [0, 9], "order": [0, 2, 5, 6], "decreas": 0, "sort": [0, 2, 5, 9], "Then": 0, "two": [0, 9], "have": [0, 5, 8, 9], "descend": 0, "index": [0, 5, 6], "preced": [0, 2, 9], "promot": [0, 2, 5, 9], "2000": 0, "200": [0, 9], "gener": [0, 1, 5], "transcript": 0, "site": 0, "tss": [0, 9], "expand": [0, 9], "around": [0, 9], "accord": [0, 5, 9], "argument": [0, 9], "nucleotid": 0, "5": [0, 8, 9], "direct": [0, 9], "full": 0, "defin": 0, "bound": [0, 9], "rank": [0, 2, 6], "its": [0, 5, 9], "reduc": [0, 2, 9], "drop_empty_rang": 0, "min_gap_width": [0, 9], "merg": [0, 5, 7], "adjac": 0, "drop": [0, 5], "separ": 0, "least": [0, 5], "resiz": [0, 2, 9], "fix": [0, 5, 9], "center": [0, 9], "anchor": [0, 9], "cannot": 0, "neither": 0, "nor": 0, "keep_all_rang": [0, 9], "keep": 0, "do": [0, 5, 7], "sampl": [0, 2, 9], "k": [0, 5, 9], "randomli": [0, 9], "set_mcol": [0, 2], "set_metadata": [0, 2], "addit": [0, 4, 5], "set_nam": [0, 2], "set_rang": [0, 2], "set_seqinfo": [0, 2], "contan": 0, "about": [0, 5, 9], "set_seqnam": [0, 2], "set_strand": [0, 2], "should": [0, 1, 5], "assum": [0, 5], "accordingli": [0, 5], "updat": [0, 4, 5, 9], "slice": [0, 6], "filter": 0, "replac": [0, 5], "equal": 0, "setdiff": [0, 2, 9], "differ": [0, 5, 9], "diff": 0, "sliding_window": [0, 2, 9], "step": [0, 5, 9], "slide": [0, 9], "along": [0, 9], "also": [0, 4, 5, 9], "checkout": [0, 5], "tile_genom": [0, 2, 9], "split": [0, 9], "gneomic": 0, "chunk": 0, "tile_by_rang": [0, 2], "n": [0, 5, 9], "subset_by_overlap": [0, 2, 9], "tile": [0, 2], "sub": [0, 8, 9], "seqlength": [0, 2, 9], "partit": [0, 9], "altern": 0, "similarli": 0, "less": [0, 4], "than": [0, 9], "instanc": 0, "to_panda": [0, 2], "convert": [0, 9], "outsid": 0, "non": [0, 5, 9], "circular": [0, 9], "obj": 0, "__next__": [0, 2], "range_length": 0, "just": [0, 5, 8, 9], "sound": [0, 8, 9], "like": [0, 5, 8, 9], "wonder": [0, 8, 9], "why": [0, 8, 9], "need": [0, 5, 8, 9], "let": [0, 8, 9], "u": [0, 5, 8, 9], "usual": [0, 8, 9], "gene": [0, 1, 8, 9], "themselv": [0, 8, 9], "made": [0, 5, 8, 9], "mani": [0, 8, 9], "g": [0, 1, 4, 5, 8, 9], "exon": [0, 8, 9], "nest": [0, 8, 9], "structur": [0, 6, 8, 9], "limit": [0, 7, 8, 9], "function": [0, 4, 5, 8, 9], "pure": [0, 9], "read": [0, 1, 9], "basic": [0, 9], "accessor": [0, 9], "typic": 0, "To": [0, 8, 9], "construct": [0, 6, 8], "simpli": [0, 9], "pass": [0, 5, 9], "chr1": [0, 8, 9], "chr2": [0, 8, 9], "chr3": [0, 8, 9], "2": [0, 5, 6, 8, 9], "4": [0, 8, 9], "10": [0, 8, 9], "30": [0, 8, 9], "50": [0, 8, 9], "60": [0, 8, 9], "b": [0, 5, 9], "chr4": [0, 8, 9], "chr5": [0, 8, 9], "6": [0, 8, 9], "grl": [0, 8, 9], "gr1": [0, 8, 9], "gr2": [0, 8, 9], "gene1": [0, 8, 9], "gene2": [0, 8, 9], "addition": 0, "attribut": 0, "individu": 0, "tupl": 0, "access": 0, "dimens": 0, "more": [0, 4, 5, 9], "genomicrangeslistit": [0, 2], "add_el": [0, 2], "element_metadata": 0, "element_nrow": [0, 2], "correspond": 0, "anoth": 0, "group": [0, 2, 5], "is_circular": [0, 2, 9], "flag": [0, 5], "is_empti": [0, 2], "grangeslist": 0, "seq_info": [0, 2, 9], "underli": [0, 4], "coerc": 0, "specif": [0, 5], "ident": 0, "wa": [0, 5], "get_genom": [0, 2], "get_is_circular": [0, 2], "get_seqlength": [0, 2], "set_genom": [0, 2], "set_is_circular": [0, 2], "Not": 0, "present": 0, "set_seqlength": [0, 2], "uniqu": 0, "merge_seqinfo": [0, 2], "take": 0, "detail": [0, 5, 8, 9], "singl": 0, "final": [0, 4], "contradictori": 0, "thei": [0, 5], "create_np_vector": [0, 2], "force_s": 0, "dont_sum": 0, "forc": 0, "size": [0, 5], "sum": 0, "increment": 0, "sanitize_strand_vector": [0, 2], "slide_interv": [0, 2], "split_interv": [0, 2], "path": [1, 5, 8, 9], "compress": 1, "skiprow": 1, "comment": 1, "file": [1, 5, 6, 7], "gzip": 1, "skip": 1, "header": 1, "charact": 1, "line": 1, "pars": [1, 8, 9], "refgen": 1, "hg19": [1, 8, 9], "shortcod": 1, "hg38": [1, 8], "mm10": 1, "etc": [1, 4], "ensgen": 1, "knowngen": 1, "ncbirefseq": 1, "except": 1, "uri": 1, "load": [1, 8, 9], "model": [1, 5], "packag": [2, 4, 5, 6, 8, 9], "subpackag": 2, "submodul": 2, "modul": [2, 5, 6], "content": [2, 5], "seqinfo": [2, 8, 9], "genomicrangeslist": [2, 6], "util": 2, "jayaram": 3, "kancherla": 3, "gmail": 3, "com": [3, 5], "releas": 4, "migrat": 4, "palat": 4, "googl": 4, "": [4, 5, 6, 8, 9], "python": [4, 5, 6, 8, 9], "style": [4, 5], "guid": [4, 5], "major": [4, 5], "modif": 4, "camelcas": 4, "now": [4, 9], "snake_cas": 4, "In": [4, 5], "docstr": [4, 5], "been": [4, 5, 8], "sphinx": [4, 5], "featur": [4, 5], "link": [4, 5], "privat": [4, 5], "special": 4, "dunder": 4, "intersphinx": 4, "depend": [4, 5], "configur": [4, 5], "flake8": [4, 5], "ruff": 4, "black": [4, 5], "ad": [4, 5], "pyproject": 4, "toml": 4, "setup": [4, 5], "cfg": [4, 5], "annoi": 4, "pyscaffold": [4, 5, 8], "myst": [4, 5], "parser": 4, "markdown": 4, "compil": [4, 5], "instead": [4, 5], "recommonmark": 4, "As": 4, "part": [4, 5], "pre": [4, 5], "commit": [4, 5], "run": [4, 5], "routin": 4, "task": [4, 6], "lint": 4, "format": 4, "befor": [4, 5], "everi": 4, "while": [4, 5], "sometim": [4, 5], "verifi": 4, "bring": 4, "consist": 4, "code": [4, 6], "implement": [4, 9], "readm": 4, "comprehens": 4, "test": [4, 5], "genomicrang": [4, 5], "suppos": 5, "TO": [5, 7, 8, 9], "BE": [5, 7], "exampl": [5, 8], "IT": 5, "servic": 5, "similar": [5, 8, 9], "github": 5, "fork": 5, "pull": 5, "request": 5, "workflow": 5, "gitlab": 5, "bitbucket": 5, "might": [5, 9], "gerrit": 5, "notic": [5, 7], "url": 5, "text": 5, "terminologi": 5, "pleas": 5, "make": 5, "sure": 5, "check": [5, 9], "assumpt": 5, "mind": 5, "thing": 5, "correct": 5, "bottom": 5, "look": 5, "contributor": 5, "especi": 5, "project": [5, 8], "open": 5, "veri": 5, "templat": 5, "few": 5, "extra": 5, "decid": 5, "includ": [5, 7], "mention": 5, "label": 5, "tracker": 5, "autom": 5, "welcom": 5, "focus": 5, "potenti": 5, "familiar": 5, "develop": [5, 6], "process": 5, "kind": [5, 7], "appreci": 5, "git": 5, "never": 5, "collabor": 5, "previous": 5, "org": [5, 8], "resourc": 5, "excel": 5, "freecodecamp": 5, "user": 5, "expect": 5, "consider": 5, "reason": 5, "respect": 5, "doubt": 5, "softwar": [5, 7], "foundat": 5, "conduct": 5, "term": 5, "behavior": 5, "guidelin": 5, "experi": 5, "bug": 5, "don": 5, "t": 5, "anyth": 5, "feel": 5, "free": [5, 7], "fire": 5, "forget": 5, "close": 5, "solut": 5, "alreadi": [5, 6], "problem": 5, "consid": 5, "solv": 5, "program": 5, "system": [5, 6], "version": [5, 6], "reproduc": 5, "try": 5, "simplifi": 5, "reproduct": 5, "minim": 5, "still": 5, "face": 5, "By": 5, "remov": 5, "factor": 5, "help": [5, 6], "root": 5, "caus": 5, "doc": 5, "them": 5, "readabl": 5, "coher": 5, "miss": 5, "mistak": 5, "main": 5, "mean": 5, "kept": 5, "done": 5, "wai": [5, 6], "markup": 5, "languag": 5, "restructuredtext": 5, "commonmark": 5, "extens": 5, "host": 5, "tip": 5, "web": 5, "interfac": 5, "quick": 5, "propos": 5, "mechan": 5, "tricki": 5, "normal": 5, "work": [5, 9], "perfectli": 5, "fine": 5, "quit": 5, "handi": 5, "navig": 5, "folder": 5, "would": 5, "click": 5, "littl": 5, "pencil": 5, "icon": 5, "top": 5, "editor": 5, "onc": 5, "finish": 5, "edit": 5, "write": 5, "messag": 5, "form": [5, 9], "page": [5, 6], "describ": [5, 9], "motiv": 5, "behind": 5, "local": 5, "machin": 5, "tox": 5, "built": 5, "server": 5, "preview": 5, "browser": 5, "http": [5, 8], "localhost": 5, "8000": 5, "python3": 5, "m": 5, "directori": 5, "_build": 5, "html": 5, "explan": 5, "intern": 5, "architectur": 5, "descript": [5, 9], "design": [5, 8], "principl": 5, "summari": 5, "concept": 5, "easi": 5, "quickli": [5, 9], "trivial": 5, "best": 5, "discuss": 5, "often": 5, "avoid": 5, "unnecessari": 5, "we": 5, "recommend": 5, "isol": 5, "virtual": 5, "instal": [5, 8], "easili": [5, 8], "via": 5, "virtualenv": 5, "venv": 5, "activ": 5, "miniconda": 5, "conda": 5, "six": 5, "pytest": 5, "cov": 5, "account": 5, "button": 5, "under": 5, "disk": 5, "yourlogin": 5, "cd": 5, "pip": [5, 6, 8], "setuptool": 5, "abl": 5, "repl": 5, "item": 5, "lot": 5, "hook": 5, "automat": 5, "being": 5, "written": 5, "branch": 5, "hold": 5, "my": 5, "public": 5, "api": [5, 8], "yourself": 5, "author": [5, 6, 7], "rst": 5, "re": [5, 9], "record": 5, "eventu": 5, "compat": 5, "unit": 5, "bugfix": 5, "moreov": 5, "highli": 5, "histori": 5, "log": 5, "graph": 5, "decor": 5, "pretti": 5, "onelin": 5, "abbrev": 5, "recur": 5, "commun": 5, "pattern": 5, "break": 5, "after": 5, "pipx": 5, "sever": 5, "av": 5, "everyth": [5, 9], "push": 5, "remot": 5, "go": 5, "send": 5, "review": 5, "uncom": 5, "paragraph": 5, "pr": 5, "draft": 5, "mark": 5, "readi": 5, "feedback": 5, "continu": 5, "integr": 5, "ci": 5, "requir": 5, "build": 5, "fetch": 5, "tag": 5, "command": 5, "script": 5, "egg": 5, "complet": [5, 8], "well": 5, "info": 5, "src": 5, "txt": 5, "recreat": 5, "reliabl": 5, "7": 5, "OR": [5, 7, 8, 9], "troubl": 5, "weird": 5, "error": 5, "upon": 5, "dedic": 5, "binari": 5, "freshli": 5, "interact": 5, "session": 5, "pdb": 5, "OF": [5, 7, 9], "THE": [5, 7], "fall": 5, "breakpoint": 5, "manual": 5, "section": 5, "pypi": [5, 6, 8], "publicli": 5, "instruct": 5, "permiss": [5, 7], "success": 5, "v1": 5, "clean": 5, "up": [5, 8], "dist": 5, "rm": 5, "rf": 5, "confus": 5, "old": 5, "dirti": 5, "hash": 5, "distribut": [5, 7], "too": 5, "big": 5, "500kb": 5, "unwant": 5, "clutter": 5, "accident": 5, "publish": [5, 7], "upload": 5, "correctli": 5, "definit": 5, "even": 5, "though": 5, "focu": 5, "idea": 5, "collect": 5, "appli": 5, "compani": 5, "proprietari": 5, "equival": [6, 9], "locat": [6, 8, 9], "analysi": [6, 8], "It": [6, 8], "effici": 6, "eco": 6, "deploi": 6, "overview": 6, "further": 6, "tutori": [6, 8], "prefer": 6, "intra": 6, "transform": 6, "inter": 6, "combin": 6, "misc": 6, "contribut": 6, "issu": 6, "report": 6, "improv": 6, "maintain": 6, "licens": 6, "changelog": 6, "mit": 7, "copyright": 7, "c": [7, 9], "2022": 7, "genentech": 7, "inc": 7, "herebi": 7, "grant": 7, "charg": 7, "person": 7, "obtain": 7, "associ": 7, "deal": 7, "without": 7, "right": 7, "sublicens": 7, "sell": 7, "permit": 7, "whom": 7, "furnish": 7, "condit": 7, "abov": 7, "shall": 7, "substanti": 7, "portion": 7, "AS": 7, "warranti": 7, "express": 7, "impli": 7, "BUT": 7, "NOT": 7, "merchant": 7, "fit": 7, "FOR": 7, "particular": 7, "purpos": 7, "AND": 7, "noninfring": 7, "IN": 7, "NO": 7, "event": 7, "holder": 7, "liabl": 7, "claim": 7, "damag": 7, "liabil": 7, "action": 7, "contract": 7, "tort": 7, "aris": 7, "connect": 7, "WITH": 7, "inclus": [8, 9], "v0": 8, "overhaul": 8, "constructor": 8, "chang": 8, "download": [8, 9], "print": [8, 9], "output": 8, "1760959": 8, "truncat": 8, "consol": 8, "relev": [8, 9], "random": [8, 9], "101": [8, 9], "106": [8, 9], "11": [8, 9], "21": [8, 9], "25": [8, 9], "gc": [8, 9], "_": [8, 9], "int64": 8, "112": [8, 9], "2593301003406461": 8, "102": [8, 9], "123": 8, "7207993213776644": 8, "103": [8, 9], "128": [8, 9], "23391468067222065": 8, "104": [8, 9], "134": [8, 9], "7671026589720187": 8, "105": [8, 9], "110": 8, "03355777784472458": 8, "common": [8, 9], "tabular": [8, 9], "dataset": [8, 9], "here": 8, "pd": [8, 9], "df": [8, 9], "109": [8, 9], "111": [8, 9], "4862658925128007": 8, "27948386889389953": 8, "5162697718607901": 8, "5979843806415466": 8, "04740781186083798": 8, "most": 8, "from_ucsc": 8, "100": [8, 9], "115": 8, "119": 8, "116": 8, "120": [8, 9], "1677082": 8, "1677083": 8, "1677084": 8, "1003411": 8, "1003412": 8, "33": 8, "52": 8, "64": 8, "56": 8, "bioc": 8, "perform": 9, "arithmet": 9, "One": 9, "necessari": 9, "coordin": 9, "store": 9, "genomeinfodb": 9, "seq_obj": 9, "seq": 9, "getter": 9, "variou": 9, "asid": 9, "someth": 9, "els": 9, "return_typ": 9, "grang": 9, "accept": 9, "5th": 9, "rownam": 9, "whose": 9, "na": 9, "flanked_gr": 9, "shifted_gr": 9, "resized_gr": 9, "narrow_gr": 9, "prom_gr": 9, "restrict_gr": 9, "114": 9, "140": 9, "trimmed_gr": 9, "is_disjoint": 9, "range_gr": 9, "reduced_gr": 9, "gapped_gr": 9, "disjoin_gr": 9, "isdisjoin": 9, "df_src": 9, "g_src": 9, "df_tgt": 9, "121": 9, "131": 9, "g_tgt": 9, "union_gr": 9, "int_gr": 9, "diff_gr": 9, "seri": 9, "bins_gr": 9, "subject_gr": 9, "binned_avg_gr": 9, "binned_scor": 9, "how": 9, "window": 9, "slidingwindow": 9, "75": 9, "res_vector": 9, "df_subject": 9, "df_queri": 9, "findoverlap": 9, "querytyp": 9, "countoverlap": 9, "subsetbyoverlap": 9, "neighbor": 9, "distance_to_nearest": 9, "distanc": 9, "find_region": 9, "1190": 9, "203": 9, "1200": 9, "query_hit": 9, "test_gr": 9, "duplic": 9, "is_unsort": 9, "unsort": 9, "sorted_gr": 9, "biocutil": 9, "concaten": 9, "combined_gr": 9, "gr3": 9, "flip": 9, "inv_gr": 9, "samp_gr": 9, "progress": 9, "grla": 9, "grlb": 9, "grlc": 9, "cgrl": 9, "later": 9}, "objects": {"": [[0, 0, 0, "-", "genomicranges"]], "genomicranges": [[0, 0, 0, "-", "GenomicRanges"], [0, 0, 0, "-", "GenomicRangesList"], [0, 0, 0, "-", "SeqInfo"], [1, 0, 0, "-", "io"], [0, 0, 0, "-", "utils"]], "genomicranges.GenomicRanges": [[0, 1, 1, "", "GenomicRanges"], [0, 1, 1, "", "GenomicRangesIter"]], "genomicranges.GenomicRanges.GenomicRanges": [[0, 2, 1, "", "__copy__"], [0, 2, 1, "", "__deepcopy__"], [0, 2, 1, "", "__getitem__"], [0, 2, 1, "", "__iter__"], [0, 2, 1, "", "__len__"], [0, 2, 1, "", "__repr__"], [0, 2, 1, "", "__setitem__"], [0, 2, 1, "", "binned_average"], [0, 2, 1, "", "copy"], [0, 2, 1, "", "count_overlaps"], [0, 2, 1, "", "coverage"], [0, 2, 1, "", "disjoin"], [0, 2, 1, "", "empty"], [0, 3, 1, "", "end"], [0, 2, 1, "", "find_overlaps"], [0, 2, 1, "", "flank"], [0, 2, 1, "", "follow"], [0, 2, 1, "", "from_pandas"], [0, 2, 1, "", "gaps"], [0, 2, 1, "", "get_end"], [0, 2, 1, "", "get_mcols"], [0, 2, 1, "", "get_metadata"], [0, 2, 1, "", "get_names"], [0, 2, 1, "", "get_ranges"], [0, 2, 1, "", "get_seqinfo"], [0, 2, 1, "", "get_seqnames"], [0, 2, 1, "", "get_start"], [0, 2, 1, "", "get_strand"], [0, 2, 1, "", "get_subset"], [0, 2, 1, "", "get_width"], [0, 2, 1, "", "intersect"], [0, 2, 1, "", "invert_strand"], [0, 2, 1, "", "match"], [0, 3, 1, "", "mcols"], [0, 3, 1, "", "metadata"], [0, 3, 1, "", "names"], [0, 2, 1, "", "narrow"], [0, 2, 1, "", "nearest"], [0, 2, 1, "", "order"], [0, 2, 1, "", "precede"], [0, 2, 1, "", "promoters"], [0, 2, 1, "", "range"], [0, 3, 1, "", "ranges"], [0, 2, 1, "", "rank"], [0, 2, 1, "", "reduce"], [0, 2, 1, "", "resize"], [0, 2, 1, "", "restrict"], [0, 2, 1, "", "sample"], [0, 3, 1, "", "seqinfo"], [0, 3, 1, "", "seqnames"], [0, 2, 1, "", "set_mcols"], [0, 2, 1, "", "set_metadata"], [0, 2, 1, "", "set_names"], [0, 2, 1, "", "set_ranges"], [0, 2, 1, "", "set_seqinfo"], [0, 2, 1, "", "set_seqnames"], [0, 2, 1, "", "set_strand"], [0, 2, 1, "", "set_subset"], [0, 2, 1, "", "setdiff"], [0, 2, 1, "", "shift"], [0, 2, 1, "", "sliding_windows"], [0, 2, 1, "", "sort"], [0, 3, 1, "", "start"], [0, 3, 1, "", "strand"], [0, 2, 1, "", "subset_by_overlaps"], [0, 2, 1, "", "tile"], [0, 2, 1, "", "tile_by_range"], [0, 2, 1, "", "tile_genome"], [0, 2, 1, "", "to_pandas"], [0, 2, 1, "", "trim"], [0, 2, 1, "", "union"], [0, 3, 1, "", "width"]], "genomicranges.GenomicRanges.GenomicRangesIter": [[0, 2, 1, "", "__iter__"], [0, 2, 1, "", "__next__"]], "genomicranges.GenomicRangesList": [[0, 1, 1, "", "GenomicRangesList"], [0, 1, 1, "", "GenomicRangesListIter"]], "genomicranges.GenomicRangesList.GenomicRangesList": [[0, 2, 1, "", "__copy__"], [0, 2, 1, "", "__deepcopy__"], [0, 2, 1, "", "__getitem__"], [0, 2, 1, "", "__iter__"], [0, 2, 1, "", "__len__"], [0, 2, 1, "", "__repr__"], [0, 2, 1, "", "add_element"], [0, 2, 1, "", "copy"], [0, 2, 1, "", "element_nrows"], [0, 2, 1, "", "empty"], [0, 3, 1, "", "end"], [0, 2, 1, "", "get_mcols"], [0, 2, 1, "", "get_metadata"], [0, 2, 1, "", "get_names"], [0, 2, 1, "", "get_ranges"], [0, 2, 1, "", "groups"], [0, 3, 1, "", "is_circular"], [0, 2, 1, "", "is_empty"], [0, 3, 1, "", "mcols"], [0, 3, 1, "", "metadata"], [0, 3, 1, "", "names"], [0, 3, 1, "", "ranges"], [0, 3, 1, "", "seq_info"], [0, 3, 1, "", "seqnames"], [0, 2, 1, "", "set_mcols"], [0, 2, 1, "", "set_metadata"], [0, 2, 1, "", "set_names"], [0, 2, 1, "", "set_ranges"], [0, 3, 1, "", "start"], [0, 3, 1, "", "strand"], [0, 2, 1, "", "to_pandas"], [0, 3, 1, "", "width"]], "genomicranges.GenomicRangesList.GenomicRangesListIter": [[0, 2, 1, "", "__iter__"], [0, 2, 1, "", "__next__"]], "genomicranges.SeqInfo": [[0, 1, 1, "", "SeqInfo"], [0, 4, 1, "", "merge_SeqInfo"]], "genomicranges.SeqInfo.SeqInfo": [[0, 2, 1, "", "__len__"], [0, 3, 1, "", "genome"], [0, 2, 1, "", "get_genome"], [0, 2, 1, "", "get_is_circular"], [0, 2, 1, "", "get_seqlengths"], [0, 2, 1, "", "get_seqnames"], [0, 3, 1, "", "is_circular"], [0, 3, 1, "", "seqlengths"], [0, 3, 1, "", "seqnames"], [0, 2, 1, "", "set_genome"], [0, 2, 1, "", "set_is_circular"], [0, 2, 1, "", "set_seqlengths"], [0, 2, 1, "", "set_seqnames"]], "genomicranges.io": [[1, 0, 0, "-", "gtf"], [1, 0, 0, "-", "ucsc"]], "genomicranges.io.gtf": [[1, 4, 1, "", "parse_gtf"], [1, 4, 1, "", "read_gtf"]], "genomicranges.io.ucsc": [[1, 4, 1, "", "access_gtf_ucsc"], [1, 4, 1, "", "read_ucsc"]], "genomicranges.utils": [[0, 4, 1, "", "create_np_vector"], [0, 4, 1, "", "sanitize_strand_vector"], [0, 4, 1, "", "slide_intervals"], [0, 4, 1, "", "split_intervals"]]}, "objtypes": {"0": "py:module", "1": "py:class", "2": "py:method", "3": "py:property", "4": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "class", "Python class"], "2": ["py", "method", "Python method"], "3": ["py", "property", "Python property"], "4": ["py", "function", "Python function"]}, "titleterms": {"genomicrang": [0, 1, 2, 6, 8, 9], "packag": [0, 1], "subpackag": 0, "submodul": [0, 1], "modul": [0, 1], "genomicrangeslist": [0, 8, 9], "seqinfo": 0, "util": 0, "content": [0, 1, 6], "io": 1, "gtf": [1, 8, 9], "ucsc": [1, 8, 9], "contributor": 3, "changelog": 4, "version": 4, "0": 4, "3": 4, "develop": 4, "2": 4, "1": 4, "todo": 5, "contribut": 5, "issu": 5, "report": 5, "document": 5, "improv": 5, "code": 5, "submit": 5, "an": 5, "creat": 5, "environ": 5, "clone": 5, "repositori": 5, "implement": 5, "your": 5, "chang": 5, "troubleshoot": 5, "maintain": 5, "task": 5, "releas": 5, "instal": 6, "indic": 6, "tabl": 6, "licens": 7, "from": [8, 9], "file": [8, 9], "irang": [8, 9], "prefer": [8, 9], "wai": [8, 9], "panda": [8, 9], "datafram": [8, 9], "interv": [8, 9], "oper": [8, 9], "further": 8, "inform": [8, 9], "note": 8, "tutori": 9, "construct": 9, "object": 9, "set": 9, "sequenc": 9, "get": 9, "method": 9, "setter": 9, "access": 9, "ani": 9, "column": 9, "rang": 9, "slice": 9, "iter": 9, "over": 9, "intra": 9, "transform": 9, "inter": 9, "genom": 9, "comput": 9, "bin": 9, "summari": 9, "stat": 9, "binned_averag": 9, "gener": 9, "tile": 9, "coverag": 9, "overlap": 9, "base": 9, "search": 9, "comparison": 9, "rank": 9, "order": 9, "combin": 9, "row": 9, "misc": 9, "properti": 9}, "envversion": {"sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.todo": 2, "sphinx.ext.viewcode": 1, "sphinx": 60}, "alltitles": {"genomicranges package": [[0, "genomicranges-package"]], "Subpackages": [[0, "subpackages"]], "Submodules": [[0, "submodules"], [1, "submodules"]], "genomicranges.GenomicRanges module": [[0, "module-genomicranges.GenomicRanges"]], "genomicranges.GenomicRangesList module": [[0, "module-genomicranges.GenomicRangesList"]], "genomicranges.SeqInfo module": [[0, "module-genomicranges.SeqInfo"]], "genomicranges.utils module": [[0, "module-genomicranges.utils"]], "Module contents": [[0, "module-genomicranges"], [1, "module-genomicranges.io"]], "genomicranges.io package": [[1, "genomicranges-io-package"]], "genomicranges.io.gtf module": [[1, "module-genomicranges.io.gtf"]], "genomicranges.io.ucsc module": [[1, "module-genomicranges.io.ucsc"]], "genomicranges": [[2, "genomicranges"]], "Contributors": [[3, "contributors"]], "Changelog": [[4, "changelog"]], "Version 0.3.0 (development)": [[4, "version-0-3-0-development"]], "Version 0.2": [[4, "version-0-2"]], "Version 0.1": [[4, "version-0-1"]], "Todo": [[5, "id1"], [5, "id2"], [5, "id3"], [5, "id5"], [5, "id6"], [5, "id7"], [5, "id8"], [5, "id9"], [5, "id10"], [5, "id11"], [5, "id12"]], "Contributing": [[5, "contributing"]], "Issue Reports": [[5, "issue-reports"]], "Documentation Improvements": [[5, "documentation-improvements"]], "Code Contributions": [[5, "code-contributions"]], "Submit an issue": [[5, "submit-an-issue"]], "Create an environment": [[5, "create-an-environment"]], "Clone the repository": [[5, "clone-the-repository"]], "Implement your changes": [[5, "implement-your-changes"]], "Submit your contribution": [[5, "submit-your-contribution"]], "Troubleshooting": [[5, "troubleshooting"]], "Maintainer tasks": [[5, "maintainer-tasks"]], "Releases": [[5, "releases"]], "GenomicRanges": [[6, "genomicranges"], [8, "genomicranges"], [8, "id1"]], "Install": [[6, "install"]], "Contents": [[6, "contents"]], "Indices and tables": [[6, "indices-and-tables"]], "License": [[7, "license"]], "From UCSC or GTF file": [[8, "from-ucsc-or-gtf-file"], [9, "from-ucsc-or-gtf-file"]], "from IRanges (Preferred way)": [[8, "from-iranges-preferred-way"], [9, "from-iranges-preferred-way"]], "Pandas DataFrame": [[8, "pandas-dataframe"], [9, "pandas-dataframe"]], "Interval Operations": [[8, "interval-operations"]], "GenomicRangesList": [[8, "genomicrangeslist"]], "Further information": [[8, "further-information"]], "Note": [[8, "note"]], "Tutorial": [[9, "tutorial"]], "Construct a GenomicRanges object": [[9, "construct-a-genomicranges-object"]], "Set sequence information": [[9, "set-sequence-information"]], "Get and Set methods": [[9, "get-and-set-methods"]], "Setters": [[9, "setters"]], "Access any column": [[9, "access-any-column"]], "Access ranges": [[9, "access-ranges"]], "Slice operations": [[9, "slice-operations"]], "Iterate over intervals": [[9, "iterate-over-intervals"]], "Intra-range transformations": [[9, "intra-range-transformations"]], "Inter-range methods": [[9, "inter-range-methods"]], "Set operations on genomic ranges": [[9, "set-operations-on-genomic-ranges"]], "Compute over bins": [[9, "compute-over-bins"]], "Summary stats for column": [[9, "summary-stats-for-column"]], "binned_average": [[9, "binned-average"]], "Generate tiles or bins from GenomicRanges": [[9, "generate-tiles-or-bins-from-genomicranges"]], "Generate tiles from Genome": [[9, "generate-tiles-from-genome"]], "Coverage": [[9, "coverage"]], "Overlap based methods": [[9, "overlap-based-methods"]], "Search operations": [[9, "search-operations"]], "Comparison, rank and order operations": [[9, "comparison-rank-and-order-operations"]], "Combine GenomicRanges objects by rows": [[9, "combine-genomicranges-objects-by-rows"]], "Misc operations": [[9, "misc-operations"]], "Construct a GenomicRangesList object.": [[9, "construct-a-genomicrangeslist-object"]], "Properties": [[9, "properties"]], "Combine GenomicRangeslist object": [[9, "combine-genomicrangeslist-object"]]}, "indexentries": {"genomicranges (class in genomicranges.genomicranges)": [[0, "genomicranges.GenomicRanges.GenomicRanges"]], "genomicrangesiter (class in genomicranges.genomicranges)": [[0, "genomicranges.GenomicRanges.GenomicRangesIter"]], "genomicrangeslist (class in genomicranges.genomicrangeslist)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList"]], "genomicrangeslistiter (class in genomicranges.genomicrangeslist)": [[0, "genomicranges.GenomicRangesList.GenomicRangesListIter"]], "seqinfo (class in genomicranges.seqinfo)": [[0, "genomicranges.SeqInfo.SeqInfo"]], "__copy__() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.__copy__"]], "__copy__() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.__copy__"]], "__deepcopy__() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.__deepcopy__"]], "__deepcopy__() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.__deepcopy__"]], "__getitem__() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.__getitem__"]], "__getitem__() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.__getitem__"]], "__iter__() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.__iter__"]], "__iter__() (genomicranges.genomicranges.genomicrangesiter method)": [[0, "genomicranges.GenomicRanges.GenomicRangesIter.__iter__"]], "__iter__() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.__iter__"]], "__iter__() (genomicranges.genomicrangeslist.genomicrangeslistiter method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesListIter.__iter__"]], "__len__() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.__len__"]], "__len__() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.__len__"]], "__len__() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.__len__"]], "__next__() (genomicranges.genomicranges.genomicrangesiter method)": [[0, "genomicranges.GenomicRanges.GenomicRangesIter.__next__"]], "__next__() (genomicranges.genomicrangeslist.genomicrangeslistiter method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesListIter.__next__"]], "__repr__() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.__repr__"]], "__repr__() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.__repr__"]], "__setitem__() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.__setitem__"]], "add_element() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.add_element"]], "binned_average() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.binned_average"]], "copy() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.copy"]], "copy() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.copy"]], "count_overlaps() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.count_overlaps"]], "coverage() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.coverage"]], "create_np_vector() (in module genomicranges.utils)": [[0, "genomicranges.utils.create_np_vector"]], "disjoin() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.disjoin"]], "element_nrows() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.element_nrows"]], "empty() (genomicranges.genomicranges.genomicranges class method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.empty"]], "empty() (genomicranges.genomicrangeslist.genomicrangeslist class method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.empty"]], "end (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.end"]], "end (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.end"]], "find_overlaps() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.find_overlaps"]], "flank() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.flank"]], "follow() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.follow"]], "from_pandas() (genomicranges.genomicranges.genomicranges class method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.from_pandas"]], "gaps() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.gaps"]], "genome (genomicranges.seqinfo.seqinfo property)": [[0, "genomicranges.SeqInfo.SeqInfo.genome"]], "genomicranges": [[0, "module-genomicranges"]], "genomicranges.genomicranges": [[0, "module-genomicranges.GenomicRanges"]], "genomicranges.genomicrangeslist": [[0, "module-genomicranges.GenomicRangesList"]], "genomicranges.seqinfo": [[0, "module-genomicranges.SeqInfo"]], "genomicranges.utils": [[0, "module-genomicranges.utils"]], "get_end() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_end"]], "get_genome() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.get_genome"]], "get_is_circular() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.get_is_circular"]], "get_mcols() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_mcols"]], "get_mcols() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.get_mcols"]], "get_metadata() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_metadata"]], "get_metadata() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.get_metadata"]], "get_names() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_names"]], "get_names() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.get_names"]], "get_ranges() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_ranges"]], "get_ranges() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.get_ranges"]], "get_seqinfo() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_seqinfo"]], "get_seqlengths() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.get_seqlengths"]], "get_seqnames() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_seqnames"]], "get_seqnames() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.get_seqnames"]], "get_start() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_start"]], "get_strand() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_strand"]], "get_subset() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_subset"]], "get_width() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.get_width"]], "groups() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.groups"]], "intersect() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.intersect"]], "invert_strand() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.invert_strand"]], "is_circular (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.is_circular"]], "is_circular (genomicranges.seqinfo.seqinfo property)": [[0, "genomicranges.SeqInfo.SeqInfo.is_circular"]], "is_empty() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.is_empty"]], "match() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.match"]], "mcols (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.mcols"]], "mcols (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.mcols"]], "merge_seqinfo() (in module genomicranges.seqinfo)": [[0, "genomicranges.SeqInfo.merge_SeqInfo"]], "metadata (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.metadata"]], "metadata (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.metadata"]], "module": [[0, "module-genomicranges"], [0, "module-genomicranges.GenomicRanges"], [0, "module-genomicranges.GenomicRangesList"], [0, "module-genomicranges.SeqInfo"], [0, "module-genomicranges.utils"], [1, "module-genomicranges.io"], [1, "module-genomicranges.io.gtf"], [1, "module-genomicranges.io.ucsc"]], "names (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.names"]], "names (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.names"]], "narrow() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.narrow"]], "nearest() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.nearest"]], "order() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.order"]], "precede() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.precede"]], "promoters() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.promoters"]], "range() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.range"]], "ranges (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.ranges"]], "ranges (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.ranges"]], "rank() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.rank"]], "reduce() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.reduce"]], "resize() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.resize"]], "restrict() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.restrict"]], "sample() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.sample"]], "sanitize_strand_vector() (in module genomicranges.utils)": [[0, "genomicranges.utils.sanitize_strand_vector"]], "seq_info (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.seq_info"]], "seqinfo (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.seqinfo"]], "seqlengths (genomicranges.seqinfo.seqinfo property)": [[0, "genomicranges.SeqInfo.SeqInfo.seqlengths"]], "seqnames (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.seqnames"]], "seqnames (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.seqnames"]], "seqnames (genomicranges.seqinfo.seqinfo property)": [[0, "genomicranges.SeqInfo.SeqInfo.seqnames"]], "set_genome() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.set_genome"]], "set_is_circular() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.set_is_circular"]], "set_mcols() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.set_mcols"]], "set_mcols() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.set_mcols"]], "set_metadata() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.set_metadata"]], "set_metadata() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.set_metadata"]], "set_names() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.set_names"]], "set_names() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.set_names"]], "set_ranges() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.set_ranges"]], "set_ranges() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.set_ranges"]], "set_seqinfo() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.set_seqinfo"]], "set_seqlengths() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.set_seqlengths"]], "set_seqnames() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.set_seqnames"]], "set_seqnames() (genomicranges.seqinfo.seqinfo method)": [[0, "genomicranges.SeqInfo.SeqInfo.set_seqnames"]], "set_strand() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.set_strand"]], "set_subset() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.set_subset"]], "setdiff() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.setdiff"]], "shift() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.shift"]], "slide_intervals() (in module genomicranges.utils)": [[0, "genomicranges.utils.slide_intervals"]], "sliding_windows() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.sliding_windows"]], "sort() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.sort"]], "split_intervals() (in module genomicranges.utils)": [[0, "genomicranges.utils.split_intervals"]], "start (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.start"]], "start (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.start"]], "strand (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.strand"]], "strand (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.strand"]], "subset_by_overlaps() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.subset_by_overlaps"]], "tile() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.tile"]], "tile_by_range() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.tile_by_range"]], "tile_genome() (genomicranges.genomicranges.genomicranges class method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.tile_genome"]], "to_pandas() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.to_pandas"]], "to_pandas() (genomicranges.genomicrangeslist.genomicrangeslist method)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.to_pandas"]], "trim() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.trim"]], "union() (genomicranges.genomicranges.genomicranges method)": [[0, "genomicranges.GenomicRanges.GenomicRanges.union"]], "width (genomicranges.genomicranges.genomicranges property)": [[0, "genomicranges.GenomicRanges.GenomicRanges.width"]], "width (genomicranges.genomicrangeslist.genomicrangeslist property)": [[0, "genomicranges.GenomicRangesList.GenomicRangesList.width"]], "access_gtf_ucsc() (in module genomicranges.io.ucsc)": [[1, "genomicranges.io.ucsc.access_gtf_ucsc"]], "genomicranges.io": [[1, "module-genomicranges.io"]], "genomicranges.io.gtf": [[1, "module-genomicranges.io.gtf"]], "genomicranges.io.ucsc": [[1, "module-genomicranges.io.ucsc"]], "parse_gtf() (in module genomicranges.io.gtf)": [[1, "genomicranges.io.gtf.parse_gtf"]], "read_gtf() (in module genomicranges.io.gtf)": [[1, "genomicranges.io.gtf.read_gtf"]], "read_ucsc() (in module genomicranges.io.ucsc)": [[1, "genomicranges.io.ucsc.read_ucsc"]]}})