<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>genomicranges.GenomicRanges &mdash; GenomicRanges 0.1 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> GenomicRanges
          </a>
              <div class="version">
                0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/modules.html">Module Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">GenomicRanges</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>genomicranges.GenomicRanges</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for genomicranges.GenomicRanges</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">ast</span> <span class="kn">import</span> <span class="n">Num</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Dict</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">ncls</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">.ucsc</span> <span class="kn">import</span> <span class="n">access_gtf_ucsc</span>
<span class="kn">from</span> <span class="nn">.gtf</span> <span class="kn">import</span> <span class="n">parse_gtf</span>
<span class="kn">from</span> <span class="nn">.utils</span> <span class="kn">import</span> <span class="n">split_pandas_df</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;jkanche&quot;</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s2">&quot;jkanche&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s2">&quot;MIT&quot;</span>


<div class="viewcode-block" id="GenomicRanges"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges">[docs]</a><span class="k">class</span> <span class="nc">GenomicRanges</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;GenomicRanges class to represent genomic locations&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">index</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="n">ncls</span><span class="o">.</span><span class="n">NCLS32</span><span class="p">,</span> <span class="n">ncls</span><span class="o">.</span><span class="n">NCLS64</span><span class="p">]],</span>
        <span class="n">indices</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Num</span><span class="p">],</span>
        <span class="n">ranges</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
        <span class="n">metadata</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Initialize an instance of GenomicRanges</span>

<span class="sd">        Args:</span>
<span class="sd">            index (Dict[str, Union[ncls.NCLS32, ncls.NCLS64]]): Nested Containment List index</span>
<span class="sd">            indices (List[Num]): indices that map between `index`, `ranges` and `metadata`</span>
<span class="sd">            ranges (pd.DataFrame): intervals as pandas dataframe</span>
<span class="sd">            metadata (Union[pd.DataFrame, None], optional): additional metadata as pandas dataframe. Defaults to None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">index</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">indices</span> <span class="o">=</span> <span class="n">indices</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span> <span class="o">=</span> <span class="n">ranges</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_iterIdx</span> <span class="o">=</span> <span class="mi">0</span>

<div class="viewcode-block" id="GenomicRanges.seqnames"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.seqnames">[docs]</a>    <span class="k">def</span> <span class="nf">seqnames</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;access sequences or chromosome names</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.Series: sequence information across all positions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span></div>

<div class="viewcode-block" id="GenomicRanges.ranges"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.ranges">[docs]</a>    <span class="k">def</span> <span class="nf">ranges</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;return the genomic positions</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: Dataframe containing the genomic positions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span></div>

<div class="viewcode-block" id="GenomicRanges.strand"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.strand">[docs]</a>    <span class="k">def</span> <span class="nf">strand</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;access strand information (if available)</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.Series: strand across all columns</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span></div>

<div class="viewcode-block" id="GenomicRanges.granges"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.granges">[docs]</a>    <span class="k">def</span> <span class="nf">granges</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;GenomicRanges object with only ranges (no metadata)</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: Genomic Ranges with only ranges</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">indices</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicRanges.mcols"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.mcols">[docs]</a>    <span class="k">def</span> <span class="nf">mcols</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Access metadata information about positions</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: metadata across all genomic positions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span></div>

<div class="viewcode-block" id="GenomicRanges.len"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.len">[docs]</a>    <span class="k">def</span> <span class="nf">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Num</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Number of intervals</span>

<span class="sd">        Returns:</span>
<span class="sd">            Num: number of genomic internals</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span></div>

<div class="viewcode-block" id="GenomicRanges.length"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.length">[docs]</a>    <span class="k">def</span> <span class="nf">length</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Num</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Number of intervals</span>

<span class="sd">        Returns:</span>
<span class="sd">            Num: number of genomic internals</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">len</span><span class="p">()</span></div>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Number of intervals</span>

<span class="sd">        Returns:</span>
<span class="sd">            Num: number of genomic internals</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">len</span><span class="p">()</span>

    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Iterator&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="fm">__next__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Get value from current iteration</span>

<span class="sd">        Raises:</span>
<span class="sd">            StopIteration: when Index is out of range</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_iterIdx</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="n">r</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_iterIdx</span><span class="p">]</span>
                <span class="n">m</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_iterIdx</span><span class="p">]</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span>
                    <span class="k">else</span> <span class="kc">None</span>
                <span class="p">)</span>
        <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_iterIdx</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">raise</span> <span class="ne">StopIteration</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_iterIdx</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_iterIdx</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">StopIteration</span><span class="p">()</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">m</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;string representation</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: what the instance holds</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;&lt;GenomicRanges&gt; contains </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="si">}</span><span class="s2"> chromosomes, </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">ranges</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> intervals&quot;</span>

<div class="viewcode-block" id="GenomicRanges.nearest"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.nearest">[docs]</a>    <span class="k">def</span> <span class="nf">nearest</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span> <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]]:</span>
        <span class="sd">&quot;&quot;&quot;Find nearest neighbors that overlap with the positions in `x`</span>

<span class="sd">        Args:</span>
<span class="sd">            x (GenomicRanges): input genomic positions to find</span>
<span class="sd">            k (int, optional): find k nearest neighbors ?. Defaults to 1.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[Union[int, List[int]]]: Possible hits</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">hits</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gr</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
            <span class="n">grhits</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">gr</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">find_overlap</span><span class="p">(</span>
                <span class="n">gr</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;starts&quot;</span><span class="p">],</span> <span class="n">gr</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">tmp_hits</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">grhits</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">counter</span> <span class="o">&lt;</span> <span class="n">k</span><span class="p">:</span>
                    <span class="n">tmp_hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>
            <span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_hits</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tmp_hits</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">hits</span></div>

<div class="viewcode-block" id="GenomicRanges.toDF"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.toDF">[docs]</a>    <span class="k">def</span> <span class="nf">toDF</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Export or Convert `GenomicRanges` to Pandas DataFrame</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: Pandas DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ranges</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">df</span></div>

<div class="viewcode-block" id="GenomicRanges.fromPandas"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.fromPandas">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">fromPandas</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Convert a pandas Dataframe to GenomicRanges</span>

<span class="sd">        Args:</span>
<span class="sd">            data (pd.DataFrame): a Pandas DataFrame object containing genomic positions.</span>
<span class="sd">                Must contain `seqname`, `start` &amp; `end` columns.</span>

<span class="sd">        Raises:</span>
<span class="sd">            Exception: Validation Error</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: An Object representing genomic positions</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># validation:</span>
        <span class="c1"># if `seqname`, `start` and `end` don&#39;t exist, abort!</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;seqname&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="p">):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;DataFrame does not contain columns: `seqname`, `start` and `end`&quot;</span>
            <span class="p">)</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;DataFrame does not contain columns: `seqname`, `start` and `end`&quot;</span>
            <span class="p">)</span>

        <span class="p">(</span><span class="n">indexes</span><span class="p">,</span> <span class="n">indices</span><span class="p">,</span> <span class="n">ranges</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span> <span class="o">=</span> <span class="n">split_pandas_df</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span><span class="n">indexes</span><span class="p">,</span> <span class="n">indices</span><span class="p">,</span> <span class="n">ranges</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicRanges.fromGTF"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.fromGTF">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">fromGTF</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Load a GTF file as GenomicRanges</span>

<span class="sd">        Args:</span>
<span class="sd">            file (str): path to gtf file</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges:  An Object representing genomic positions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">compressed</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;gz&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="kc">False</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">parse_gtf</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="n">compressed</span><span class="p">)</span>

        <span class="p">(</span><span class="n">indexes</span><span class="p">,</span> <span class="n">indices</span><span class="p">,</span> <span class="n">ranges</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span> <span class="o">=</span> <span class="n">split_pandas_df</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span><span class="n">indexes</span><span class="p">,</span> <span class="n">indices</span><span class="p">,</span> <span class="n">ranges</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span></div>

<div class="viewcode-block" id="GenomicRanges.fromUCSC"><a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.fromUCSC">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">fromUCSC</span><span class="p">(</span><span class="n">genome</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;refGene&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Load a GTF file from UCSC as GenomicRanges</span>

<span class="sd">        Args:</span>
<span class="sd">            genome (str): genome shortcode; e.g. hg19, hg38, mm10 etc</span>
<span class="sd">            type (str): One of refGene, ensGene, knownGene or ncbiRefSeq</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges:  An Object representing genomic positions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">path</span> <span class="o">=</span> <span class="n">access_gtf_ucsc</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">type</span><span class="p">)</span>
        <span class="n">compressed</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">parse_gtf</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="n">compressed</span><span class="p">)</span>

        <span class="p">(</span><span class="n">indexes</span><span class="p">,</span> <span class="n">indices</span><span class="p">,</span> <span class="n">ranges</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span> <span class="o">=</span> <span class="n">split_pandas_df</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span><span class="n">indexes</span><span class="p">,</span> <span class="n">indices</span><span class="p">,</span> <span class="n">ranges</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span></div></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, jkanche.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>