<!doctype html>
<html class="no-js" lang="en" data-content_root="../../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" />

    <!-- Generated with Sphinx 7.2.5 and Furo 2023.08.19 -->
        <title>genomicranges.GenomicRanges - GenomicRanges 0.3.2 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">GenomicRanges 0.3.2 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  
  <span class="sidebar-brand-text">GenomicRanges 0.3.2 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#construct-a-genomicranges-object">Construct a <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#get-and-set-methods">Get and Set methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#slice-operations">Slice operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#iterate-over-rows">Iterate over rows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#intra-range-transformations">Intra-range transformations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#inter-range-methods">Inter-range methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#set-operations-on-genomic-ranges">Set operations on genomic ranges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#compute-over-bins">Compute over bins</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#overlap-based-methods">Overlap based methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#search-operations">Search operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#comparison-rank-and-order-operations">Comparison, rank and order operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#concatenate-genomicranges-objects-by-rows">Concatenate <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> objects by rows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#misc-operations">Misc operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html#construct-a-genomicrangeslist-object">Construct a <code class="docutils literal notranslate"><span class="pre">GenomicRangesList</span></code> object.</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/modules.html">Module Reference</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/genomicranges.html">genomicranges package</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of genomicranges package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/genomicranges.io.html">genomicranges.io package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <h1>Source code for genomicranges.GenomicRanges</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">Any</span><span class="p">,</span>
    <span class="n">Callable</span><span class="p">,</span>
    <span class="n">Dict</span><span class="p">,</span>
    <span class="n">List</span><span class="p">,</span>
    <span class="n">Literal</span><span class="p">,</span>
    <span class="n">Mapping</span><span class="p">,</span>
    <span class="n">MutableMapping</span><span class="p">,</span>
    <span class="n">Optional</span><span class="p">,</span>
    <span class="n">Sequence</span><span class="p">,</span>
    <span class="n">Tuple</span><span class="p">,</span>
    <span class="n">Union</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">warnings</span> <span class="kn">import</span> <span class="n">warn</span>

<span class="kn">from</span> <span class="nn">biocframe</span> <span class="kn">import</span> <span class="n">BiocFrame</span>
<span class="kn">from</span> <span class="nn">numpy</span> <span class="kn">import</span> <span class="n">concatenate</span><span class="p">,</span> <span class="n">count_nonzero</span><span class="p">,</span> <span class="n">ndarray</span><span class="p">,</span> <span class="nb">sum</span><span class="p">,</span> <span class="n">zeros</span>
<span class="kn">from</span> <span class="nn">pandas</span> <span class="kn">import</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">concat</span><span class="p">,</span> <span class="n">isna</span>
<span class="kn">from</span> <span class="nn">prettytable</span> <span class="kn">import</span> <span class="n">PrettyTable</span>

<span class="kn">from</span> <span class="nn">.io</span> <span class="kn">import</span> <span class="n">from_pandas</span>
<span class="kn">from</span> <span class="nn">.SeqInfo</span> <span class="kn">import</span> <span class="n">SeqInfo</span>
<span class="kn">from</span> <span class="nn">.utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">OVERLAP_QUERY_TYPES</span><span class="p">,</span>
    <span class="n">calc_row_gapwidth</span><span class="p">,</span>
    <span class="n">compute_mean</span><span class="p">,</span>
    <span class="n">create_np_interval_vector</span><span class="p">,</span>
    <span class="n">find_diff</span><span class="p">,</span>
    <span class="n">find_disjoin</span><span class="p">,</span>
    <span class="n">find_gaps</span><span class="p">,</span>
    <span class="n">find_intersect</span><span class="p">,</span>
    <span class="n">find_nearest</span><span class="p">,</span>
    <span class="n">find_overlaps</span><span class="p">,</span>
    <span class="n">find_union</span><span class="p">,</span>
    <span class="n">slide_intervals</span><span class="p">,</span>
    <span class="n">split_intervals</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;jkanche&quot;</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s2">&quot;jkanche&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s2">&quot;MIT&quot;</span>


<div class="viewcode-block" id="GenomicRanges">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges">[docs]</a>
<span class="k">class</span> <span class="nc">GenomicRanges</span><span class="p">(</span><span class="n">BiocFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;``GenomicRanges`` provides functionality to represent and operate over genomic regions and annotations.</span>

<span class="sd">    **Note: Intervals are inclusive on both ends and start at 1.**</span>

<span class="sd">    Additionally, ``GenomicRanges`` may also contain `Sequence Information` (checkout</span>
<span class="sd">    :py:class:`~genomicranges.SeqInfo.SeqInfo`) as part of its metadata. It contains for each</span>
<span class="sd">    sequence name (or chromosome) in the gene model, its length. Additionally, (checkout</span>
<span class="sd">    :py:class:`~genomicranges.SeqInfo.SeqInfo`) might also contain metadata about the</span>
<span class="sd">    genome, e.g. if its circular (`is_circular`) or not.</span>

<span class="sd">    Note: The documentation for some of the methods come from the</span>
<span class="sd">    `GenomicRanges R/Bioconductor package &lt;https://github.com/Bioconductor/GenomicRanges&gt;`_.</span>

<span class="sd">    Typical usage example:</span>

<span class="sd">    To construct a **GenomicRanges** object, simply pass in the column representation as a</span>
<span class="sd">    dictionary. This dictionary must contain &quot;seqnames&quot;, &quot;starts&quot;, &quot;ends&quot; columns and optionally</span>
<span class="sd">    specify &quot;strand&quot;. If &quot;strand&quot; column is not provided, &quot;*&quot; is used as the default value for</span>
<span class="sd">    each genomic interval.</span>

<span class="sd">    .. code-block:: python</span>

<span class="sd">        gr = GenomicRanges(</span>
<span class="sd">            {</span>
<span class="sd">                &quot;seqnames&quot;: [&quot;chr1&quot;, &quot;chr2&quot;, &quot;chr3&quot;],</span>
<span class="sd">                &quot;starts&quot;: [100, 115, 119],</span>
<span class="sd">                &quot;ends&quot;: [103, 116, 120],</span>
<span class="sd">            }</span>
<span class="sd">        )</span>

<span class="sd">    Alternatively, you may also convert a :py:class:`~pandas.DataFrame` to ``GenomicRanges``.</span>

<span class="sd">    .. code-block:: python</span>

<span class="sd">        df = pd.DataFrame(</span>
<span class="sd">            {</span>
<span class="sd">                &quot;seqnames&quot;: [&quot;chr1&quot;, &quot;chr2&quot;, &quot;chr3&quot;],</span>
<span class="sd">                &quot;starts&quot;: [100, 115, 119],</span>
<span class="sd">                &quot;ends&quot;: [103, 116, 120],</span>
<span class="sd">            }</span>
<span class="sd">        )</span>

<span class="sd">        gr = genomicranges.from_pandas(df)</span>

<span class="sd">    All columns other than &quot;seqnames&quot;, &quot;starts&quot;, &quot;ends&quot; and &quot;strand&quot; are considered</span>
<span class="sd">    metadata columns and can be accessed by</span>
<span class="sd">    :py:meth:`~genomicranges.GenomicRanges.GenomicRanges.mcols`.</span>

<span class="sd">    .. code-block:: python</span>

<span class="sd">        gr.mcols()</span>

<span class="sd">    or slice the object</span>

<span class="sd">    .. code-block:: python</span>

<span class="sd">        sliced_gr = gr[1:2, [True, False, False]]</span>

<span class="sd">    Attributes:</span>
<span class="sd">        data (Mapping[str, Union[List[Any], Mapping]]):</span>
<span class="sd">            Columns as dictionary, Must contain &quot;seqnames&quot;, &quot;starts&quot;, &quot;ends&quot; and &quot;strand&quot; columns.</span>
<span class="sd">        number_of_rows (int, optional): Number of genomic intervals (or rows). Defaults to None.</span>
<span class="sd">        row_names (Sequence[str], optional): Row index. Defaults to None.</span>
<span class="sd">        column_names (Sequence[str], optional): Column names, automatically inferred from `data`.</span>
<span class="sd">            Defaults to None.</span>
<span class="sd">        metadata (Mapping, optional): Additional metadata. Defaults to None.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">required_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">data</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="n">Sequence</span><span class="p">,</span> <span class="n">Mapping</span><span class="p">]],</span>
        <span class="n">number_of_rows</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">row_names</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Sequence</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">column_names</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Sequence</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">metadata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Mapping</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initialize a `GenomicRanges` object.&quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">number_of_rows</span><span class="p">,</span> <span class="n">row_names</span><span class="p">,</span> <span class="n">column_names</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_validate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Internal function to validate ``GenomicRanges``.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;strand&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;*&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">])</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">)</span>

        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">_validate</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_ranges</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_validate_ranges</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Internal function to validate all columns of ``GenomicRanges``.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If missing required columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">missing</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">required_columns</span><span class="p">)</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">)))</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">missing</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;`data` must contain </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">required_columns</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="sa">f</span><span class="s2">&quot;missing </span><span class="si">{</span><span class="n">missing</span><span class="si">}</span><span class="s2"> column</span><span class="si">{</span><span class="s1">&#39;s&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">missing</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mi">1</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">seqnames</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get sequence or chromosome names.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: List of all chromosome names.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">start</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get sequence or chromosome start positions.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[int]: List of all chromosome start positions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">end</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get sequence or chromosome end positions.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[int]: List of all chromosome end positions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="GenomicRanges.ranges">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.ranges">[docs]</a>
    <span class="k">def</span> <span class="nf">ranges</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">return_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Callable</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get genomic positions.</span>

<span class="sd">        Args:</span>
<span class="sd">            ignore_strand (bool): Whether to ignore strands. Defaults to False.</span>
<span class="sd">            return_type (Callable, optional): Format to return genomic positions.</span>
<span class="sd">                Defaults to a dictionary representation, supports `DataFrame`</span>
<span class="sd">                or any callable representation that takes a dictionary as an input.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If ``return_type`` is not supported.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Union[DataFrame, MutableMapping, &quot;GenomicRanges&quot;, Any]: Genomic regions.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">obj</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">),</span>
            <span class="s2">&quot;starts&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">),</span>
            <span class="s2">&quot;ends&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">),</span>
            <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">),</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="n">ignore_strand</span><span class="p">:</span>
            <span class="n">obj</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;*&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">obj</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">return_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">obj</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">return_type</span><span class="p">(</span><span class="n">obj</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">return_type</span><span class="si">}</span><span class="s2"> is not supported, </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">strand</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get strand information.</span>

<span class="sd">        If ``strand`` is not provided, we use &#39;*&#39; as the default value for each interval.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: Strand across all positions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">width</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get widths of each interval.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[int]: Widths of each interval.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">widths</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="n">widths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">widths</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">seq_info</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqInfo</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get sequence information, if available.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (SeqInfo, optional): Sequence information.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="ow">and</span> <span class="s2">&quot;seq_info&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;seq_info&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@seq_info</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">seq_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_info</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqInfo</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set sequence information.</span>

<span class="sd">        Args:</span>
<span class="sd">            (SeqInfo, optional): Sequence information.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If `seq_info` is not a `SeqInfo` class.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">seq_info</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seq_info</span><span class="p">,</span> <span class="n">SeqInfo</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;seq_info is not a `SeqInfo` class.&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;seq_info&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_info</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">seqlengths</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get length of each chromosome, if available.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (Dict[str, int], optional): A dictionary where keys are chromosome names and values</span>
<span class="sd">            specify lengths of each chromsome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="s2">&quot;seq_info&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;seq_info&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">seqlengths</span>

        <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">score</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Sequence</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get &quot;score&quot; column (if available) for each genomic interval.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (Sequence[Union[int, float]], optional): Score column.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="s2">&quot;score&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;score&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@score</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">score</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set score for each position.</span>

<span class="sd">        Args:</span>
<span class="sd">            score (Sequence[Union[int, float]]): Score values to set.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If length of ``score`` does not</span>
<span class="sd">                match the number of intervals in the object.</span>
<span class="sd">            TypeError: If `score` is not a list.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">score</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`score` must be a list!&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">score</span><span class="p">)</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Length of `score` must be the same as number of intervals&quot;</span>
                <span class="sa">f</span><span class="s2">&quot;must be </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">, but provided </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">score</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="p">[</span><span class="s2">&quot;score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">score</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">is_circular</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Whether the sequences/chromosomes are circular (only if available).</span>

<span class="sd">        Returns:</span>
<span class="sd">            (Dict[str, bool], optional): A dictionary with chromosome names as keys and a</span>
<span class="sd">            boolean value indicating if its circular or not.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="s2">&quot;seqInfo&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;seqInfo&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">is_circular</span>

        <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genome</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get genome information, if available.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (str, optional): Genome information if available else None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_metadata</span> <span class="ow">and</span> <span class="s2">&quot;seqInfo&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_metadata</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_metadata</span><span class="p">[</span><span class="s2">&quot;seqInfo&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">genome</span>

        <span class="k">return</span> <span class="kc">None</span>

<div class="viewcode-block" id="GenomicRanges.granges">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.granges">[docs]</a>
    <span class="k">def</span> <span class="nf">granges</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Creates a new ``GenomicRanges`` object with only ranges (`seqnames`, `starts, `ends` and `strand`).</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` with only ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">),</span>
                <span class="s2">&quot;starts&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">),</span>
                <span class="s2">&quot;ends&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">),</span>
                <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">),</span>
            <span class="p">}</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.mcols">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.mcols">[docs]</a>
    <span class="k">def</span> <span class="nf">mcols</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">return_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Callable</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">MutableMapping</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get metadata across all genomic intervals.</span>

<span class="sd">        All columns other than &quot;seqnames&quot;, &quot;starts&quot;, &quot;ends&quot; and `&quot;strand&quot;</span>
<span class="sd">        are considered metadata for each interval.</span>

<span class="sd">        Args:</span>
<span class="sd">            return_type (Callable, optional): Format to return metadata.</span>
<span class="sd">                Defaults to dictionary representation, supports `DataFrame`</span>
<span class="sd">                or any callable representation that takes a dictionary as an input.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If ``return_type`` is not supported.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Union[DataFrame, MutableMapping, Any]: Metadata columns without</span>
<span class="sd">            genomic positions.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">required_columns</span><span class="p">:</span>
                <span class="n">new_data</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">return_type</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">new_data</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">return_type</span><span class="p">(</span><span class="n">new_data</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">return_type</span><span class="si">}</span><span class="s2"> not supported, </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.__repr__">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.__repr__">[docs]</a>
    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">PrettyTable</span><span class="p">(</span><span class="n">padding_width</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">table</span><span class="o">.</span><span class="n">field_names</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">col</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">]</span>

        <span class="n">_rows</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">rows_to_show</span> <span class="o">=</span> <span class="mi">2</span>
        <span class="n">_top</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">_top</span> <span class="o">&gt;</span> <span class="n">rows_to_show</span><span class="p">:</span>
            <span class="n">_top</span> <span class="o">=</span> <span class="n">rows_to_show</span>

        <span class="c1"># top three rows</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">_top</span><span class="p">):</span>
            <span class="n">_row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">row</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">_row</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
            <span class="n">res</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">vals</span><span class="p">]</span>
            <span class="n">_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">rows_to_show</span><span class="p">:</span>
            <span class="c1"># add ...</span>
            <span class="n">_rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s2">&quot;...&quot;</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">))])</span>

        <span class="n">_last</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">rows_to_show</span>
        <span class="k">if</span> <span class="n">_last</span> <span class="o">&lt;=</span> <span class="n">rows_to_show</span><span class="p">:</span>
            <span class="n">_last</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">_top</span>

        <span class="c1"># last three rows</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">_last</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)):</span>
            <span class="n">_row</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">row</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">_row</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
            <span class="n">res</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">vals</span><span class="p">]</span>
            <span class="n">_rows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>

        <span class="n">table</span><span class="o">.</span><span class="n">add_rows</span><span class="p">(</span><span class="n">_rows</span><span class="p">)</span>

        <span class="n">pattern</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Class GenomicRanges with </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">dims</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> intervals and &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">dims</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="mi">4</span><span class="si">}</span><span class="s2"> metadata columns </span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;contains row names?: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="w"> </span><span class="ow">is</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="kc">None</span><span class="si">}</span><span class="s2"> </span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">table</span><span class="o">.</span><span class="n">get_string</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">pattern</span></div>


<div class="viewcode-block" id="GenomicRanges.__getitem__">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.__getitem__">[docs]</a>
    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">args</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Sequence</span><span class="p">,</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Sequence</span><span class="p">]]]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Subset the object.</span>

<span class="sd">        This operation returns a new object with the same type as caller.</span>
<span class="sd">        If you need to access specific rows or columns, use the</span>
<span class="sd">        :py:meth:`~biocframe.BiocFrame.BiocFrame.row` or</span>
<span class="sd">        :py:meth:`~biocframe.BiocFrame.BiocFrame.column`</span>
<span class="sd">        methods.</span>

<span class="sd">        Usage:</span>

<span class="sd">        .. code-block:: python</span>

<span class="sd">            # made up chromosome locations and ensembl ids.</span>
<span class="sd">            obj = {</span>
<span class="sd">                &quot;ensembl&quot;: [&quot;ENS00001&quot;, &quot;ENS00002&quot;, &quot;ENS00002&quot;],</span>
<span class="sd">                &quot;symbol&quot;: [&quot;MAP1A&quot;, &quot;BIN1&quot;, &quot;ESR1&quot;],</span>
<span class="sd">                &quot;ranges&quot;: BiocFrame({</span>
<span class="sd">                    &quot;chr&quot;: [&quot;chr1&quot;, &quot;chr2&quot;, &quot;chr3&quot;]</span>
<span class="sd">                    &quot;start&quot;: [1000, 1100, 5000],</span>
<span class="sd">                    &quot;end&quot;: [1100, 4000, 5500]</span>
<span class="sd">                ),</span>
<span class="sd">            }</span>
<span class="sd">            gr = GenomicRanges(obj)</span>

<span class="sd">            # different ways to slice.</span>
<span class="sd">            gr[0:2, 0:2]</span>
<span class="sd">            gr[[0,2], [True, False, False]]</span>
<span class="sd">            gr[&lt;List of column names&gt;]</span>

<span class="sd">        Args:</span>
<span class="sd">            args (SlicerArgTypes): A Tuple of slicer arguments to subset rows and</span>
<span class="sd">                columns. An element in ``args`` may be,</span>

<span class="sd">                - List of booleans, True to keep the row/column, False to remove.</span>
<span class="sd">                    The length of the boolean vector must be the same as number of rows/columns.</span>

<span class="sd">                - List of integer positions along rows/columns to keep.</span>

<span class="sd">                - A :py:class:`~slice` object specifying the list of positions to keep.</span>

<span class="sd">                - A list of index names to keep. For rows, the object must contain unique</span>
<span class="sd">                    :py:attr:`~biocframe.BiocFrame.BiocFrame.row_names` and for columns must</span>
<span class="sd">                    contain unique :py:attr:`~biocframe.BiocFrame.BiocFrame.column_names`.</span>

<span class="sd">                - A scalar integer to subset either a single row or column position.</span>
<span class="sd">                    Alternatively, you might want to use</span>
<span class="sd">                    :py:meth:`~biocframe.BiocFrame.BiocFrame.row` or</span>
<span class="sd">                    :py:meth:`~biocframe.BiocFrame.BiocFrame.column` methods.</span>

<span class="sd">                - A singular string to subset either a single row or column label/index.</span>
<span class="sd">                    Alternatively, you might want to use</span>
<span class="sd">                    :py:meth:`~biocframe.BiocFrame.BiocFrame.row` or</span>
<span class="sd">                    :py:meth:`~biocframe.BiocFrame.BiocFrame.column` methods.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: Too many slices provided.</span>
<span class="sd">            TypeError: If provided ``args`` are not an expected type.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The same type as caller with the subsetted rows and columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__getitem__</span><span class="p">(</span><span class="n">args</span><span class="p">)</span></div>


    <span class="c1"># intra-range methods</span>
<div class="viewcode-block" id="GenomicRanges.flank">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.flank">[docs]</a>
    <span class="k">def</span> <span class="nf">flank</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">width</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">start</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">both</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compute flanking ranges for each range. The logic for this comes from the `R/GenomicRanges` &amp; `IRanges`</span>
<span class="sd">        packages.</span>

<span class="sd">        If ``start`` is ``True`` for a given range, the flanking occurs at the `start`,</span>
<span class="sd">        otherwise the `end`.</span>
<span class="sd">        The `widths` of the flanks are given by the ``width`` parameter.</span>

<span class="sd">        ``width`` can be negative, in which case the flanking region is</span>
<span class="sd">        reversed so that it represents a prefix or suffix of the range.</span>

<span class="sd">        Usage:</span>

<span class="sd">            `gr.flank(3, True)`, where &quot;x&quot; indicates a range in ``gr`` and &quot;-&quot; indicates the</span>
<span class="sd">            resulting flanking region:</span>
<span class="sd">                ---xxxxxxx</span>
<span class="sd">            If ``start`` were ``False``, the range in ``gr`` becomes</span>
<span class="sd">                xxxxxxx---</span>
<span class="sd">            For negative width, i.e. `gr.flank(x, -3, FALSE)`, where &quot;*&quot; indicates the overlap</span>
<span class="sd">            between &quot;x&quot; and the result:</span>
<span class="sd">                xxxx***</span>
<span class="sd">            If ``both`` is ``True``, then, for all ranges in &quot;x&quot;, the flanking regions are</span>
<span class="sd">            extended into (or out of, if ``width`` is negative) the range, so that the result</span>
<span class="sd">            straddles the given endpoint and has twice the width given by width.</span>

<span class="sd">            This is illustrated below for `gr.flank(3, both=TRUE)`:</span>
<span class="sd">                ---***xxxx</span>

<span class="sd">        Args:</span>
<span class="sd">            width (int): Width to flank by. May be negative.</span>
<span class="sd">            start (bool, optional): Whether to only flank starts. Defaults to True.</span>
<span class="sd">            both (bool, optional): Whether to flank both starts and ends. Defaults to False.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new :py:class:`~.GenomicRanges` object with the flanked ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">new_starts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_ends</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">all_starts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">)</span>
        <span class="n">all_ends</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">)</span>
        <span class="n">all_strands</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">)</span>

        <span class="c1"># figure out which position to pin, start or end?</span>
        <span class="n">start_flags</span> <span class="o">=</span> <span class="p">[</span><span class="n">start</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_strands</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">ignore_strand</span><span class="p">:</span>
            <span class="n">start_flags</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">start</span> <span class="o">!=</span> <span class="p">(</span><span class="n">all_strands</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_strands</span><span class="p">))</span>
            <span class="p">]</span>

        <span class="c1"># if both is true, then depending on start_flag, we extend it out</span>
        <span class="c1"># I couldn&#39;t understand the scenario&#39;s with witdh &lt;=0,</span>
        <span class="c1"># so refer to the R implementation here</span>
        <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">start_flags</span><span class="p">)):</span>
            <span class="n">sf</span> <span class="o">=</span> <span class="n">start_flags</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="n">tstart</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">both</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">tstart</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">all_starts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">-</span> <span class="nb">abs</span><span class="p">(</span><span class="n">width</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">sf</span>
                    <span class="k">else</span> <span class="n">all_ends</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">-</span> <span class="nb">abs</span><span class="p">(</span><span class="n">width</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">width</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">tstart</span> <span class="o">=</span> <span class="n">all_starts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">-</span> <span class="nb">abs</span><span class="p">(</span><span class="n">width</span><span class="p">)</span> <span class="k">if</span> <span class="n">sf</span> <span class="k">else</span> <span class="n">all_ends</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">tstart</span> <span class="o">=</span> <span class="n">all_starts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">if</span> <span class="n">sf</span> <span class="k">else</span> <span class="n">all_ends</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">+</span> <span class="n">width</span> <span class="o">+</span> <span class="mi">1</span>

            <span class="n">new_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tstart</span><span class="p">)</span>
            <span class="n">new_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tstart</span> <span class="o">+</span> <span class="p">(</span><span class="n">width</span> <span class="o">*</span> <span class="p">(</span><span class="mi">2</span> <span class="k">if</span> <span class="n">both</span> <span class="k">else</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_starts</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_ends</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="n">new_data</span><span class="p">,</span>
            <span class="n">number_of_rows</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">row_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="p">,</span>
            <span class="n">column_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.resize">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.resize">[docs]</a>
    <span class="k">def</span> <span class="nf">resize</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">width</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">fix</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Resize ranges to the specified ``width`` where either the ``start``, ``end``, or ``center`` is used as an</span>
<span class="sd">        anchor.</span>

<span class="sd">        Args:</span>
<span class="sd">            width (int): Width to resize, cannot be negative!</span>
<span class="sd">            fix (Literal[&quot;start&quot;, &quot;end&quot;, &quot;center&quot;], optional): Fix positions by</span>
<span class="sd">                &quot;start&quot;, &quot;end&quot;, &quot;center&quot;. Defaults to &quot;start&quot;.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If parameter ``fix`` is neither `start`, `end` or `center`.</span>
<span class="sd">            ValueError: If ``width`` is negative.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` object with the resized ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">width</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`width` cannot be negative!&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">fix</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;center&quot;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;`fix` must be either &#39;start&#39;, &#39;end&#39; or &#39;center&#39;, provided </span><span class="si">{</span><span class="n">fix</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="n">new_starts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_ends</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="n">ts</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">te</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="k">if</span> <span class="n">ignore_strand</span> <span class="ow">is</span> <span class="kc">True</span> <span class="ow">or</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">fix</span> <span class="o">==</span> <span class="s2">&quot;start&quot;</span><span class="p">:</span>
                    <span class="n">ts</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span>
                    <span class="n">te</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">width</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">fix</span> <span class="o">==</span> <span class="s2">&quot;center&quot;</span><span class="p">:</span>
                    <span class="n">tmid</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                    <span class="n">twidthby2</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">width</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span>
                        <span class="k">else</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">width</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="n">ts</span> <span class="o">=</span> <span class="n">tmid</span> <span class="o">-</span> <span class="n">twidthby2</span>
                    <span class="n">te</span> <span class="o">=</span> <span class="n">ts</span> <span class="o">+</span> <span class="n">width</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">te</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span>
                    <span class="n">ts</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">width</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">fix</span> <span class="o">==</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span>
                    <span class="n">ts</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span>
                    <span class="n">te</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">width</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">fix</span> <span class="o">==</span> <span class="s2">&quot;center&quot;</span><span class="p">:</span>
                    <span class="n">tmid</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                    <span class="n">twidthby2</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">width</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                    <span class="n">ts</span> <span class="o">=</span> <span class="n">tmid</span> <span class="o">-</span> <span class="n">twidthby2</span>
                    <span class="n">te</span> <span class="o">=</span> <span class="n">ts</span> <span class="o">+</span> <span class="n">width</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">te</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span>
                    <span class="n">ts</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">width</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;strand must be either +, - or *, contains &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> at index: </span><span class="si">{</span><span class="n">idx</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

            <span class="n">new_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
            <span class="n">new_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">te</span><span class="p">)</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_starts</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_ends</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="n">new_data</span><span class="p">,</span>
            <span class="n">number_of_rows</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">row_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="p">,</span>
            <span class="n">column_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.shift">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.shift">[docs]</a>
    <span class="k">def</span> <span class="nf">shift</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">shift</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Shift all intervals by parameter ``shift``.</span>

<span class="sd">        Args:</span>
<span class="sd">            shift (int, optional): Shift interval. Defaults to 0.</span>
<span class="sd">                If shift is 0, the current object is returned.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` object with the shifted ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">shift</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="n">all_starts</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">shift</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">)]</span>
        <span class="n">all_ends</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">shift</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">)]</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_starts</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_ends</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="n">new_data</span><span class="p">,</span>
            <span class="n">number_of_rows</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">row_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="p">,</span>
            <span class="n">column_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.promoters">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.promoters">[docs]</a>
    <span class="k">def</span> <span class="nf">promoters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">upstream</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">downstream</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extend intervals to promoter regions.</span>

<span class="sd">        Generates promoter ranges relative to the transcription start site (TSS),</span>
<span class="sd">        where TSS is start(x). The promoter range is expanded around the TSS</span>
<span class="sd">        according to the upstream and downstream arguments. Upstream represents</span>
<span class="sd">        the number of nucleotides in the 5&#39; direction and downstream the number</span>
<span class="sd">        in the 3&#39; direction. The full range is defined as, (`start(x) - upstream`)</span>
<span class="sd">        to (`start(x) + downstream - 1`).</span>

<span class="sd">        Args:</span>
<span class="sd">            upstream (int, optional): Number of positions to extend in the 5&#39; direction.</span>
<span class="sd">                Defaults to 2000.</span>
<span class="sd">            downstream (int, optional): Number of positions to extend in the 3&#39; direction.</span>
<span class="sd">                Defaults to 200.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` object with ranges replaced</span>
<span class="sd">            by promoter regions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_starts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">)</span>
        <span class="n">all_ends</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">)</span>
        <span class="n">all_strands</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">)</span>

        <span class="n">start_flags</span> <span class="o">=</span> <span class="p">[</span><span class="n">all_strands</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_strands</span><span class="p">))]</span>

        <span class="n">new_starts</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">(</span>
                <span class="n">all_starts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">-</span> <span class="n">upstream</span>
                <span class="k">if</span> <span class="n">start_flags</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
                <span class="k">else</span> <span class="n">all_ends</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">-</span> <span class="n">downstream</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">start_flags</span><span class="p">))</span>
        <span class="p">]</span>
        <span class="n">new_ends</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">(</span>
                <span class="n">all_starts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">+</span> <span class="n">downstream</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">start_flags</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
                <span class="k">else</span> <span class="n">all_ends</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">+</span> <span class="n">upstream</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">start_flags</span><span class="p">))</span>
        <span class="p">]</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_starts</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_ends</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="n">new_data</span><span class="p">,</span>
            <span class="n">number_of_rows</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">row_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="p">,</span>
            <span class="n">column_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.restrict">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.restrict">[docs]</a>
    <span class="k">def</span> <span class="nf">restrict</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">start</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">end</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">keep_all_ranges</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Restrict ranges to a given start and end positions.</span>

<span class="sd">        Args:</span>
<span class="sd">            start (int, optional): Start position. Defaults to None.</span>
<span class="sd">            end (int, optional): End position. Defaults to None.</span>
<span class="sd">            keep_all_ranges (bool, optional): Whether to keep intervals that do</span>
<span class="sd">                not overlap with start and end. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` object with restricted ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_starts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">)</span>
        <span class="n">all_ends</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">)</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">new_starts</span> <span class="o">=</span> <span class="n">all_starts</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
            <span class="n">new_starts</span> <span class="o">=</span> <span class="p">[(</span><span class="n">start</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&lt;</span> <span class="n">start</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_starts</span><span class="p">]</span>

        <span class="n">new_ends</span> <span class="o">=</span> <span class="n">all_ends</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">end</span><span class="p">:</span>
            <span class="n">new_ends</span> <span class="o">=</span> <span class="p">[(</span><span class="n">end</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="n">end</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_ends</span><span class="p">]</span>

        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_starts</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_ends</span>

        <span class="n">new_row_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">keep_all_ranges</span><span class="p">:</span>
            <span class="n">keepers</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="c1"># find intervals indices that overlap with the given start and end</span>
            <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_starts</span><span class="p">)):</span>
                <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">if</span> <span class="n">start</span> <span class="ow">and</span> <span class="p">(</span><span class="n">all_starts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">start</span> <span class="o">&lt;=</span> <span class="n">all_ends</span><span class="p">[</span><span class="n">idx</span><span class="p">]):</span>
                    <span class="n">keep</span> <span class="o">=</span> <span class="kc">True</span>

                <span class="k">if</span> <span class="n">end</span> <span class="ow">and</span> <span class="p">(</span><span class="n">all_starts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">end</span> <span class="o">&lt;=</span> <span class="n">all_ends</span><span class="p">[</span><span class="n">idx</span><span class="p">]):</span>
                    <span class="n">keep</span> <span class="o">=</span> <span class="kc">True</span>

                <span class="k">if</span> <span class="n">keep</span><span class="p">:</span>
                    <span class="n">keepers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">:</span>
                <span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">][</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">keepers</span><span class="p">]</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="p">:</span>
                <span class="n">new_row_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">new_row_names</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">keepers</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="n">new_data</span><span class="p">,</span>
            <span class="n">row_names</span><span class="o">=</span><span class="n">new_row_names</span><span class="p">,</span>
            <span class="n">column_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.trim">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.trim">[docs]</a>
    <span class="k">def</span> <span class="nf">trim</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Trim sequences outside of bounds for non-circular chromosomes.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` object with trimmed ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_info</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot trim ranges. `seqinfo` is not available.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot trim ranges. `seqinfo` is not available.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;seq_info&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;seq_info&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot trim ranges. `seqinfo` is not available.&quot;</span><span class="p">)</span>

        <span class="n">seqinfos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_info</span>
        <span class="n">seqlengths</span> <span class="o">=</span> <span class="n">seqinfos</span><span class="o">.</span><span class="n">seqlengths</span>
        <span class="n">is_circular</span> <span class="o">=</span> <span class="n">seqinfos</span><span class="o">.</span><span class="n">is_circular</span>

        <span class="k">if</span> <span class="n">seqlengths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot trim ranges. `seqlengths` is not available.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">is_circular</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">warn</span><span class="p">(</span><span class="s2">&quot;considering all sequences as non-circular...&quot;</span><span class="p">)</span>

        <span class="n">all_chrs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">)</span>
        <span class="n">all_ends</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">)</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">new_row_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">keepers</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_chrs</span><span class="p">)):</span>
            <span class="n">keep</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">t_chr</span> <span class="o">=</span> <span class="n">all_chrs</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">is_circular</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                <span class="ow">and</span> <span class="n">is_circular</span><span class="p">[</span><span class="n">t_chr</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">False</span>
                <span class="ow">and</span> <span class="n">all_ends</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">seqlengths</span><span class="p">[</span><span class="n">t_chr</span><span class="p">]</span>
            <span class="p">):</span>
                <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="k">if</span> <span class="n">keep</span><span class="p">:</span>
                <span class="n">keepers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">:</span>
            <span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">][</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">keepers</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="p">:</span>
            <span class="n">new_row_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">new_row_names</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">keepers</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="n">new_data</span><span class="p">,</span>
            <span class="n">row_names</span><span class="o">=</span><span class="n">new_row_names</span><span class="p">,</span>
            <span class="n">column_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


    <span class="c1"># TODO: needs checks when relative - {start, width and end} do not agree</span>
<div class="viewcode-block" id="GenomicRanges.narrow">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.narrow">[docs]</a>
    <span class="k">def</span> <span class="nf">narrow</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">start</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">width</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">end</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Narrow genomic positions by provided ``start``, ``width`` and ``end`` parameters.</span>

<span class="sd">        Important: these parameters are relative shift in positions for each range.</span>

<span class="sd">        Args:</span>
<span class="sd">            start (int, optional): Relative start position. Defaults to None.</span>
<span class="sd">            width (int, optional): Relative end position. Defaults to None.</span>
<span class="sd">            end (int, optional): Relative width of the interval. Defaults to None.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If `width` is provided, either `start` or `end` must be provided.</span>
<span class="sd">            ValueError: Provide two of the three parameters - `start`, `end` and `width`</span>
<span class="sd">                but not all.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges:  A new `GenomicRanges` object with narrow ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">width</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Only provide two of the three parameters - `start`, &quot;</span>
                <span class="s2">&quot;`end` and `width` but not all!&quot;</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">width</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">end</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;If width is provided, either start or end must be provided.&quot;</span>
                <span class="p">)</span>

        <span class="n">new_starts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_ends</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">all_starts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">)</span>
        <span class="n">all_ends</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
            <span class="n">new_starts</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">start</span> <span class="o">-</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_starts</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_starts</span> <span class="o">=</span> <span class="n">all_starts</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">end</span><span class="p">:</span>
            <span class="n">new_ends</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_starts</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_ends</span> <span class="o">=</span> <span class="n">all_ends</span>

        <span class="k">if</span> <span class="n">width</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">new_starts</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">-</span> <span class="n">width</span> <span class="o">+</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_ends</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">end</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">new_ends</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="n">width</span> <span class="o">-</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_starts</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;If `width` is provided, either `start` or `end` must be provided.&quot;</span>
                <span class="p">)</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_starts</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_ends</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="n">new_data</span><span class="p">,</span>
            <span class="n">number_of_rows</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">row_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="p">,</span>
            <span class="n">column_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">_calc_gap_widths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Internal method to calculate gap widths.</span>

<span class="sd">        Args:</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[int]: Gap widths for each range.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">obj</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">),</span>
            <span class="s2">&quot;starts&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">),</span>
            <span class="s2">&quot;ends&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">),</span>
            <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">),</span>
            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">)))],</span>
        <span class="p">}</span>

        <span class="n">df_gr</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="p">(</span><span class="n">obj</span><span class="p">)</span>
        <span class="n">df_gr</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">ignore_strand</span><span class="p">:</span>
            <span class="n">obj</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;*&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">))</span>

        <span class="n">gapwidths</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span>
            <span class="o">.</span><span class="n">rolling</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
            <span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">calc_row_gapwidth</span><span class="p">(</span>
                    <span class="n">df_gr</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="p">:],</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="p">:]</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">gapwidths</span>

    <span class="c1"># inter range methods</span>

    <span class="c1"># TODO: implement dropEmptyRanges</span>
    <span class="c1"># TODO: this is a very ineffecient implementation, can do a better job later.</span>
<div class="viewcode-block" id="GenomicRanges.reduce">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.reduce">[docs]</a>
    <span class="k">def</span> <span class="nf">reduce</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">with_reverse_map</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">min_gap_width</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Reduce orders the ranges, then merges overlapping or adjacent ranges.</span>

<span class="sd">        Args:</span>
<span class="sd">            with_reverse_map (bool, optional): Whether to return map of indices back to</span>
<span class="sd">                original object. Defaults to False.</span>
<span class="sd">            min_gap_width (int, optional): Ranges separated by a gap of</span>
<span class="sd">                at least ``min_gap_width`` positions are not merged. Defaults to 1.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` object with reduced intervals.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;gapwidths&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calc_gap_widths</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">)</span>
        <span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;gapwidth_flag&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">(</span><span class="kc">True</span> <span class="k">if</span> <span class="p">(</span><span class="n">isna</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">x</span> <span class="o">==</span> <span class="mi">0</span><span class="p">))</span> <span class="k">else</span> <span class="n">x</span> <span class="o">&lt;</span> <span class="n">min_gap_width</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;gapwidths&quot;</span><span class="p">]</span>
        <span class="p">]</span>

        <span class="n">gaps_to_merge</span> <span class="o">=</span> <span class="n">df_gr</span><span class="p">[</span><span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;gapwidth_flag&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span><span class="p">]</span>  <span class="c1"># noqa: E712</span>

        <span class="n">gaps_merged</span> <span class="o">=</span> <span class="n">gaps_to_merge</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;gapwidth_flag&quot;</span><span class="p">],</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">starts</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="nb">min</span><span class="p">),</span> <span class="n">ends</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">,</span> <span class="nb">max</span><span class="p">),</span> <span class="n">revmap</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">))</span>

        <span class="n">gaps_merged</span> <span class="o">=</span> <span class="n">gaps_merged</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

        <span class="n">gaps_not_merged</span> <span class="o">=</span> <span class="n">df_gr</span><span class="p">[</span><span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;gapwidth_flag&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>  <span class="c1"># noqa: E712</span>
        <span class="n">gaps_not_merged</span><span class="p">[</span><span class="s2">&quot;revmap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gaps_not_merged</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[</span><span class="n">x</span><span class="p">])</span>
        <span class="n">gaps_not_merged</span> <span class="o">=</span> <span class="n">gaps_not_merged</span><span class="p">[</span><span class="n">gaps_merged</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

        <span class="n">finale</span> <span class="o">=</span> <span class="n">concat</span><span class="p">([</span><span class="n">gaps_merged</span><span class="p">,</span> <span class="n">gaps_not_merged</span><span class="p">])</span>

        <span class="n">columns_to_keep</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">with_reverse_map</span><span class="p">:</span>
            <span class="n">columns_to_keep</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;revmap&quot;</span><span class="p">)</span>

        <span class="n">finale</span> <span class="o">=</span> <span class="n">finale</span><span class="p">[</span><span class="n">columns_to_keep</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">finale</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.range">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.range">[docs]</a>
    <span class="k">def</span> <span class="nf">range</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">with_reverse_map</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate ranges for each chromosome. (minimum of all starts, maximum of all ends) in the object.</span>

<span class="sd">        Technically its same as :py:meth:`~genomicranges.GenomicRanges.GenomicRanges.reduce`</span>
<span class="sd">        with a ridiculously high ``min_gap_width``.</span>

<span class="sd">        Args:</span>
<span class="sd">            with_reverse_map (bool, optional): return map of indices back to</span>
<span class="sd">                original object?. Defaults to False.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: a new `GenomicRanges` object with the new ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">reduce</span><span class="p">(</span>
            <span class="n">with_reverse_map</span><span class="o">=</span><span class="n">with_reverse_map</span><span class="p">,</span>
            <span class="n">min_gap_width</span><span class="o">=</span><span class="mi">100000000000</span><span class="p">,</span>
            <span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.compute_seqlengths">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.compute_seqlengths">[docs]</a>
    <span class="k">def</span> <span class="nf">compute_seqlengths</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get seqlengths either from the :py:class:`~genomicranges.SeqInfo.SeqInfo` object or computes one from the</span>
<span class="sd">        current ranges.</span>

<span class="sd">        Note: If computed, they are specific to this `GenomicRanges` object</span>
<span class="sd">        and may not represent the seqlenths of the genome.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict[str, int]: A dict with chromosome names as keys and their lengths as values.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">seqlengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqlengths</span>

        <span class="k">if</span> <span class="n">seqlengths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">seqlengths</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">ranges</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">range</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ranges</span><span class="p">:</span>
                <span class="n">seqlengths</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">seqlengths</span></div>


<div class="viewcode-block" id="GenomicRanges.gaps">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.gaps">[docs]</a>
    <span class="k">def</span> <span class="nf">gaps</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">end</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">MutableMapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Identify gaps in genomic positions for each distinct `seqname` (chromosome) and `strand` combination.</span>

<span class="sd">        Args:</span>
<span class="sd">            start (int, optional): Restrict chromosome start position. Defaults to 1.</span>
<span class="sd">            end (MutableMapping[str, int], optional): Restrict end</span>
<span class="sd">                position for each chromosome. Defaults to None. If None, it uses the</span>
<span class="sd">                :py:class:`~genomicranges.SeqInfo.SeqInfo` object if available.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` containing</span>
<span class="sd">            gaps across chromosome and strand. If there are no gaps, returns None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># seqlengths = self._computeSeqLengths()</span>
        <span class="n">seqlengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seqlengths</span>

        <span class="k">if</span> <span class="n">end</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">seqlengths</span>

        <span class="n">df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">groups</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">])</span>

        <span class="n">gap_intervals</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="n">group</span><span class="p">[</span><span class="s2">&quot;iindex&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">group</span><span class="p">))</span>

            <span class="n">all_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">end_limit</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">end</span> <span class="ow">and</span> <span class="n">end</span><span class="p">[</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">end_limit</span> <span class="o">=</span> <span class="n">end</span><span class="p">[</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

            <span class="n">tgaps</span> <span class="o">=</span> <span class="n">find_gaps</span><span class="p">(</span><span class="n">all_intvals</span><span class="p">,</span> <span class="n">start_limit</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end_limit</span><span class="o">=</span><span class="n">end_limit</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">td</span> <span class="ow">in</span> <span class="n">tgaps</span><span class="p">:</span>
                <span class="n">tmp_start</span> <span class="o">=</span> <span class="n">td</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">tmp_start</span> <span class="o">&lt;</span> <span class="n">start</span><span class="p">:</span>
                    <span class="n">tmp_start</span> <span class="o">=</span> <span class="n">start</span>

                <span class="n">tmp_end</span> <span class="o">=</span> <span class="n">td</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">end_limit</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">tmp_end</span> <span class="o">&gt;</span> <span class="n">end_limit</span><span class="p">:</span>
                    <span class="n">tmp_end</span> <span class="o">=</span> <span class="n">end_limit</span>

                <span class="n">td_res</span> <span class="o">=</span> <span class="p">(</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">tmp_start</span><span class="p">,</span> <span class="n">tmp_end</span><span class="p">)</span>

                <span class="n">gap_intervals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">td_res</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
                <span class="n">strands</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
                <span class="n">missing_strands</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">strands</span><span class="p">)))</span>
                <span class="k">for</span> <span class="n">ms</span> <span class="ow">in</span> <span class="n">missing_strands</span><span class="p">:</span>
                    <span class="n">gap_intervals</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">name</span><span class="p">,</span> <span class="n">ms</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">[</span><span class="n">name</span><span class="p">]))</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gap_intervals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">gap_intervals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.disjoin">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.disjoin">[docs]</a>
    <span class="k">def</span> <span class="nf">disjoin</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">with_reverse_map</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate disjoint genomic positions for each distinct `seqname` (chromosome) and `strand` combination.</span>

<span class="sd">        Args:</span>
<span class="sd">            with_reverse_map (bool, optional): Whether to return map of indices back to</span>
<span class="sd">                original object. Defaults to False.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` containing</span>
<span class="sd">            disjoint ranges across chromosome and strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">groups</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">])</span>

        <span class="n">disjoin_intervals</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="n">group</span><span class="p">[</span><span class="s2">&quot;iindex&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">group</span><span class="p">))</span>

            <span class="n">all_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">tdisjoin</span> <span class="o">=</span> <span class="n">find_disjoin</span><span class="p">(</span><span class="n">all_intvals</span><span class="p">,</span> <span class="n">with_reverse_map</span><span class="o">=</span><span class="n">with_reverse_map</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">td</span> <span class="ow">in</span> <span class="n">tdisjoin</span><span class="p">:</span>
                <span class="n">td_res</span> <span class="o">=</span> <span class="p">(</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">td</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">td</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">with_reverse_map</span><span class="p">:</span>
                    <span class="n">td_res</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">td</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                <span class="n">disjoin_intervals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">td_res</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">disjoin_intervals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">with_reverse_map</span><span class="p">:</span>
            <span class="n">columns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;revmap&quot;</span><span class="p">)</span>

        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">disjoin_intervals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.is_disjoint">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.is_disjoint">[docs]</a>
    <span class="k">def</span> <span class="nf">is_disjoint</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Whether all ranges (for each chromosome, strand) are disjoint.</span>

<span class="sd">        Args:</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool: True if the intervals are disjoint.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">new_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">disjoin</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">new_gr</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shape</span></div>


<div class="viewcode-block" id="GenomicRanges.disjoint_bins">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.disjoint_bins">[docs]</a>
    <span class="k">def</span> <span class="nf">disjoint_bins</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span></div>


    <span class="c1"># set operations</span>
<div class="viewcode-block" id="GenomicRanges.union">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.union">[docs]</a>
    <span class="k">def</span> <span class="nf">union</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find union of genomic intervals with `other`.</span>

<span class="sd">        Args:</span>
<span class="sd">            other (GenomicRanges): The other `GenomicRanges` object.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``other`` is not a `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` object with all ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`other` is not a `GenomicRanges` object&quot;</span><span class="p">)</span>

        <span class="n">a_df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">b_df_gr</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">df_gr</span> <span class="o">=</span> <span class="n">concat</span><span class="p">([</span><span class="n">a_df_gr</span><span class="p">,</span> <span class="n">b_df_gr</span><span class="p">])</span>
        <span class="n">df_gr</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="n">groups</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">])</span>

        <span class="n">union_intervals</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="n">group</span><span class="p">[</span><span class="s2">&quot;iindex&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">group</span><span class="p">))</span>

            <span class="n">all_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">tunion</span> <span class="o">=</span> <span class="n">find_union</span><span class="p">(</span><span class="n">all_intvals</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">td</span> <span class="ow">in</span> <span class="n">tunion</span><span class="p">:</span>
                <span class="n">td_res</span> <span class="o">=</span> <span class="p">(</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">td</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">td</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">union_intervals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">td_res</span><span class="p">)</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">union_intervals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.intersect">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.intersect">[docs]</a>
    <span class="k">def</span> <span class="nf">intersect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find intersection of genomic intervals with `other`.</span>

<span class="sd">        Args:</span>
<span class="sd">            other (GenomicRanges): The other `GenomicRanges` object.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``other`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` object with intersecting ranges.</span>
<span class="sd">            If there are no overlapping intervals, returns None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`other` is not a `GenomicRanges` object&quot;</span><span class="p">)</span>

        <span class="n">a_df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">b_df_gr</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">df_gr</span> <span class="o">=</span> <span class="n">concat</span><span class="p">([</span><span class="n">a_df_gr</span><span class="p">,</span> <span class="n">b_df_gr</span><span class="p">])</span>
        <span class="n">df_gr</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="n">groups</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">])</span>

        <span class="n">intersect_intervals</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="n">group</span><span class="p">[</span><span class="s2">&quot;iindex&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">group</span><span class="p">))</span>

            <span class="n">all_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">tintersect</span> <span class="o">=</span> <span class="n">find_intersect</span><span class="p">(</span><span class="n">all_intvals</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">td</span> <span class="ow">in</span> <span class="n">tintersect</span><span class="p">:</span>
                <span class="n">td_res</span> <span class="o">=</span> <span class="p">(</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">td</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">td</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">intersect_intervals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">td_res</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">intersect_intervals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">intersect_intervals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.setdiff">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.setdiff">[docs]</a>
    <span class="k">def</span> <span class="nf">setdiff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find set difference of genomic intervals with `other`.</span>

<span class="sd">        Args:</span>
<span class="sd">            other (GenomicRanges): The other `GenomicRanges` object.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``other`` is not of type `GenomicRanges`.</span>


<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` object with the diff ranges.</span>
<span class="sd">            If there are no diff intervals, returns None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`other` is not a `GenomicRanges` object&quot;</span><span class="p">)</span>

        <span class="n">a_df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">b_df_gr</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">only_seqnames</span> <span class="o">=</span> <span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span><span class="n">a_df_gr</span><span class="p">[[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">]],</span> <span class="n">b_df_gr</span><span class="p">[[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">]]]</span>
        <span class="p">)</span>
        <span class="n">only_seqnames</span><span class="p">[</span><span class="s2">&quot;seqstrand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">only_seqnames</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">only_seqnames</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>

        <span class="n">unique_seqs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">only_seqnames</span><span class="p">[</span><span class="s2">&quot;seqstrand&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="n">diff_ints</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">unique_seqs</span><span class="p">:</span>
            <span class="n">ustrand</span> <span class="o">=</span> <span class="n">name</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">name</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

            <span class="n">a_set</span> <span class="o">=</span> <span class="n">a_df_gr</span><span class="p">[</span>
                <span class="p">(</span><span class="n">a_df_gr</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">a_df_gr</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ustrand</span><span class="p">)</span>
            <span class="p">]</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">a_set</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="n">a_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">a_set</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">a_set</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">b_set</span> <span class="o">=</span> <span class="n">b_df_gr</span><span class="p">[</span>
                <span class="p">(</span><span class="n">b_df_gr</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">b_df_gr</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ustrand</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="n">b_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">b_set</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">b_set</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">tdiff</span> <span class="o">=</span> <span class="n">find_diff</span><span class="p">(</span><span class="n">a_intvals</span><span class="p">,</span> <span class="n">b_intvals</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">td</span> <span class="ow">in</span> <span class="n">tdiff</span><span class="p">:</span>
                <span class="n">td_res</span> <span class="o">=</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">ustrand</span><span class="p">,</span> <span class="n">td</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">td</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">diff_ints</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">td_res</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">diff_ints</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">diff_ints</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.binned_average">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.binned_average">[docs]</a>
    <span class="k">def</span> <span class="nf">binned_average</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">scorename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">bins</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span> <span class="n">outname</span><span class="p">:</span> <span class="nb">str</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate average for a column across all bins, then set a column called ``outname`` with those values.</span>

<span class="sd">        Args:</span>
<span class="sd">            scorename (str): Score column to compute averages on.</span>
<span class="sd">            bins (GenomicRanges): Bins you want to use.</span>
<span class="sd">            outname (str): New column name to add to the object.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If ``scorename`` column does not exist.</span>
<span class="sd">            Exception: ``scorename`` is not all ints or floats.</span>
<span class="sd">            TypeError: If ``bins`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` similar to bins,</span>
<span class="sd">            with a new column containing the averages.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">bins</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`bins` is not a `GenomicRanges` object.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">scorename</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">scorename</span><span class="si">}</span><span class="s2">&#39; is not a valid column name&quot;</span><span class="p">)</span>

        <span class="n">values</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">scorename</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">((</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">float</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">values</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">scorename</span><span class="si">}</span><span class="s2">&#39; values must be either ints or floats.&quot;</span><span class="p">)</span>

        <span class="n">df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;values&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">values</span>

        <span class="n">tgt_gr</span> <span class="o">=</span> <span class="n">bins</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">tgt_groups</span> <span class="o">=</span> <span class="n">tgt_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">cache_intvals</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">tgt_groups</span><span class="p">:</span>
            <span class="n">src_intervals</span> <span class="o">=</span> <span class="n">df_gr</span><span class="p">[</span><span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">name</span><span class="p">]</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">src_intervals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">name</span><span class="p">,</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">],</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">],</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cache_intvals</span><span class="p">:</span>
                <span class="n">all_intvals</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                        <span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span>
                        <span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span>
                    <span class="p">)</span>
                <span class="p">]</span>

                <span class="n">_</span><span class="p">,</span> <span class="n">np_sum</span> <span class="o">=</span> <span class="n">compute_mean</span><span class="p">(</span>
                    <span class="n">intervals</span><span class="o">=</span><span class="n">all_intvals</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;values&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
                <span class="p">)</span>

                <span class="n">cache_intvals</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">np_sum</span>

            <span class="n">np_sum</span> <span class="o">=</span> <span class="n">cache_intvals</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">tint</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                <span class="n">vec</span> <span class="o">=</span> <span class="n">np_sum</span><span class="p">[</span><span class="n">tint</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">tint</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]]</span>
                <span class="n">vec_mean</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">vec</span><span class="p">)</span> <span class="o">/</span> <span class="n">count_nonzero</span><span class="p">(</span><span class="n">vec</span><span class="p">)</span>

                <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">name</span><span class="p">,</span> <span class="n">tint</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">],</span> <span class="n">tint</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">],</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="n">vec_mean</span><span class="p">))</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="n">outname</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.subtract">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.subtract">[docs]</a>
    <span class="k">def</span> <span class="nf">subtract</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span> <span class="n">min_overlap</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">):</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span></div>


    <span class="c1"># integer range methods</span>
<div class="viewcode-block" id="GenomicRanges.coverage">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.coverage">[docs]</a>
    <span class="k">def</span> <span class="nf">coverage</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">shift</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">width</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">weight</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">ndarray</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate coverage for each chromosome, For each position, counts the number of ranges that cover it.</span>

<span class="sd">        Args:</span>
<span class="sd">            shift (int, optional): Shift all genomic positions. Defaults to 0.</span>
<span class="sd">            width (int, optional): Restrict the width of all</span>
<span class="sd">                chromosomes. Defaults to None.</span>
<span class="sd">            weight (int, optional): Weight to use. Defaults to 1.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict[str, ndarray]:  A dictionary with chromosome names as keys and the</span>
<span class="sd">            coverage vector as value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">groups</span> <span class="o">=</span> <span class="n">df_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">)</span>

        <span class="n">shift_arr</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">shift</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">shift_arr</span> <span class="o">=</span> <span class="n">zeros</span><span class="p">(</span><span class="n">shift</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="n">all_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">cov</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">create_np_interval_vector</span><span class="p">(</span>
                <span class="n">intervals</span><span class="o">=</span><span class="n">all_intvals</span><span class="p">,</span> <span class="n">with_reverse_map</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">shift</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">cov</span> <span class="o">=</span> <span class="n">concatenate</span><span class="p">((</span><span class="n">shift_arr</span><span class="p">,</span> <span class="n">cov</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">weight</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">cov</span> <span class="o">=</span> <span class="n">cov</span> <span class="o">*</span> <span class="n">weight</span>

            <span class="k">if</span> <span class="n">width</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">cov</span> <span class="o">=</span> <span class="n">cov</span><span class="p">[:</span><span class="n">width</span><span class="p">]</span>

            <span class="n">result</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">cov</span>

        <span class="k">return</span> <span class="n">result</span></div>


    <span class="c1"># search based methods</span>
<div class="viewcode-block" id="GenomicRanges.find_overlaps">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.find_overlaps">[docs]</a>
    <span class="k">def</span> <span class="nf">find_overlaps</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span>
        <span class="n">query_type</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;any&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;within&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;any&quot;</span><span class="p">,</span>
        <span class="n">max_gap</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">min_overlap</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find overlaps between ``subject`` (self) and a ``query`` `GenomicRanges` object.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Query `GenomicRanges`.</span>
<span class="sd">            query_type (Literal[&quot;any&quot;, &quot;start&quot;, &quot;end&quot;, &quot;within&quot;], optional): Overlap query type,</span>
<span class="sd">                must be one of</span>

<span class="sd">                - &quot;any&quot;: Any overlap is good</span>
<span class="sd">                - &quot;start&quot;: Overlap at the beginning of the intervals</span>
<span class="sd">                - &quot;end&quot;: Must overlap at the end of the intervals</span>
<span class="sd">                - &quot;within&quot;: Fully contain the query interval</span>

<span class="sd">                Defaults to &quot;any&quot;.</span>
<span class="sd">            max_gap (int, optional): Maximum gap allowed in the overlap.</span>
<span class="sd">                Defaults to -1 (no gap allowed).</span>
<span class="sd">            min_overlap (int, optional): Minimum overlap with query. Defaults to 1.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``query`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` object</span>
<span class="sd">            with the same length as query, containing hits to overlapping indices.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`query` is not a `GenomicRanges` object.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">query_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">OVERLAP_QUERY_TYPES</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;&#39;</span><span class="si">{</span><span class="n">query_type</span><span class="si">}</span><span class="s2">&#39; must be one of </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">OVERLAP_QUERY_TYPES</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>

        <span class="n">df_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">tgt_gr</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">tgt_groups</span> <span class="o">=</span> <span class="n">tgt_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">])</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">tgt_groups</span><span class="p">:</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">ustrand</span> <span class="o">=</span> <span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">src_intervals</span> <span class="o">=</span> <span class="n">df_gr</span><span class="p">[</span>
                <span class="p">(</span><span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df_gr</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ustrand</span><span class="p">)</span>
            <span class="p">]</span>

            <span class="n">src_intvals_map</span> <span class="o">=</span> <span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">src_intervals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">ustrand</span><span class="p">,</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">],</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">],</span> <span class="p">[]))</span>
                <span class="k">continue</span>

            <span class="n">subject_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                    <span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
                <span class="p">)</span>
            <span class="p">]</span>

            <span class="n">query_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">thits</span> <span class="o">=</span> <span class="n">find_overlaps</span><span class="p">(</span>
                <span class="n">subject_intvals</span><span class="p">,</span>
                <span class="n">query_intvals</span><span class="p">,</span>
                <span class="n">max_gap</span><span class="o">=</span><span class="n">max_gap</span><span class="p">,</span>
                <span class="n">min_overlap</span><span class="o">=</span><span class="n">min_overlap</span><span class="p">,</span>
                <span class="n">query_type</span><span class="o">=</span><span class="n">query_type</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">th</span> <span class="ow">in</span> <span class="n">thits</span><span class="p">:</span>
                <span class="n">tindices</span> <span class="o">=</span> <span class="p">[</span><span class="n">src_intvals_map</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">th</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span>
                <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">ustrand</span><span class="p">,</span> <span class="n">th</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">th</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">tindices</span><span class="p">))</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">,</span> <span class="s2">&quot;hits&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.count_overlaps">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.count_overlaps">[docs]</a>
    <span class="k">def</span> <span class="nf">count_overlaps</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span>
        <span class="n">query_type</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;any&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;within&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;any&quot;</span><span class="p">,</span>
        <span class="n">max_gap</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">min_overlap</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Count overlaps between ``subject`` (self) and a ``query`` `GenomicRanges` object.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Query `GenomicRanges`.</span>
<span class="sd">            query_type (Literal[&quot;any&quot;, &quot;start&quot;, &quot;end&quot;, &quot;within&quot;], optional): Overlap query type,</span>
<span class="sd">                must be one of</span>

<span class="sd">                - &quot;any&quot;: Any overlap is good</span>
<span class="sd">                - &quot;start&quot;: Overlap at the beginning of the intervals</span>
<span class="sd">                - &quot;end&quot;: Must overlap at the end of the intervals</span>
<span class="sd">                - &quot;within&quot;: Fully contain the query interval</span>

<span class="sd">                Defaults to &quot;any&quot;.</span>
<span class="sd">            max_gap (int, optional): Maximum gap allowed in the overlap.</span>
<span class="sd">                Defaults to -1 (no gap allowed).</span>
<span class="sd">            min_overlap (int, optional): Minimum overlap with query. Defaults to 1.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``query`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[int]: Number of overlaps for each range in query.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`query` is not a `GenomicRanges` object.&quot;</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">find_overlaps</span><span class="p">(</span>
            <span class="n">query</span><span class="o">=</span><span class="n">query</span><span class="p">,</span>
            <span class="n">query_type</span><span class="o">=</span><span class="n">query_type</span><span class="p">,</span>
            <span class="n">max_gap</span><span class="o">=</span><span class="n">max_gap</span><span class="p">,</span>
            <span class="n">min_overlap</span><span class="o">=</span><span class="n">min_overlap</span><span class="p">,</span>
            <span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">result</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">hits</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;hits&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span> <span class="k">for</span> <span class="n">ht</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">]</span></div>


<div class="viewcode-block" id="GenomicRanges.subset_by_overlaps">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.subset_by_overlaps">[docs]</a>
    <span class="k">def</span> <span class="nf">subset_by_overlaps</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span>
        <span class="n">query_type</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span><span class="s2">&quot;any&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;within&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;any&quot;</span><span class="p">,</span>
        <span class="n">max_gap</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">min_overlap</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Subset ``subject`` (self) with overlaps in ``query`` `GenomicRanges` object.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Query `GenomicRanges`.</span>
<span class="sd">            query_type (Literal[&quot;any&quot;, &quot;start&quot;, &quot;end&quot;, &quot;within&quot;], optional): Overlap query type,</span>
<span class="sd">                must be one of</span>

<span class="sd">                - &quot;any&quot;: Any overlap is good</span>
<span class="sd">                - &quot;start&quot;: Overlap at the beginning of the intervals</span>
<span class="sd">                - &quot;end&quot;: Must overlap at the end of the intervals</span>
<span class="sd">                - &quot;within&quot;: Fully contain the query interval</span>

<span class="sd">                Defaults to &quot;any&quot;.</span>
<span class="sd">            max_gap (int, optional): Maximum gap allowed in the overlap.</span>
<span class="sd">                Defaults to -1 (no gap allowed).</span>
<span class="sd">            min_overlap (int, optional): Minimum overlap with query. Defaults to 1.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``query`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` object</span>
<span class="sd">            containing only subsets.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`query` is not a `GenomicRanges` object.&quot;</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">find_overlaps</span><span class="p">(</span>
            <span class="n">query</span><span class="o">=</span><span class="n">query</span><span class="p">,</span>
            <span class="n">query_type</span><span class="o">=</span><span class="n">query_type</span><span class="p">,</span>
            <span class="n">max_gap</span><span class="o">=</span><span class="n">max_gap</span><span class="p">,</span>
            <span class="n">min_overlap</span><span class="o">=</span><span class="n">min_overlap</span><span class="p">,</span>
            <span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">result</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">hits</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;hits&quot;</span><span class="p">)</span>
        <span class="n">hit_counts</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span> <span class="k">for</span> <span class="n">ht</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">]</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="p">[</span><span class="n">idx</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">hit_counts</span><span class="p">))</span> <span class="k">if</span> <span class="n">hit_counts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">query</span><span class="p">[</span><span class="n">indices</span><span class="p">,</span> <span class="p">:]</span></div>


    <span class="k">def</span> <span class="nf">_generic_search</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">stepstart</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
        <span class="n">stepend</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Internal function to search ``self`` and a ``query`` `GenomicRanges` object.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``query`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` object that has</span>
<span class="sd">            the same length as query but contains `hits` to</span>
<span class="sd">            `indices` and `distance`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`query` is not a `GenomicRanges` object.&quot;</span><span class="p">)</span>

        <span class="n">subject_gr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">query_gr</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">query_groups</span> <span class="o">=</span> <span class="n">query_gr</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">])</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">query_groups</span><span class="p">:</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">ustrand</span> <span class="o">=</span> <span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">src_intervals</span> <span class="o">=</span> <span class="n">subject_gr</span><span class="p">[</span>
                <span class="p">(</span><span class="n">subject_gr</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">subject_gr</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ustrand</span><span class="p">)</span>
            <span class="p">]</span>

            <span class="n">src_intvals_map</span> <span class="o">=</span> <span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">src_intervals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">group</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">ustrand</span><span class="p">,</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">],</span> <span class="n">g</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">],</span> <span class="p">[],</span> <span class="kc">None</span><span class="p">))</span>
                <span class="k">continue</span>

            <span class="n">subject_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                    <span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">src_intervals</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
                <span class="p">)</span>
            <span class="p">]</span>

            <span class="n">query_intvals</span> <span class="o">=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
            <span class="p">]</span>

            <span class="n">thits</span> <span class="o">=</span> <span class="n">find_nearest</span><span class="p">(</span>
                <span class="n">subject_intvals</span><span class="p">,</span> <span class="n">query_intvals</span><span class="p">,</span> <span class="n">stepstart</span><span class="o">=</span><span class="n">stepstart</span><span class="p">,</span> <span class="n">stepend</span><span class="o">=</span><span class="n">stepend</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">th</span> <span class="ow">in</span> <span class="n">thits</span><span class="p">:</span>
                <span class="n">tindices</span> <span class="o">=</span> <span class="p">[</span><span class="n">src_intvals_map</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">th</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span>
                <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">ustrand</span><span class="p">,</span> <span class="n">th</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">th</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">tindices</span><span class="p">,</span> <span class="n">th</span><span class="p">[</span><span class="mi">3</span><span class="p">]))</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">,</span> <span class="s2">&quot;hits&quot;</span><span class="p">,</span> <span class="s2">&quot;distance&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span>

<div class="viewcode-block" id="GenomicRanges.nearest">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.nearest">[docs]</a>
    <span class="k">def</span> <span class="nf">nearest</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Search nearest positions both upstream and downstream that overlap with each range in ``query``.</span>

<span class="sd">        Adds a new column (&quot;hits&quot;) to ``query`` with the nearest matches.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Query `GenomicRanges` to find nearest positions.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strand. Defaults to False.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``query`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` containing list of</span>
<span class="sd">            possible &quot;hits&quot; indices for each range in ``query``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_search</span><span class="p">(</span><span class="n">query</span><span class="o">=</span><span class="n">query</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.precede">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.precede">[docs]</a>
    <span class="k">def</span> <span class="nf">precede</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Search nearest positions only downstream that overlap with each range in ``query``.</span>

<span class="sd">        Adds a new column (&quot;hits&quot;) to ``query`` with the nearest matches.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Query `GenomicRanges` to find nearest positions.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strand. Defaults to False.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``query`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` containing list of</span>
<span class="sd">            possible &quot;hits&quot; indices for each range in ``query``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_search</span><span class="p">(</span><span class="n">query</span><span class="o">=</span><span class="n">query</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">stepend</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.follow">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.follow">[docs]</a>
    <span class="k">def</span> <span class="nf">follow</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Search nearest positions only upstream that overlap with each range in ``query``.</span>

<span class="sd">        Adds a new column (&quot;hits&quot;) to ``query`` with the nearest matches.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Query `GenomicRanges` to find nearest positions.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strand. Defaults to False.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``query`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): a new `GenomicRanges` containing list of</span>
<span class="sd">            possible &quot;hits&quot; indices for each range in ``query``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_search</span><span class="p">(</span>
            <span class="n">query</span><span class="o">=</span><span class="n">query</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">stepstart</span><span class="o">=</span><span class="mi">0</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.distance_to_nearest">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.distance_to_nearest">[docs]</a>
    <span class="k">def</span> <span class="nf">distance_to_nearest</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">,</span>
        <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="s2">&quot;GenomicRanges&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Search nearest positions only downstream that overlap with each range in ``query``.</span>

<span class="sd">        Adds a new column (&quot;hits&quot;) to ``query`` with the nearest matches.</span>

<span class="sd">        Note: Technically same as :py:meth:`~genomicranges.GenomicRanges.GenomicRanges.nearest`</span>
<span class="sd">        since we also return `distances`.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Query `GenomicRanges` to find nearest positions.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strand. Defaults to False.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If query is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (&quot;GenomicRanges&quot;, optional): A new `GenomicRanges` containing list of</span>
<span class="sd">            possible &quot;hits&quot; indices for each range in ``query``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_search</span><span class="p">(</span><span class="n">query</span><span class="o">=</span><span class="n">query</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">)</span></div>


    <span class="c1"># compare and order methods</span>
<div class="viewcode-block" id="GenomicRanges.duplicated">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.duplicated">[docs]</a>
    <span class="k">def</span> <span class="nf">duplicated</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">bool</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Element wise comparison to find duplicate ranges.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[bool]: True if duplicated else False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span></div>


<div class="viewcode-block" id="GenomicRanges.match">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.match">[docs]</a>
    <span class="k">def</span> <span class="nf">match</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">query</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Element wise comparison to find exact match ranges.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Input `GenomicRanges` to search matches.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If ``query`` is not of type `GenomicRanges`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (List[int], optional): List contianing index positions if able to match else None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;`query` is not a `GenomicRanges` object.&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">hits</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">query</span><span class="p">:</span>
            <span class="n">sliced</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span>
                <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">])</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">])</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">])</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">])</span>
            <span class="p">]</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sliced</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">sliced</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
        <span class="k">return</span> <span class="n">hits</span></div>


    <span class="k">def</span> <span class="nf">_generic_pandas_ranges</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Internal function to create a :py:class:`~pandas.DataFrame` from ranges.</span>

<span class="sd">        Args:</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>
<span class="sd">            sort (bool, optional): Whether to sort by region. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: a pandas `DataFrame` of the genomic ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">obj</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">),</span>
            <span class="s2">&quot;starts&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;starts&quot;</span><span class="p">),</span>
            <span class="s2">&quot;ends&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;ends&quot;</span><span class="p">),</span>
            <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">),</span>
            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">))),</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="n">ignore_strand</span><span class="p">:</span>
            <span class="n">obj</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;*&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;seqnames&quot;</span><span class="p">))</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="p">(</span><span class="n">obj</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">sort</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">df</span>

    <span class="k">def</span> <span class="nf">_generic_order</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Internal function to compute order.</span>

<span class="sd">        Args:</span>
<span class="sd">            query (GenomicRanges): Query `GenomicRanges` to search matches.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[int, optional]: Sorted index order.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">sorted</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_pandas_ranges</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">new_order</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">new_order</span>

<div class="viewcode-block" id="GenomicRanges.is_unsorted">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.is_unsorted">[docs]</a>
    <span class="k">def</span> <span class="nf">is_unsorted</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Whether the genomic positions are unsorted.</span>

<span class="sd">        Args:</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool: True if unsorted else False.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">order</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_order</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">)</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">order</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span>

        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">diff</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">True</span>

        <span class="k">return</span> <span class="kc">False</span></div>


<div class="viewcode-block" id="GenomicRanges.order">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.order">[docs]</a>
    <span class="k">def</span> <span class="nf">order</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">decreasing</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the order of indices for sorting.</span>

<span class="sd">        Args:</span>
<span class="sd">            decreasing (bool, optional): Whether to sort in descending order. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[int]: List with order of indices.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">order</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_order</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">decreasing</span><span class="p">:</span>
            <span class="n">order</span> <span class="o">=</span> <span class="n">order</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">order</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span></div>


<div class="viewcode-block" id="GenomicRanges.sort">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.sort">[docs]</a>
    <span class="k">def</span> <span class="nf">sort</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">decreasing</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">ignore_strand</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Sort the `GenomicRanges` object.</span>

<span class="sd">        Args:</span>
<span class="sd">            decreasing (bool, optional): Whether to sort in descending order. Defaults to False.</span>
<span class="sd">            ignore_strand (bool, optional): Whether to ignore strands. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            &quot;GenomicRanges&quot;: A new sorted `GenomicRanges` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">order</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_order</span><span class="p">(</span><span class="n">ignore_strand</span><span class="o">=</span><span class="n">ignore_strand</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">decreasing</span><span class="p">:</span>
            <span class="n">order</span> <span class="o">=</span> <span class="n">order</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">new_order</span> <span class="o">=</span> <span class="n">order</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="n">new_order</span><span class="p">,</span> <span class="p">:]</span></div>


<div class="viewcode-block" id="GenomicRanges.rank">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.rank">[docs]</a>
    <span class="k">def</span> <span class="nf">rank</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get rank of the `GenomicRanges` object.</span>

<span class="sd">        For each range identifies its position is a sorted order.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[int]: List of indices identifying rank.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">order</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generic_order</span><span class="p">()</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
        <span class="n">rank</span> <span class="o">=</span> <span class="p">[</span><span class="n">order</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">order</span><span class="p">))]</span>
        <span class="k">return</span> <span class="n">rank</span></div>


    <span class="c1"># windowing functions</span>
<div class="viewcode-block" id="GenomicRanges.tile_by_range">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.tile_by_range">[docs]</a>
    <span class="k">def</span> <span class="nf">tile_by_range</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">width</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Split each range into chunks by ``n`` (number of intervals) or ``width`` (intervals with equal width).</span>

<span class="sd">        Note: Either ``n`` or ``width`` must be provided, but not both.</span>

<span class="sd">        Also, checkout :py:func:`~genomicranges.io.tiling.tile_genome` for splitting</span>
<span class="sd">        a gneomic into chunks.</span>

<span class="sd">        Args:</span>
<span class="sd">            n (int, optional): Number of intervals to split into.</span>
<span class="sd">                Defaults to None.</span>
<span class="sd">            width (int, optional): Width of each interval. Defaults to None.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If both ``n`` or ``width`` are provided.</span>

<span class="sd">        Returns:</span>
<span class="sd">            &quot;GenomicRanges&quot;: A new `GenomicRanges` with the split ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">width</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Either `n` or `width` must be provided but not both.&quot;</span><span class="p">)</span>

        <span class="n">ranges</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">range</span><span class="p">()</span>

        <span class="n">all_intervals</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">ranges</span><span class="p">:</span>
            <span class="n">twidth</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">twidth</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">val</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">val</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="n">n</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">width</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">twidth</span> <span class="o">=</span> <span class="n">width</span>

            <span class="n">all_intervals</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                <span class="n">split_intervals</span><span class="p">(</span>
                    <span class="n">val</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">],</span> <span class="n">twidth</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">all_intervals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.tile">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.tile">[docs]</a>
    <span class="k">def</span> <span class="nf">tile</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">width</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Split each interval by ``n`` (number of sub intervals) or ``width`` (intervals with equal width).</span>

<span class="sd">        Note: Either ``n`` or ``width`` must be provided but not both.</span>

<span class="sd">        Also, checkout :py:func:`~genomicranges.io.tiling.tile_genome` for splitting</span>
<span class="sd">        a gneomic into chunks, or</span>
<span class="sd">        :py:meth:`~genomicranges.GenomicRanges.GenomicRanges.tile_by_range`.</span>

<span class="sd">        Args:</span>
<span class="sd">            n (int, optional): Number of intervals to split into. Defaults to None.</span>
<span class="sd">            width (int, optional): Width of each interval. Defaults to None.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If both ``n`` and ``width`` are provided.</span>

<span class="sd">        Returns:</span>
<span class="sd">            &quot;GenomicRanges&quot;: A new `GenomicRanges` with the split ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">width</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;either `n` or `width` must be provided but not both&quot;</span><span class="p">)</span>

        <span class="n">all_intervals</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="n">twidth</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">twidth</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">((</span><span class="n">val</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">val</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">n</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">width</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">twidth</span> <span class="o">=</span> <span class="n">width</span>

            <span class="n">all_intervals</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                <span class="n">split_intervals</span><span class="p">(</span>
                    <span class="n">val</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">],</span> <span class="n">twidth</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">all_intervals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.sliding_windows">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.sliding_windows">[docs]</a>
    <span class="k">def</span> <span class="nf">sliding_windows</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">width</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">step</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Slide along each range by ``width`` (intervals with equal ``width``) and ``step``.</span>

<span class="sd">        Also, checkout :py:func:`~genomicranges.io.tiling.tile_genome` for splitting</span>
<span class="sd">        a gneomic into chunks, or</span>
<span class="sd">        :py:meth:`~genomicranges.GenomicRanges.GenomicRanges.tile_by_range`.</span>

<span class="sd">        Args:</span>
<span class="sd">            width (int, optional): Width of each interval. Defaults to None.</span>
<span class="sd">            step (int, optional): Step interval, Defaults to 1.</span>

<span class="sd">        Returns:</span>
<span class="sd">            &quot;GenomicRanges&quot;: a new `GenomicRanges` with the sliding ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_intervals</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
            <span class="n">all_intervals</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                <span class="n">slide_intervals</span><span class="p">(</span>
                    <span class="n">val</span><span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">],</span>
                    <span class="n">val</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">],</span>
                    <span class="n">val</span><span class="p">[</span><span class="s2">&quot;starts&quot;</span><span class="p">],</span>
                    <span class="n">val</span><span class="p">[</span><span class="s2">&quot;ends&quot;</span><span class="p">],</span>
                    <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span>
                    <span class="n">step</span><span class="o">=</span><span class="n">step</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">]</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">all_intervals</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">final_df</span> <span class="o">=</span> <span class="n">final_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;seqnames&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;starts&quot;</span><span class="p">,</span> <span class="s2">&quot;ends&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">from_pandas</span><span class="p">(</span><span class="n">final_df</span><span class="p">)</span></div>


    <span class="c1"># misc methods</span>
<div class="viewcode-block" id="GenomicRanges.sample">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.sample">[docs]</a>
    <span class="k">def</span> <span class="nf">sample</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Randomly sample ``k`` intervals.</span>

<span class="sd">        Args:</span>
<span class="sd">            k (int, optional): Number of intervals to sample. Defaults to 5.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` with randomly sampled ranges.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)),</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="n">sample</span><span class="p">,</span> <span class="p">:]</span></div>


<div class="viewcode-block" id="GenomicRanges.invert_strand">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.invert_strand">[docs]</a>
    <span class="k">def</span> <span class="nf">invert_strand</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Invert strand information for each interval.</span>

<span class="sd">        Conversion map:</span>
<span class="sd">            - &quot;+&quot; map to &quot;-&quot;</span>
<span class="sd">            - &quot;-&quot; becomes &quot;+&quot;</span>
<span class="sd">            - &quot;*&quot; stays the same</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new `GenomicRanges` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">convertor</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;+&quot;</span><span class="p">:</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">:</span> <span class="s2">&quot;*&quot;</span><span class="p">}</span>
        <span class="n">inverts</span> <span class="o">=</span> <span class="p">[</span><span class="n">convertor</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">)]</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">new_data</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">inverts</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span>
            <span class="n">new_data</span><span class="p">,</span>
            <span class="n">number_of_rows</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">row_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">row_names</span><span class="p">,</span>
            <span class="n">column_names</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">,</span>
            <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="GenomicRanges.concat">
<a class="viewcode-back" href="../../api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.concat">[docs]</a>
    <span class="k">def</span> <span class="nf">concat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">granges</span><span class="p">:</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;GenomicRanges&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Row-wise concatenate multiple `GenomicRanges` objects.</span>

<span class="sd">        Args:</span>
<span class="sd">            granges (GenomicRanges): Objects to concatenate.</span>

<span class="sd">        Raises:</span>
<span class="sd">            TypeError: If any ``granges`` are not &quot;GenomicRanges&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GenomicRanges: A new concatenated `GenomicRanges` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_granges</span> <span class="o">=</span> <span class="p">[</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">gr</span><span class="p">,</span> <span class="n">GenomicRanges</span><span class="p">)</span> <span class="k">for</span> <span class="n">gr</span> <span class="ow">in</span> <span class="n">granges</span><span class="p">]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">all_granges</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;all provided objects are not GenomicRanges objects&quot;</span><span class="p">)</span>

        <span class="n">all_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">gr</span><span class="o">.</span><span class="n">column_names</span> <span class="k">for</span> <span class="n">gr</span> <span class="ow">in</span> <span class="n">granges</span><span class="p">]</span>
        <span class="n">all_columns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">)</span>
        <span class="n">all_unique_columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">([</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">all_columns</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">])</span>
        <span class="p">)</span>

        <span class="n">new_data</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">all_unique_columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">new_data</span><span class="p">:</span>
                <span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">for</span> <span class="n">gr</span> <span class="ow">in</span> <span class="n">granges</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">gr</span><span class="o">.</span><span class="n">column_names</span><span class="p">:</span>
                    <span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">col</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">gr</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span><span class="p">:</span>
                <span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">col</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">new_data</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">GenomicRanges</span><span class="p">(</span><span class="n">new_data</span><span class="p">,</span> <span class="n">column_names</span><span class="o">=</span><span class="n">all_unique_columns</span><span class="p">)</span></div>
</div>

</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, jkanche
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../_static/documentation_options.js?v=71d9d8e6"></script>
    <script src="../../_static/doctools.js?v=888ff710"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/furo.js?v=32e29ea5"></script>
    </body>
</html>