<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html"><link rel="search" title="Search" href="../search.html"><link rel="next" title="genomicranges.io package" href="genomicranges.io.html"><link rel="prev" title="genomicranges" href="modules.html">

    <!-- Generated with Sphinx 9.1.0 and Furo 2025.12.19 -->
        <title>genomicranges package - GenomicRanges 0.8.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=7bdb33bb" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">GenomicRanges 0.8.4 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <span class="sidebar-brand-text">GenomicRanges 0.8.4 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html"><code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code>: Genomic analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#construct-a-genomicranges-object">Construct a <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#sequence-information">Sequence information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#subset-operations">Subset operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#intra-range-transformations">Intra-range transformations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#inter-range-methods">Inter-range methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#set-operations">Set operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#compute-over-bins">Compute over bins</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#generate-tiles-or-bins">Generate tiles or bins</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#coverage">Coverage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#overlap-based-methods">Overlap based methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#search-operations">Search operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#comparison-rank-and-order-operations">Comparison, rank and order operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#combine-genomicranges-objects-by-rows">Combine <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> objects by rows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#misc-operations">Misc operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html#compressedgenomicrangeslist-class"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList</span></code> class</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul class="current">
<li class="toctree-l2 current has-children current-page"><a class="current reference internal" href="#">genomicranges package</a><input aria-label="Toggle navigation of genomicranges package" checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="genomicranges.io.html">genomicranges.io package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/api/genomicranges.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="genomicranges-package">
<h1>genomicranges package<a class="headerlink" href="#genomicranges-package" title="Link to this heading">¶</a></h1>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Link to this heading">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="genomicranges.io.html">genomicranges.io package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="genomicranges.io.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="genomicranges.io.html#module-genomicranges.io.gtf">genomicranges.io.gtf module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="genomicranges.io.html#genomicranges.io.gtf.parse_gtf"><code class="docutils literal notranslate"><span class="pre">parse_gtf()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="genomicranges.io.html#genomicranges.io.gtf.read_gtf"><code class="docutils literal notranslate"><span class="pre">read_gtf()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="genomicranges.io.html#module-genomicranges.io.ucsc">genomicranges.io.ucsc module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="genomicranges.io.html#genomicranges.io.ucsc.access_gtf_ucsc"><code class="docutils literal notranslate"><span class="pre">access_gtf_ucsc()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="genomicranges.io.html#genomicranges.io.ucsc.read_ucsc"><code class="docutils literal notranslate"><span class="pre">read_ucsc()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="genomicranges.io.html#module-genomicranges.io">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading">¶</a></h2>
</section>
<section id="module-genomicranges.GenomicRanges">
<span id="genomicranges-genomicranges-module"></span><h2>genomicranges.GenomicRanges module<a class="headerlink" href="#module-genomicranges.GenomicRanges" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GRanges">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-prename descclassname"><span class="pre">genomicranges.GenomicRanges.</span></span><span class="sig-name descname"><span class="pre">GRanges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqinfo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_validate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GRanges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GRanges" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-prename descclassname"><span class="pre">genomicranges.GenomicRanges.</span></span><span class="sig-name descname"><span class="pre">GenomicRanges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqinfo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_validate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.biocobject.BiocObject" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocObject</span></code></a></p>
<p><code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> provides a container class to represent and operate over genomic regions and annotations.</p>
<p>Note: The documentation for some of the methods are derived from the
<a class="reference external" href="https://github.com/Bioconductor/GenomicRanges">GenomicRanges R/Bioconductor package</a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><span class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{}</span></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__annotations__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__copy__">
<span class="sig-name descname"><span class="pre">__copy__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.__copy__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__copy__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A shallow copy of the current <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__deepcopy__">
<span class="sig-name descname"><span class="pre">__deepcopy__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_nil</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.__deepcopy__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__deepcopy__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A deep copy of the current <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__getitem__" title="Link to this definition">¶</a></dt>
<dd><p>Alias to <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_subset" title="genomicranges.GenomicRanges.GenomicRanges.get_subset"><code class="xref py py-attr docutils literal notranslate"><span class="pre">get_subset</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mcols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqinfo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_validate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__init__" title="Link to this definition">¶</a></dt>
<dd><p>Initialize a <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqnames</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – List of sequence or chromosome names.</p></li>
<li><p><strong>ranges</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/IRanges/api/iranges.html#iranges.IRanges.IRanges" title="(in IRanges v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">IRanges</span></code></a></span>) – Genomic positions and widths of each position. Must have the same length as <code class="docutils literal notranslate"><span class="pre">seqnames</span></code>.</p></li>
<li><p><strong>strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>Strand information for each genomic range. This should be 0 (any strand),
1 (forward strand) or -1 (reverse strand). If None, all genomic ranges
are assumed to be 0 (any strand).</p>
<p>May be provided as a list of strings representing the strand;
“+” for forward strand, “-” for reverse strand, or “*” for any strand and will be mapped
to 1, -1 or 0 respectively.</p>
</p></li>
<li><p><strong>names</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Names</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Names for each genomic range.
Defaults to None, which means the ranges are unnamed.</p></li>
<li><p><strong>mcols</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>]</span>) – A ~py:class:<cite>~biocframe.BiocFrame.BiocFrame</cite> with the number of rows same as
number of genomic ranges, containing per-range annotation.
Defaults to None, in which case an empty BiocFrame object is created.</p></li>
<li><p><strong>seqinfo</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a>]</span>) – Sequence information. Defaults to None, in which case a
<code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code> object is created with the unique set of
chromosome names from <code class="docutils literal notranslate"><span class="pre">seqnames</span></code>.</p></li>
<li><p><strong>metadata</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>], <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Additional metadata.
Defaults to None, and is assigned to an empty dictionary.</p></li>
<li><p><strong>validate</strong> – Internal use only.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__iter__" title="Link to this definition">¶</a></dt>
<dd><p>Iterator over rows.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRangesIter" title="genomicranges.GenomicRanges.GenomicRangesIter"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRangesIter</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.__len__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__len__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of rows.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__repr__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A string representation of this <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.__setitem__">
<span class="sig-name descname"><span class="pre">__setitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.__setitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.__setitem__" title="Link to this definition">¶</a></dt>
<dd><p>Alias to <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.set_subset" title="genomicranges.GenomicRanges.GenomicRanges.set_subset"><code class="xref py py-attr docutils literal notranslate"><span class="pre">set_subset</span></code></a>.</p>
<p>This operation modifies object in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.binned_average">
<span class="sig-name descname"><span class="pre">binned_average</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scorename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'binned_average'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.binned_average"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.binned_average" title="Link to this definition">¶</a></dt>
<dd><p>Calculate average for a column across all regions in <code class="docutils literal notranslate"><span class="pre">bins</span></code>, then set a column specified by ‘outname’ with
those values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scorename</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Score column to compute averages on.</p></li>
<li><p><strong>bins</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Bins you want to use.</p></li>
<li><p><strong>outname</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – New column name to add to the object.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify <code class="docutils literal notranslate"><span class="pre">bins</span></code> in place.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3.12/library/exceptions.html#ValueError" title="(in Python v3.12)"><strong>ValueError</strong></a> – If <code class="docutils literal notranslate"><span class="pre">scorename</span></code> column does not exist.
    <code class="docutils literal notranslate"><span class="pre">scorename</span></code> is not all ints or floats.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3.12/library/exceptions.html#TypeError" title="(in Python v3.12)"><strong>TypeError</strong></a> – If <code class="docutils literal notranslate"><span class="pre">bins</span></code> is not of type <cite>GenomicRanges</cite>.</p></li>
</ul>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A modified <code class="docutils literal notranslate"><span class="pre">bins</span></code> object with the computed averages,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.copy" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__copy__" title="genomicranges.GenomicRanges.GenomicRanges.__copy__"><code class="xref py py-meth docutils literal notranslate"><span class="pre">__copy__()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.count_overlaps">
<span class="sig-name descname"><span class="pre">count_overlaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'any'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.count_overlaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.count_overlaps" title="Link to this definition">¶</a></dt>
<dd><p>Count overlaps between subject (self) and a <code class="docutils literal notranslate"><span class="pre">query</span></code> <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Query <cite>GenomicRanges</cite>.</p></li>
<li><p><strong>query_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'any'</span></code>, <code class="docutils literal notranslate"><span class="pre">'start'</span></code>, <code class="docutils literal notranslate"><span class="pre">'end'</span></code>, <code class="docutils literal notranslate"><span class="pre">'within'</span></code>]</span>) – <p>Overlap query type, must be one of</p>
<ul>
<li><p>”any”: Any overlap is good</p></li>
<li><p>”start”: Overlap at the beginning of the intervals</p></li>
<li><p>”end”: Must overlap at the end of the intervals</p></li>
<li><p>”within”: Fully contain the query interval</p></li>
</ul>
<p>Defaults to “any”.</p>
</p></li>
<li><p><strong>max_gap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum gap allowed in the overlap.
Defaults to -1 (no gap allowed).</p></li>
<li><p><strong>min_overlap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum overlap with query. Defaults to 0.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.
Defaults to 1.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.12/library/exceptions.html#TypeError" title="(in Python v3.12)"><strong>TypeError</strong></a> – If <code class="docutils literal notranslate"><span class="pre">query</span></code> is not of type <cite>GenomicRanges</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>NumPy vector with length matching query,
value represents the number of overlaps in <cite>self</cite> for each query.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.coverage">
<span class="sig-name descname"><span class="pre">coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.coverage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.coverage" title="Link to this definition">¶</a></dt>
<dd><p>Calculate coverage for each chromosome. For each position, this method
counts the number of ranges that cover it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>shift</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Shift all genomic positions. Defaults to 0.</p></li>
<li><p><strong>width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Restrict the width of all chromosomes. Defaults to None.</p></li>
<li><p><strong>weight</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Weight to use for each range. Defaults to 1.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strand, effectively combining
all strands for coverage calculation. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary with chromosome names as keys and the
coverage vector as values. The dictionary is sorted
by chromosome name.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.disjoin">
<span class="sig-name descname"><span class="pre">disjoin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">with_reverse_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.disjoin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.disjoin" title="Link to this definition">¶</a></dt>
<dd><p>Calculate disjoint genomic positions for each distinct (seqname, strand) pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>with_reverse_map</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to return a map of indices back to the original object.
Defaults to False.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> containing disjoint ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.disjoint_bins">
<span class="sig-name descname"><span class="pre">disjoint_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.disjoint_bins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.disjoint_bins" title="Link to this definition">¶</a></dt>
<dd><p>Split ranges into a set of bins so that the ranges in each bin are disjoint.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An ndarray indicating the bin index for each range.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.distance">
<span class="sig-name descname"><span class="pre">distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.distance" title="Link to this definition">¶</a></dt>
<dd><p>Compute the pair-wise distance with intervals in query.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>query</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a>, <a class="reference external" href="https://biocpy.github.io/IRanges/api/iranges.html#iranges.IRanges.IRanges" title="(in IRanges v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">IRanges</span></code></a>]</span>) – Query <cite>GenomicRanges</cite> or <cite>IRanges</cite>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Numpy vector containing distances for each interval in query.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.empty">
<span class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.empty"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.empty" title="Link to this definition">¶</a></dt>
<dd><p>Create an zero-length <cite>GenomicRanges</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>same type as caller, in this case a <cite>GenomicRanges</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.end">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">end</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><span class="pre">ndarray</span></a></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.end" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_end" title="genomicranges.GenomicRanges.GenomicRanges.get_end"><code class="xref py py-attr docutils literal notranslate"><span class="pre">get_end</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.extract_groups_by_seqnames">
<span class="sig-name descname"><span class="pre">extract_groups_by_seqnames</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.extract_groups_by_seqnames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.extract_groups_by_seqnames" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.find_overlaps">
<span class="sig-name descname"><span class="pre">find_overlaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'any'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.find_overlaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.find_overlaps" title="Link to this definition">¶</a></dt>
<dd><p>Find overlaps between subject (self) and a <code class="docutils literal notranslate"><span class="pre">query</span></code> <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Query <cite>GenomicRanges</cite>.</p></li>
<li><p><strong>query_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'any'</span></code>, <code class="docutils literal notranslate"><span class="pre">'start'</span></code>, <code class="docutils literal notranslate"><span class="pre">'end'</span></code>, <code class="docutils literal notranslate"><span class="pre">'within'</span></code>]</span>) – <p>Overlap query type, must be one of</p>
<ul>
<li><p>”any”: Any overlap is good</p></li>
<li><p>”start”: Overlap at the beginning of the intervals</p></li>
<li><p>”end”: Must overlap at the end of the intervals</p></li>
<li><p>”within”: Fully contain the query interval</p></li>
</ul>
<p>Defaults to “any”.</p>
</p></li>
<li><p><strong>select</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'all'</span></code>, <code class="docutils literal notranslate"><span class="pre">'first'</span></code>, <code class="docutils literal notranslate"><span class="pre">'last'</span></code>, <code class="docutils literal notranslate"><span class="pre">'arbitrary'</span></code>]</span>) – Determine what hit to choose when
there are multiple hits for an interval in <code class="docutils literal notranslate"><span class="pre">subject</span></code>.</p></li>
<li><p><strong>max_gap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum gap allowed in the overlap.
Defaults to -1 (no gap allowed).</p></li>
<li><p><strong>min_overlap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum overlap with query. Defaults to 0.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.
Defaults to 1.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.12/library/exceptions.html#TypeError" title="(in Python v3.12)"><strong>TypeError</strong></a> – If <code class="docutils literal notranslate"><span class="pre">query</span></code> is not of type <cite>GenomicRanges</cite>.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p>query_hits: Indices into query ranges</p></li>
<li><p>self_hits: Corresponding indices into self ranges that are upstream</p></li>
</ul>
<p>Each row represents a query-self pair where self overlaps query.</p>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>A BiocFrame with two columns</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.flank">
<span class="sig-name descname"><span class="pre">flank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">both</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.flank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.flank" title="Link to this definition">¶</a></dt>
<dd><p>Compute flanking ranges for each range. The logic for this comes from the <cite>R/GenomicRanges</cite> &amp; <cite>IRanges</cite>
packages.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">start</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code> for a given range, the flanking occurs at the <cite>start</cite>,
otherwise the <cite>end</cite>.
The <cite>widths</cite> of the flanks are given by the <code class="docutils literal notranslate"><span class="pre">width</span></code> parameter.</p>
<p><code class="docutils literal notranslate"><span class="pre">width</span></code> can be negative, in which case the flanking region is
reversed so that it represents a prefix or suffix of the range.</p>
<p>Usage:</p>
<blockquote>
<div><p><cite>gr.flank(3, True)</cite>, where “x” indicates a range in <code class="docutils literal notranslate"><span class="pre">gr</span></code> and “-” indicates the
resulting flanking region:</p>
<blockquote>
<div><p>—xxxxxxx</p>
</div></blockquote>
<dl class="simple">
<dt>If <code class="docutils literal notranslate"><span class="pre">start</span></code> were <code class="docutils literal notranslate"><span class="pre">False</span></code>, the range in <code class="docutils literal notranslate"><span class="pre">gr</span></code> becomes</dt><dd><p>xxxxxxx—</p>
</dd>
</dl>
<p>For negative width, i.e. <cite>gr.flank(x, -3, FALSE)</cite>, where “*” indicates the overlap
between “x” and the result:</p>
<blockquote>
<div><p>xxxx***</p>
</div></blockquote>
<p>If <code class="docutils literal notranslate"><span class="pre">both</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, then, for all ranges in “x”, the flanking regions are
extended into (or out of, if <code class="docutils literal notranslate"><span class="pre">width</span></code> is negative) the range, so that the result
straddles the given endpoint and has twice the width given by width.</p>
<dl class="simple">
<dt>This is illustrated below for <cite>gr.flank(3, both=TRUE)</cite>:</dt><dd><p>—<a href="#id1"><span class="problematic" id="id2">**</span></a><a href="#id3"><span class="problematic" id="id4">*</span></a>xxxx</p>
</dd>
</dl>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Width to flank by. May be negative.</p></li>
<li><p><strong>start</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>]]</span>) – <p>Whether to only flank starts.
Defaults to True.</p>
<p>Alternatively, you may provide a list of start values,
whose length is the same as the number of ranges.</p>
</p></li>
<li><p><strong>both</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to flank both starts and ends. Defaults to False.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the flanked regions,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.follow">
<span class="sig-name descname"><span class="pre">follow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'last'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.follow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.follow" title="Link to this definition">¶</a></dt>
<dd><p>Search nearest positions only upstream that overlap with each range in <code class="docutils literal notranslate"><span class="pre">query</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Query <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> to find nearest positions.</p></li>
<li><p><strong>select</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'all'</span></code>, <code class="docutils literal notranslate"><span class="pre">'last'</span></code>]</span>) – Whether to return “all” hits or just “last”.
Defaults to “last”.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strand. Defaults to False.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.
Defaults to 1.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A numpy array of integers with length matching query, containing indices</dt><dd><p>into self for the closest downstream position of each query range. Value may be
None if there are no matches.</p>
</dd>
<dt>If select=”all”:</dt><dd><p>A BiocFrame with two columns:
- query_hits: Indices into query ranges
- self_hits: Corresponding indices into self ranges that are upstream
Each row represents a query-self pair where self follows query.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>If select=”last”</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.from_pandas">
<span class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">from_pandas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.from_pandas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.from_pandas" title="Link to this definition">¶</a></dt>
<dd><p>Create an <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object from a <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>input</strong> – Input data. Must contain columns ‘seqnames’, ‘starts’ and ‘widths’ or “ends”.
‘ends’ are expected to be inclusive.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.from_polars">
<span class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">from_polars</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.from_polars"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.from_polars" title="Link to this definition">¶</a></dt>
<dd><p>Create an <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object from a <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>input</strong> – Input polars DataFrame. Must contain columns ‘seqnames’, ‘starts’ and ‘widths’ or “ends”.
‘ends’ are expected to be inclusive.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.gaps">
<span class="sig-name descname"><span class="pre">gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.gaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.gaps" title="Link to this definition">¶</a></dt>
<dd><p>Identify complemented ranges for each distinct (seqname, strand) pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Restrict chromosome start position. Defaults to 1.</p></li>
<li><p><strong>end</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>Restrict end position for each chromosome.
Defaults to None. If None, extracts sequence information from
<a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.seqinfo" title="genomicranges.GenomicRanges.GenomicRanges.seqinfo"><code class="xref py py-attr docutils literal notranslate"><span class="pre">seqinfo</span></code></a> object if available.</p>
<p>Alternatively, you may provide a dictionary with seqnames as
keys and the values specifying the ends.</p>
</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> with complement ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_end">
<span class="sig-name descname"><span class="pre">get_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_end" title="Link to this definition">¶</a></dt>
<dd><p>Get all end positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>NumPy array of 32-bit signed integers containing the end
positions for all ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_mcols">
<span class="sig-name descname"><span class="pre">get_mcols</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_mcols"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_mcols" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ~py:class:<cite>~biocframe.BiocFrame.BiocFrame</cite> containing per-range annotations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_names">
<span class="sig-name descname"><span class="pre">get_names</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_names" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Names</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of names for each genomic range.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_out_of_bound_index">
<span class="sig-name descname"><span class="pre">get_out_of_bound_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_out_of_bound_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_out_of_bound_index" title="Link to this definition">¶</a></dt>
<dd><p>Find indices of genomic ranges that are out of bounds.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A numpy array of integer indices where ranges are out of bounds.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_ranges">
<span class="sig-name descname"><span class="pre">get_ranges</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_ranges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_ranges" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/IRanges/api/iranges.html#iranges.IRanges.IRanges" title="(in IRanges v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">IRanges</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An <code class="docutils literal notranslate"><span class="pre">IRanges</span></code> object containing the positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_seqinfo">
<span class="sig-name descname"><span class="pre">get_seqinfo</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_seqinfo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_seqinfo" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ~py:class:<cite>~genomicranges.SeqInfo.SeqInfo</cite> containing sequence information.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_seqlengths">
<span class="sig-name descname"><span class="pre">get_seqlengths</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_seqlengths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_seqlengths" title="Link to this definition">¶</a></dt>
<dd><p>Get sequence lengths for each genomic range.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An ndarray containing the sequence lengths.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_seqnames">
<span class="sig-name descname"><span class="pre">get_seqnames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">as_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'list'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_seqnames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_seqnames" title="Link to this definition">¶</a></dt>
<dd><p>Access sequence names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>as_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'factor'</span></code>, <code class="docutils literal notranslate"><span class="pre">'list'</span></code>]</span>) – <p>Access seqnames as factor tuple, in which case, levels and codes
are returned.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">list</span></code>, then codes are mapped to levels and returned.</p>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Factor.Factor" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Factor</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">biocutils.Factor</span></code> if <cite>as_type=”factor”</cite>.
Otherwise a list of sequence names.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_start">
<span class="sig-name descname"><span class="pre">get_start</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_start"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_start" title="Link to this definition">¶</a></dt>
<dd><p>Get all start positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>NumPy array of 32-bit signed integers containing the start
positions for all ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_strand">
<span class="sig-name descname"><span class="pre">get_strand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">as_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'numpy'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_strand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_strand" title="Link to this definition">¶</a></dt>
<dd><p>Access strand information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>as_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'numpy'</span></code>, <code class="docutils literal notranslate"><span class="pre">'factor'</span></code>, <code class="docutils literal notranslate"><span class="pre">'list'</span></code>]</span>) – <dl class="simple">
<dt>Access seqnames as factor codes, in which case, a numpy</dt><dd><p>vector is retuned.</p>
</dd>
</dl>
<p>If <code class="docutils literal notranslate"><span class="pre">factor</span></code>, a tuple with codes as the strand vector
and levels a dictionary containing the mapping.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">list</span></code>, then codes are mapped to levels and returned.</p>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Tuple" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>A numpy vector representing strand, 0
for any strand, -1 for reverse strand
and 1 for forward strand.</p>
<p>A tuple of codes and levels.</p>
<p>A list of “+”, “-”, or “*” for each range.</p>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_subset">
<span class="sig-name descname"><span class="pre">get_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_subset" title="Link to this definition">¶</a></dt>
<dd><p>Subset <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code>, based on their indices or names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>subset</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – <p>Indices to be extracted. This may be an integer, boolean, string,
or any sequence thereof, as supported by
<code class="xref py py-meth docutils literal notranslate"><span class="pre">normalize_subscript()</span></code>.
Scalars are treated as length-1 sequences.</p>
<p>Strings may only be used if :py:attr:<code class="docutils literal notranslate"><span class="pre">~names</span></code> are available (see
<a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_names" title="genomicranges.GenomicRanges.GenomicRanges.get_names"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_names()</span></code></a>). The first occurrence of each string
in the names is used for extraction.</p>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the ranges of interest.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.get_width">
<span class="sig-name descname"><span class="pre">get_width</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.get_width"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.get_width" title="Link to this definition">¶</a></dt>
<dd><p>Get width of each genomic range.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>NumPy array of 32-bit signed integers containing the width
for all ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.intersect">
<span class="sig-name descname"><span class="pre">intersect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.intersect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.intersect" title="Link to this definition">¶</a></dt>
<dd><p>Find intersecting genomic intervals with <cite>other</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – The other <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with intersecting ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.intersect_ncls">
<span class="sig-name descname"><span class="pre">intersect_ncls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delete_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.intersect_ncls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.intersect_ncls" title="Link to this definition">¶</a></dt>
<dd><p>Find intersecting genomic intervals with <cite>other</cite> (uses NCLS index).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – The other <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p></li>
<li><p><strong>delete_index</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Defaults to True, to delete the cached ncls index.
Set to False, to reuse the index across multiple queries.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.
Defaults to 1.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with intersecting ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.invert_strand">
<span class="sig-name descname"><span class="pre">invert_strand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.invert_strand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.invert_strand" title="Link to this definition">¶</a></dt>
<dd><p>Invert strand for each range.</p>
<dl class="simple">
<dt>Conversion map:</dt><dd><ul class="simple">
<li><p>“+” map to “-”</p></li>
<li><p>“-” becomes “+”</p></li>
<li><p>“*” stays the same</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the object in place. Defaults to False.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the trimmed regions,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.is_disjoint">
<span class="sig-name descname"><span class="pre">is_disjoint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.is_disjoint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.is_disjoint" title="Link to this definition">¶</a></dt>
<dd><p>Calculate disjoint genomic positions for each distinct (seqname, strand) pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>True if all ranges are disjoint, otherwise False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.match">
<span class="sig-name descname"><span class="pre">match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.match"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.match" title="Link to this definition">¶</a></dt>
<dd><p>Element wise comparison to find exact match ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Query <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> to search for matches.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strand. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A NumPy array with length matching query
containing the matched indices.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.mcols">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">mcols</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><span class="pre">BiocFrame</span></a></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.mcols" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_mcols" title="genomicranges.GenomicRanges.GenomicRanges.get_mcols"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_mcols()</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.names">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">names</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.3.3)"><span class="pre">Names</span></a></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.names" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_names" title="genomicranges.GenomicRanges.GenomicRanges.get_names"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_names()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.narrow">
<span class="sig-name descname"><span class="pre">narrow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.narrow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.narrow" title="Link to this definition">¶</a></dt>
<dd><p>Narrow genomic positions by provided <code class="docutils literal notranslate"><span class="pre">start</span></code>, <code class="docutils literal notranslate"><span class="pre">width</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> parameters.</p>
<p>Important: these parameters are relative shift in positions for each range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Relative start position. Defaults to None.</p></li>
<li><p><strong>width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Relative end position. Defaults to None.</p></li>
<li><p><strong>end</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Relative width of the interval. Defaults to None.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the trimmed regions,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.nearest">
<span class="sig-name descname"><span class="pre">nearest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'arbitrary'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjacent_equals_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.nearest"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.nearest" title="Link to this definition">¶</a></dt>
<dd><p>Search nearest positions both upstream and downstream that overlap with each range in <code class="docutils literal notranslate"><span class="pre">query</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Query <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> to find nearest positions.</p></li>
<li><p><strong>select</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'all'</span></code>, <code class="docutils literal notranslate"><span class="pre">'arbitrary'</span></code>]</span>) – Determine what hit to choose when there are
multiple hits for an interval in <code class="docutils literal notranslate"><span class="pre">query</span></code>.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strand. Defaults to False.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.
Defaults to 1.</p></li>
<li><p><strong>adjacent_equals_overlap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – <p>Whether to consider immediately-adjacent subject intervals to be
equally “nearest” to the query as an overlapping subject interval.</p>
<p>If true, both overlapping and immediately-adjacent subject intervals
(i.e., a gap of zero) will be reported in matches.</p>
<p>Otherwise, immediately-adjacent subjects will only be reported if
overlapping subjects are not present.</p>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A numpy array of integers with length matching query, containing indices</dt><dd><p>into self for the closest for each query range. Value may be None if there
are no matches.</p>
</dd>
<dt>If select=”all”:</dt><dd><p>A BiocFrame with two columns:
- query_hits: Indices into query ranges
- self_hits: Corresponding indices into self ranges that are upstream
Each row represents a query-self pair where subject is nearest to query.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>If select=”arbitrary”</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.order">
<span class="sig-name descname"><span class="pre">order</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">decreasing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.order"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.order" title="Link to this definition">¶</a></dt>
<dd><p>Get the order of indices for sorting.</p>
<p>Order orders the genomic ranges by chromosome and strand.
Strand is ordered by reverse first (-1), any strand (0) and
forward strand (-1). Then by the start positions and width if
two regions have the same start.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>decreasing</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to sort in descending order. Defaults to False.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>NumPy vector containing index positions in the sorted order.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.pintersect">
<span class="sig-name descname"><span class="pre">pintersect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.pintersect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.pintersect" title="Link to this definition">¶</a></dt>
<dd><p>Parallel intersection of genomic ranges.</p>
<p>Computes the intersection for each parallel pair of ranges in <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>.
If seqnames mismatch or strands are incompatible (and not ignored), the result
for that index is an empty range (width 0).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – The other <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object. Must have the same length as <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.precede">
<span class="sig-name descname"><span class="pre">precede</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'first'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.precede"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.precede" title="Link to this definition">¶</a></dt>
<dd><p>Search nearest positions only downstream that overlap with each range in <code class="docutils literal notranslate"><span class="pre">query</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Query <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> to find nearest positions.</p></li>
<li><p><strong>select</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'all'</span></code>, <code class="docutils literal notranslate"><span class="pre">'first'</span></code>]</span>) – Whether to return “all” hits or just “first”.
Defaults to “first”.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strand. Defaults to False.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.
Defaults to 1.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A numpy array of integers with length matching query, containing indices</dt><dd><p>into self for the closest upstream position of each query range. Value may be
None if there are no matches.</p>
</dd>
<dt>If select=”all”:</dt><dd><p>A BiocFrame with two columns:
- query_hits: Indices into query ranges
- self_hits: Corresponding indices into self ranges that are upstream
Each row represents a query-self pair where self precedes query.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>If select=”first”</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.promoters">
<span class="sig-name descname"><span class="pre">promoters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">upstream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downstream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.promoters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.promoters" title="Link to this definition">¶</a></dt>
<dd><p>Extend ranges to promoter regions.</p>
<p>Generates promoter ranges relative to the transcription start site (TSS).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>upstream</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of positions to extend in the 5’ direction.
Defaults to 2000.</p></li>
<li><p><strong>downstream</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of positions to extend in the 3’ direction.
Defaults to 200.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the promoter regions,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.range">
<span class="sig-name descname"><span class="pre">range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">with_reverse_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.range"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.range" title="Link to this definition">¶</a></dt>
<dd><p>Calculate range bounds for each distinct (seqname, strand) pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>with_reverse_map</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to return map of indices back to
original object. Defaults to False.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the range bounds.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.ranges">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">ranges</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/IRanges/api/iranges.html#iranges.IRanges.IRanges" title="(in IRanges v0.7.2)"><span class="pre">IRanges</span></a></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.ranges" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_ranges" title="genomicranges.GenomicRanges.GenomicRanges.get_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_ranges()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.rank">
<span class="sig-name descname"><span class="pre">rank</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.rank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.rank" title="Link to this definition">¶</a></dt>
<dd><p>Get rank of the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p>
<p>For each range identifies its position is a sorted order.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Numpy vector containing rank.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.reduce">
<span class="sig-name descname"><span class="pre">reduce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">with_reverse_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_empty_ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_gap_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.reduce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.reduce" title="Link to this definition">¶</a></dt>
<dd><p>Reduce orders the ranges, then merges overlapping or adjacent ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>with_reverse_map</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to return map of indices back to
original object. Defaults to False.</p></li>
<li><p><strong>drop_empty_ranges</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to drop empty ranges. Defaults to False.</p></li>
<li><p><strong>min_gap_width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Ranges separated by a gap of
at least <code class="docutils literal notranslate"><span class="pre">min_gap_width</span></code> positions are not merged. Defaults to 1.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with reduced intervals.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.resize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.resize" title="Link to this definition">¶</a></dt>
<dd><p>Resize ranges to the specified <code class="docutils literal notranslate"><span class="pre">width</span></code> where either the <code class="docutils literal notranslate"><span class="pre">start</span></code>, <code class="docutils literal notranslate"><span class="pre">end</span></code>, or <code class="docutils literal notranslate"><span class="pre">center</span></code> is used as an
anchor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>]</span>) – Width to resize, cannot be negative!</p></li>
<li><p><strong>fix</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'start'</span></code>, <code class="docutils literal notranslate"><span class="pre">'end'</span></code>, <code class="docutils literal notranslate"><span class="pre">'center'</span></code>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'start'</span></code>, <code class="docutils literal notranslate"><span class="pre">'end'</span></code>, <code class="docutils literal notranslate"><span class="pre">'center'</span></code>]]]</span>) – <p>Fix positions by “start”, “end”, or “center”.
Defaults to “start”.</p>
<p>Alternatively, may provide a list of fix positions
for each genomic range.</p>
</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.12/library/exceptions.html#ValueError" title="(in Python v3.12)"><strong>ValueError</strong></a> – If parameter <code class="docutils literal notranslate"><span class="pre">fix</span></code> is neither <cite>start</cite>, <cite>end</cite>, nor <cite>center</cite>.
    If <code class="docutils literal notranslate"><span class="pre">width</span></code> is negative.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the resized regions,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.restrict">
<span class="sig-name descname"><span class="pre">restrict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_all_ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.restrict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.restrict" title="Link to this definition">¶</a></dt>
<dd><p>Restrict ranges to a given start and end positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>Start position. Defaults to None.</p>
<p>Alternatively may provide a dictionary
mapping sequence names to starts, or array of starts.</p>
</p></li>
<li><p><strong>end</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>End position. Defaults to None.</p>
<p>Alternatively may provide a dictionary
mapping sequence names to starts, or array of starts.</p>
</p></li>
<li><p><strong>keep_all_ranges</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to keep intervals that do not overlap with start and end.
Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the restricted regions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.sample">
<span class="sig-name descname"><span class="pre">sample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.sample"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.sample" title="Link to this definition">¶</a></dt>
<dd><p>Randomly sample <code class="docutils literal notranslate"><span class="pre">k</span></code> intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>k</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of intervals to sample. Defaults to 5.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> with randomly sampled ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.seqinfo">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">seqinfo</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><span class="pre">ndarray</span></a></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.seqinfo" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_seqinfo" title="genomicranges.GenomicRanges.GenomicRanges.get_seqinfo"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_seqinfo()</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.seqnames">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">seqnames</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><span class="pre">ndarray</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.seqnames" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_seqnames" title="genomicranges.GenomicRanges.GenomicRanges.get_seqnames"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_seqnames()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.set_mcols">
<span class="sig-name descname"><span class="pre">set_mcols</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mcols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.set_mcols"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.set_mcols" title="Link to this definition">¶</a></dt>
<dd><p>Set new range metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mcols</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>]</span>) – <p>A ~py:class:<cite>~biocframe.BiocFrame.BiocFrame</cite> with length same as the number
of ranges, containing per-range annotations.</p>
<p>May be None to remove range metadata.</p>
</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.set_names">
<span class="sig-name descname"><span class="pre">set_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.set_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.set_names" title="Link to this definition">¶</a></dt>
<dd><p>Set new names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>names</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Names</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Names for each genomic range.
May be <cite>None</cite> to remove names.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.set_ranges">
<span class="sig-name descname"><span class="pre">set_ranges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ranges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.set_ranges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.set_ranges" title="Link to this definition">¶</a></dt>
<dd><p>Set new ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ranges</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/IRanges/api/iranges.html#iranges.IRanges.IRanges" title="(in IRanges v0.7.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">IRanges</span></code></a></span>) – IRanges containing the genomic positions and widths.
Must have the same length as <code class="docutils literal notranslate"><span class="pre">seqnames</span></code>.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.set_seqinfo">
<span class="sig-name descname"><span class="pre">set_seqinfo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqinfo</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.set_seqinfo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.set_seqinfo" title="Link to this definition">¶</a></dt>
<dd><p>Set new sequence information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqinfo</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a>]</span>) – <p>A ~py:class:<cite>~genomicranges.SeqInfo.SeqInfo</cite> object containing information
about sequences in <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.seqnames" title="genomicranges.GenomicRanges.GenomicRanges.seqnames"><code class="xref py py-attr docutils literal notranslate"><span class="pre">seqnames</span></code></a>.</p>
<p>May be None to remove sequence information. This would then generate a
new sequence information based on the current sequence names.</p>
</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.set_seqnames">
<span class="sig-name descname"><span class="pre">set_seqnames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.set_seqnames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.set_seqnames" title="Link to this definition">¶</a></dt>
<dd><p>Set new sequence names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqnames</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>]</span>) – List of sequence or chromosome names.
Optionally can be a numpy array with indices mapped
to :py:attr:<code class="docutils literal notranslate"><span class="pre">~seqinfo</span></code>.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.set_strand">
<span class="sig-name descname"><span class="pre">set_strand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.set_strand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.set_strand" title="Link to this definition">¶</a></dt>
<dd><p>Set new strand information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>Strand information for each genomic range. This should be 0 (any strand),
1 (forward strand) or -1 (reverse strand).</p>
<p>Alternatively, may provide a list of strings representing the strand;
“+” for forward strand, “-” for reverse strand, or “*” for any strand
and will be automatically mapped to 1, -1 or 0 respectively.</p>
<p>May be set to None; in which case all genomic ranges are assumed to be 0.</p>
</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.set_subset">
<span class="sig-name descname"><span class="pre">set_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.set_subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.set_subset" title="Link to this definition">¶</a></dt>
<dd><p>Udate positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#slice" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">slice</span></code></a>, <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.range" title="genomicranges.GenomicRanges.GenomicRanges.range"><code class="xref py py-class docutils literal notranslate"><span class="pre">range</span></code></a>]</span>) – Integer indices, a boolean filter, or (if the current object is
named) names specifying the ranges to be replaced, see
<code class="xref py py-meth docutils literal notranslate"><span class="pre">normalize_subscript()</span></code>.</p></li>
<li><p><strong>value</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – An <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object of length equal to the number of ranges
to be replaced, as specified by <code class="docutils literal notranslate"><span class="pre">subset</span></code>.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.setdiff">
<span class="sig-name descname"><span class="pre">setdiff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.setdiff"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.setdiff" title="Link to this definition">¶</a></dt>
<dd><p>Find set difference of genomic intervals with <cite>other</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – The other <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the diff ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.shift">
<span class="sig-name descname"><span class="pre">shift</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.shift"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.shift" title="Link to this definition">¶</a></dt>
<dd><p>Shift all intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>shift</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>]</span>) – Shift interval. If shift is 0, the current
object is returned. Defaults to 0.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the shifted regions,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.sliding_windows">
<span class="sig-name descname"><span class="pre">sliding_windows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.sliding_windows"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.sliding_windows" title="Link to this definition">¶</a></dt>
<dd><p>Slide along each range by <code class="docutils literal notranslate"><span class="pre">width</span></code> (intervals with equal <code class="docutils literal notranslate"><span class="pre">width</span></code>) and <code class="docutils literal notranslate"><span class="pre">step</span></code>.</p>
<p>Also, checkout <code class="xref py py-func docutils literal notranslate"><span class="pre">tile_genome()</span></code> for splitting
a gneomic into chunks, or
<code class="xref py py-meth docutils literal notranslate"><span class="pre">tile_by_range()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> – Number of intervals to split into.
Defaults to None.</p></li>
<li><p><strong>width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Width of each interval. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a>]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> with the sliding ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.sort">
<span class="sig-name descname"><span class="pre">sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">decreasing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.sort"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.sort" title="Link to this definition">¶</a></dt>
<dd><p>Get the order of indices for sorting.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>decreasing</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to sort in descending order. Defaults to False.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the object in place. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the trimmed regions,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.split">
<span class="sig-name descname"><span class="pre">split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">groups</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.split"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.split" title="Link to this definition">¶</a></dt>
<dd><p>Split the <cite>GenomicRanges</cite> object into a <a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList" title="genomicranges.grangeslist.CompressedGenomicRangesList"><code class="xref py py-class docutils literal notranslate"><span class="pre">CompressedGenomicRangesList</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>groups</strong> (<span class="sphinx_autodoc_typehints-type">list</span>) – <p>A list specifying the groups or factors to split by.</p>
<p>Must have the same length as the number of genomic elements
in the object.</p>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type">CompressedGenomicRangesList</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <cite>CompressedGenomicRangesList</cite> containing the groups and their
corresponding elements.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.start">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">start</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><span class="pre">ndarray</span></a></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.start" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_start" title="genomicranges.GenomicRanges.GenomicRanges.get_start"><code class="xref py py-attr docutils literal notranslate"><span class="pre">get_start</span></code></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.strand">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">strand</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><span class="pre">ndarray</span></a></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.strand" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_strand" title="genomicranges.GenomicRanges.GenomicRanges.get_strand"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_strand()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.subset_by_overlaps">
<span class="sig-name descname"><span class="pre">subset_by_overlaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'any'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.subset_by_overlaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.subset_by_overlaps" title="Link to this definition">¶</a></dt>
<dd><p>Subset <code class="docutils literal notranslate"><span class="pre">subject</span></code> (self) with overlaps in <code class="docutils literal notranslate"><span class="pre">query</span></code> <cite>GenomicRanges</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Query <cite>GenomicRanges</cite>.</p></li>
<li><p><strong>query_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'any'</span></code>, <code class="docutils literal notranslate"><span class="pre">'start'</span></code>, <code class="docutils literal notranslate"><span class="pre">'end'</span></code>, <code class="docutils literal notranslate"><span class="pre">'within'</span></code>]</span>) – <p>Overlap query type, must be one of</p>
<ul>
<li><p>”any”: Any overlap is good</p></li>
<li><p>”start”: Overlap at the beginning of the intervals</p></li>
<li><p>”end”: Must overlap at the end of the intervals</p></li>
<li><p>”within”: Fully contain the query interval</p></li>
</ul>
<p>Defaults to “any”.</p>
</p></li>
<li><p><strong>max_gap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Maximum gap allowed in the overlap.
Defaults to -1 (no gap allowed).</p></li>
<li><p><strong>min_overlap</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Minimum overlap with query. Defaults to 0.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.
Defaults to 1.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.12/library/exceptions.html#TypeError" title="(in Python v3.12)"><strong>TypeError</strong></a> – If <code class="docutils literal notranslate"><span class="pre">query</span></code> is not of type <cite>GenomicRanges</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object containing overlapping ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.subtract">
<span class="sig-name descname"><span class="pre">subtract</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.subtract"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.subtract" title="Link to this definition">¶</a></dt>
<dd><p>Subtract searches for features in <code class="docutils literal notranslate"><span class="pre">x</span></code> that overlap <code class="docutils literal notranslate"><span class="pre">self</span></code> by at least the number of base pairs given by
<code class="docutils literal notranslate"><span class="pre">min_overlap</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<span class="sphinx_autodoc_typehints-type">GenomicRanges</span>) – Object to subtract.</p></li>
<li><p><strong>min_overlap</strong> (<span class="sphinx_autodoc_typehints-type">int</span>) – Minimum overlap with query.
Defaults to 1.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type">bool</span>) – Whether to ignore strands.
Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type">CompressedGenomicRangesList</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <cite>CompressedGenomicRangesList</cite> with the same size as <code class="docutils literal notranslate"><span class="pre">self</span></code> containing
the subtracted regions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.terminators">
<span class="sig-name descname"><span class="pre">terminators</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">upstream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downstream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.terminators"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.terminators" title="Link to this definition">¶</a></dt>
<dd><p>Extend ranges to termiantor regions.</p>
<p>Generates terminator ranges relative to the transcription end site (TES).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>upstream</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of positions to extend in the 5’ direction.
Defaults to 2000.</p></li>
<li><p><strong>downstream</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of positions to extend in the 3’ direction.
Defaults to 200.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the promoter regions,
either as a copy of the original or as a reference to the
(in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.tile">
<span class="sig-name descname"><span class="pre">tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.tile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.tile" title="Link to this definition">¶</a></dt>
<dd><p>Split each interval by <code class="docutils literal notranslate"><span class="pre">n</span></code> (number of sub intervals) or <code class="docutils literal notranslate"><span class="pre">width</span></code> (intervals with equal width).</p>
<p>Note: Either <code class="docutils literal notranslate"><span class="pre">n</span></code> or <code class="docutils literal notranslate"><span class="pre">width</span></code> must be provided but not both.</p>
<p>Also, checkout <code class="xref py py-func docutils literal notranslate"><span class="pre">tile_genome()</span></code> for splitting
a genome into chunks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of intervals to split into.
Defaults to None.</p></li>
<li><p><strong>width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Width of each interval. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.12/library/exceptions.html#ValueError" title="(in Python v3.12)"><strong>ValueError</strong></a> – If both <code class="docutils literal notranslate"><span class="pre">n</span></code> and <code class="docutils literal notranslate"><span class="pre">width</span></code> are provided.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> with the split ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.tile_genome">
<span class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">tile_genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqlengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ntile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tilewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_last_tile_in_chrom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.tile_genome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.tile_genome" title="Link to this definition">¶</a></dt>
<dd><p>Tile genome into approximately equal-sized regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqlengths</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Dictionary of sequence lengths by chromosome.</p></li>
<li><p><strong>ntile</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of tiles (exclusive with tilewidth).</p></li>
<li><p><strong>tilewidth</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Width of tiles (exclusive with ntile).</p></li>
<li><p><strong>cut_last_tile_in_chrom</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to cut the last tile in each chromosome.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>GenomicRanges</cite> object with ranges and bin numbers in mcols.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.to_pandas">
<span class="sig-name descname"><span class="pre">to_pandas</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.to_pandas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.to_pandas" title="Link to this definition">¶</a></dt>
<dd><p>Convert this <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> to a <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></a> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></a> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.to_polars">
<span class="sig-name descname"><span class="pre">to_polars</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.to_polars"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.to_polars" title="Link to this definition">¶</a></dt>
<dd><p>Convert this <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> to a <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.trim">
<span class="sig-name descname"><span class="pre">trim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.trim"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.trim" title="Link to this definition">¶</a></dt>
<dd><p>Trim sequences outside of bounds for non-circular chromosomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object in place.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with the trimmed regions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.union">
<span class="sig-name descname"><span class="pre">union</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRanges.union"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.union" title="Link to this definition">¶</a></dt>
<dd><p>Find union of genomic intervals with <cite>other</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – The other <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object.</p></li>
<li><p><strong>ignore_strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to ignore strands. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code> object with all ranges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRanges.width">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">width</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><span class="pre">ndarray</span></a></span><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRanges.width" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_width" title="genomicranges.GenomicRanges.GenomicRanges.get_width"><code class="xref py py-attr docutils literal notranslate"><span class="pre">get_width</span></code></a>.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRangesIter">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-prename descclassname"><span class="pre">genomicranges.GenomicRanges.</span></span><span class="sig-name descname"><span class="pre">GenomicRangesIter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRangesIter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRangesIter" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>An iterator to a <a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a> object.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRangesIter.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRangesIter.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRangesIter.__init__" title="Link to this definition">¶</a></dt>
<dd><p>Initialize the iterator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>obj</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – Source object to iterate.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRangesIter.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRangesIter.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRangesIter.__iter__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.GenomicRanges.GenomicRangesIter.__next__">
<span class="sig-name descname"><span class="pre">__next__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/GenomicRanges.html#GenomicRangesIter.__next__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.GenomicRanges.GenomicRangesIter.__next__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-genomicranges.grangeslist">
<span id="genomicranges-grangeslist-module"></span><h2>genomicranges.grangeslist module<a class="headerlink" href="#module-genomicranges.grangeslist" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGRangesList">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-prename descclassname"><span class="pre">genomicranges.grangeslist.</span></span><span class="sig-name descname"><span class="pre">CompressedGRangesList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">unlist_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partitioning</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">element_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/grangeslist.html#CompressedGRangesList"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.grangeslist.CompressedGRangesList" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList" title="genomicranges.grangeslist.CompressedGenomicRangesList"><code class="xref py py-class docutils literal notranslate"><span class="pre">CompressedGenomicRangesList</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGRangesList.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><span class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{}</span></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGRangesList.__annotations__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-prename descclassname"><span class="pre">genomicranges.grangeslist.</span></span><span class="sig-name descname"><span class="pre">CompressedGenomicRangesList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">unlist_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partitioning</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">element_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/grangeslist.html#CompressedGenomicRangesList"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://biocpy.github.io/compressed-lists/api/compressed_lists.html#compressed_lists.base.CompressedList" title="(in compressed-lists v0.4.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CompressedList</span></code></a></p>
<p>Just as it sounds, a <cite>CompressedGenomicRangesList</cite> is a named-list like object.</p>
<p>If you are wondering why you need this class, a <cite>GenomicRanges</cite> object lets us specify multiple
genomic elements, usually where the genes start and end. Genes are themselves made of many sub
regions, e.g. exons. <cite>CompressedGenomicRangesList</cite> allows us to represent this nested structure.</p>
<p>Currently, this class is limited in functionality, purely a read-only class with basic accessors.</p>
<p>Typical usage:</p>
<p>To construct a <strong>CompressedGenomicRangesList</strong> object, simply pass in a list of
<a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomicranges.GenomicRanges.GenomicRanges</span></code></a> objects and Optionally <code class="docutils literal notranslate"><span class="pre">names</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="n">GenomicRanges</span><span class="p">(</span>
    <span class="n">seqnames</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;chr1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;chr2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;chr1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">ranges</span><span class="o">=</span><span class="n">IRanges</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="mi">1</span><span class="p">,</span>
            <span class="mi">3</span><span class="p">,</span>
            <span class="mi">2</span><span class="p">,</span>
            <span class="mi">4</span><span class="p">,</span>
        <span class="p">],</span>
        <span class="p">[</span>
            <span class="mi">10</span><span class="p">,</span>
            <span class="mi">30</span><span class="p">,</span>
            <span class="mi">50</span><span class="p">,</span>
            <span class="mi">60</span><span class="p">,</span>
        <span class="p">],</span>
    <span class="p">),</span>
    <span class="n">strand</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;-&quot;</span><span class="p">,</span>
        <span class="s2">&quot;+&quot;</span><span class="p">,</span>
        <span class="s2">&quot;*&quot;</span><span class="p">,</span>
        <span class="s2">&quot;+&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">mcols</span><span class="o">=</span><span class="n">BiocFrame</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="mi">1</span><span class="p">,</span>
                <span class="mi">2</span><span class="p">,</span>
                <span class="mi">3</span><span class="p">,</span>
                <span class="mi">4</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">}</span>
    <span class="p">),</span>
<span class="p">)</span>

<span class="n">b</span> <span class="o">=</span> <span class="n">GenomicRanges</span><span class="p">(</span>
    <span class="n">seqnames</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;chr2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;chr4&quot;</span><span class="p">,</span>
        <span class="s2">&quot;chr5&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">ranges</span><span class="o">=</span><span class="n">IRanges</span><span class="p">(</span>
        <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
        <span class="p">[</span>
            <span class="mi">30</span><span class="p">,</span>
            <span class="mi">50</span><span class="p">,</span>
            <span class="mi">60</span><span class="p">,</span>
        <span class="p">],</span>
    <span class="p">),</span>
    <span class="n">strand</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;-&quot;</span><span class="p">,</span>
        <span class="s2">&quot;+&quot;</span><span class="p">,</span>
        <span class="s2">&quot;*&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">mcols</span><span class="o">=</span><span class="n">BiocFrame</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="mi">2</span><span class="p">,</span>
                <span class="mi">3</span><span class="p">,</span>
                <span class="mi">4</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">}</span>
    <span class="p">),</span>
<span class="p">)</span>

<span class="n">grl</span> <span class="o">=</span> <span class="n">CompressedGenomicRangesList</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span>
    <span class="n">lst</span><span class="o">=</span><span class="p">[</span>
        <span class="n">gr1</span><span class="p">,</span>
        <span class="n">gr2</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">names</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;gene1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gene2&quot;</span><span class="p">,</span>
    <span class="p">],</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Additionally, you may also provide metadata about the genomic elements in the dictionary
using mcols attribute.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><span class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{}</span></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.__annotations__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">unlist_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partitioning</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">element_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/grangeslist.html#CompressedGenomicRangesList.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.__init__" title="Link to this definition">¶</a></dt>
<dd><p>Initialize a CompressedIRangesList.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>unlist_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span>) – GenomicRanges object.</p></li>
<li><p><strong>partitioning</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/compressed-lists/api/compressed_lists.html#compressed_lists.partition.Partitioning" title="(in compressed-lists v0.4.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Partitioning</span></code></a></span>) – Partitioning object defining element boundaries.</p></li>
<li><p><strong>element_metadata</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span>) – Optional metadata for elements.</p></li>
<li><p><strong>metadata</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>], <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Optional general metadata.</p></li>
<li><p><strong>kwargs</strong> – Additional arguments.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/grangeslist.html#CompressedGenomicRangesList.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.__repr__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A string representation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.empty">
<span class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/grangeslist.html#CompressedGenomicRangesList.empty"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.empty" title="Link to this definition">¶</a></dt>
<dd><p>Create an zero-length <cite>CompressedGenomicRangesList</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of elements.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>same type as caller, in this case a <cite>CompressedGenomicRangesList</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.end">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">end</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/compressed-lists/api/compressed_lists.html#compressed_lists.numpy_list.CompressedNumpyList" title="(in compressed-lists v0.4.4)"><span class="pre">CompressedNumpyList</span></a></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.end" title="Link to this definition">¶</a></dt>
<dd><p>Access end positions for all elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A CompressedList of end positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.extract_range">
<span class="sig-name descname"><span class="pre">extract_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/grangeslist.html#CompressedGenomicRangesList.extract_range"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.extract_range" title="Link to this definition">¶</a></dt>
<dd><p>Extract a range from <cite>unlist_data</cite>.</p>
<p>This method must be implemented by subclasses to handle
type-specific extraction from <cite>unlist_data</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Start index (inclusive).</p></li>
<li><p><strong>end</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – End index (exclusive).</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Extracted element.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.from_list">
<span class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">from_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/grangeslist.html#CompressedGenomicRangesList.from_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.from_list" title="Link to this definition">¶</a></dt>
<dd><p>Create a <cite>CompressedIRangesList</cite> from a regular list.</p>
<p>This concatenates the list of <cite>GenomicRanges</cite> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a>]</span>) – <p>List of <cite>GenomicRanges</cite> objects.</p>
<p>Must have the same number and names of columns.</p>
</p></li>
<li><p><strong>names</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.Names.Names" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Names</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Optional names for list elements.</p></li>
<li><p><strong>metadata</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>], <a class="reference external" href="https://biocpy.github.io/BiocUtils/api/biocutils.html#biocutils.NamedList.NamedList" title="(in biocutils v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NamedList</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Optional metadata.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList" title="genomicranges.grangeslist.CompressedGenomicRangesList"><code class="xref py py-class docutils literal notranslate"><span class="pre">CompressedGenomicRangesList</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <cite>CompressedList</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.ranges">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">ranges</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/IRanges/api/iranges.html#iranges.irangeslist.CompressedIRangesList" title="(in IRanges v0.7.2)"><span class="pre">CompressedIRangesList</span></a></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.ranges" title="Link to this definition">¶</a></dt>
<dd><p>Access range information for each element in the list.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A CompressedList containing IRanges objects.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.seqinfo">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">seqinfo</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><span class="pre">SeqInfo</span></a></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.seqinfo" title="Link to this definition">¶</a></dt>
<dd><p>Access sequence information.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The SeqInfo object from the underlying GenomicRanges.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.seqnames">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">seqnames</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/compressed-lists/api/compressed_lists.html#compressed_lists.string_list.CompressedCharacterList" title="(in compressed-lists v0.4.4)"><span class="pre">CompressedCharacterList</span></a></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.seqnames" title="Link to this definition">¶</a></dt>
<dd><p>Access sequence names for each element in the list.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A CompressedList containing sequence names.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.stack">
<span class="sig-name descname"><span class="pre">stack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index_column_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'index'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/grangeslist.html#CompressedGenomicRangesList.stack"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.stack" title="Link to this definition">¶</a></dt>
<dd><p>Stack all elements into a single GenomicRanges object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index_column_name</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – Name of the column to store the original index/names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges" title="genomicranges.GenomicRanges.GenomicRanges"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A GenomicRanges object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.start">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">start</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/compressed-lists/api/compressed_lists.html#compressed_lists.numpy_list.CompressedNumpyList" title="(in compressed-lists v0.4.4)"><span class="pre">CompressedNumpyList</span></a></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.start" title="Link to this definition">¶</a></dt>
<dd><p>Access start positions for all elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A CompressedList of start positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.strand">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">strand</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/compressed-lists/api/compressed_lists.html#compressed_lists.string_list.CompressedCharacterList" title="(in compressed-lists v0.4.4)"><span class="pre">CompressedCharacterList</span></a></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.strand" title="Link to this definition">¶</a></dt>
<dd><p>Access strand information for each element in the list.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A CompressedList containing strand information.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.grangeslist.CompressedGenomicRangesList.width">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">width</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://biocpy.github.io/compressed-lists/api/compressed_lists.html#compressed_lists.numpy_list.CompressedNumpyList" title="(in compressed-lists v0.4.4)"><span class="pre">CompressedNumpyList</span></a></span><a class="headerlink" href="#genomicranges.grangeslist.CompressedGenomicRangesList.width" title="Link to this definition">¶</a></dt>
<dd><p>Access widths for all elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A CompressedList of widths.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-genomicranges.sequence_info">
<span id="genomicranges-sequence-info-module"></span><h2>genomicranges.sequence_info module<a class="headerlink" href="#module-genomicranges.sequence_info" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-prename descclassname"><span class="pre">genomicranges.sequence_info.</span></span><span class="sig-name descname"><span class="pre">SeqInfo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqlengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Information about the reference sequences, specifically the name and length of each sequence, whether it is a
circular, and the identity of the genome from which it was derived.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.__copy__">
<span class="sig-name descname"><span class="pre">__copy__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.__copy__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.__copy__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A shallow copy of the current <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.__deepcopy__">
<span class="sig-name descname"><span class="pre">__deepcopy__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_nil</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.__deepcopy__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.__deepcopy__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A deep copy of the current <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.__getitem__" title="Link to this definition">¶</a></dt>
<dd><p>Alias to <a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_subset" title="genomicranges.sequence_info.SeqInfo.get_subset"><code class="xref py py-attr docutils literal notranslate"><span class="pre">get_subset</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqlengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.__init__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqnames</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Names of all reference sequences, should be unique.</p></li>
<li><p><strong>seqlengths</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>Lengths of all sequences in base pairs. This should contain
non-negative values and have the same number of elements as
<code class="docutils literal notranslate"><span class="pre">seqnames</span></code>. Entries may also be None if no lengths are
available for that sequence.</p>
<p>Alternatively, a dictionary where keys are the sequence names
and values are the lengths. If a name is missing from this
dictionary, the length of the sequence is set to None.</p>
<p>Alternatively a single integer, if all sequences are of the
same length.</p>
<p>Alternatively None, if no length information is available
for any sequence.</p>
</p></li>
<li><p><strong>is_circular</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>Whether each sequence is circular. This should have the same
number of elements as <code class="docutils literal notranslate"><span class="pre">seqnames</span></code>. Entries may also be None
if no information is available for that sequence.</p>
<p>Alternatively, a dictionary where keys are the sequence names
and values are the circular flags. If a name is missing from
this dictionary, the flag for the sequence is set to None.</p>
<p>Alternatively a single boolean, if all sequences have the same
circular flag.</p>
<p>Alternatively None, if no flags are available for any sequence.</p>
</p></li>
<li><p><strong>genome</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>The genome build containing each reference sequence. This
should have the same number of elements as <code class="docutils literal notranslate"><span class="pre">seqnames</span></code>.
Entries may also be None if no information is available.</p>
<p>Alternatively, a dictionary where keys are the sequence names
and values are the genomes. If a name is missing from this
dictionary, the genome is set to None.</p>
<p>Alternatively a single string, if all sequences are derived
from the same genome.</p>
<p>Alternatively None, if no genome information is available
for any sequence.</p>
</p></li>
<li><p><strong>validate</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to validate the arguments, internal use only.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.__iter__" title="Link to this definition">¶</a></dt>
<dd><p>Iterator over sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfoIterator" title="genomicranges.sequence_info.SeqInfoIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfoIterator</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.__len__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.__len__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of sequences in this object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.__repr__" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A string representation of this <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.copy" title="Link to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.__copy__" title="genomicranges.sequence_info.SeqInfo.__copy__"><code class="xref py py-meth docutils literal notranslate"><span class="pre">__copy__()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.empty">
<span class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.empty"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.empty" title="Link to this definition">¶</a></dt>
<dd><p>Create an zero-length <cite>SeqInfo</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>same type as caller, in this case a <cite>SeqInfo</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.genome">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">genome</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.genome" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.get_genome">
<span class="sig-name descname"><span class="pre">get_genome</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.get_genome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.get_genome" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of strings is returned containing the genome identity for
all sequences in <a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_seqnames" title="genomicranges.sequence_info.SeqInfo.get_seqnames"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_seqnames()</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.get_is_circular">
<span class="sig-name descname"><span class="pre">get_is_circular</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.get_is_circular"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.get_is_circular" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of booleans is returned specifying whether each sequence
(from <a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_seqnames" title="genomicranges.sequence_info.SeqInfo.get_seqnames"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_seqnames()</span></code></a>) is circular.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.get_seqlengths">
<span class="sig-name descname"><span class="pre">get_seqlengths</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.get_seqlengths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.get_seqlengths" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of integers is returned containing the lengths of all
sequences, in the same order as the sequence names from
<a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_seqnames" title="genomicranges.sequence_info.SeqInfo.get_seqnames"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_seqnames()</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.get_seqnames">
<span class="sig-name descname"><span class="pre">get_seqnames</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.get_seqnames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.get_seqnames" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of all chromosome names.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.get_subset">
<span class="sig-name descname"><span class="pre">get_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.get_subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.get_subset" title="Link to this definition">¶</a></dt>
<dd><p>Subset <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code>, based on their indices or seqnames.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>subset</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – <p>Indices to be extracted. This may be an integer, boolean, string,
or any sequence thereof, as supported by
<code class="xref py py-meth docutils literal notranslate"><span class="pre">normalize_subscript()</span></code>.
Scalars are treated as length-1 sequences.</p>
<p>Strings may only be used if :py:attr:<code class="docutils literal notranslate"><span class="pre">~seqnames</span></code> are available (see
<a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_seqnames" title="genomicranges.sequence_info.SeqInfo.get_seqnames"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_seqnames()</span></code></a>). The first occurrence of each string
in the seqnames is used for extraction.</p>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object with the sequences of interest.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.is_circular">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">is_circular</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a><span class="p"><span class="pre">]</span></span></span><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.is_circular" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.seqlengths">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">seqlengths</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a><span class="p"><span class="pre">]</span></span></span><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.seqlengths" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.seqnames">
<span class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">seqnames</span></span><span class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.seqnames" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.set_genome">
<span class="sig-name descname"><span class="pre">set_genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.set_genome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.set_genome" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – List of genomes, of length equal to the number of names in this
<code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object. Values may be None or strings.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.set_is_circular">
<span class="sig-name descname"><span class="pre">set_is_circular</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_circular</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.set_is_circular"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.set_is_circular" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>is_circular</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>List of circular flags, of length equal to the number of
names in this <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object. Values may be None or
booleans.</p>
<p>Alternatively, a dictionary where keys are the sequence
names and values are the flags. Not all names need to be
present in which case the flag is assumed to be None.</p>
</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.set_seqlengths">
<span class="sig-name descname"><span class="pre">set_seqlengths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqlengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.set_seqlengths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.set_seqlengths" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqlengths</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – <p>List of sequence lengths, of length equal to the number of
names in this <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object. Values may be None or
non-negative integers.</p>
<p>Alternatively, a dictionary where keys are the sequence
names and values are the lengths. Not all names need to be
present in which case the length is assumed to be None.</p>
</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfo.set_seqnames">
<span class="sig-name descname"><span class="pre">set_seqnames</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_place</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfo.set_seqnames"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfo.set_seqnames" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqnames</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – List of sequence names, of length equal to the number of names
in this <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object. All names should be unique strings.</p></li>
<li><p><strong>in_place</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to modify the <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object in place.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A modified <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object, either as a copy of the original
or as a reference to the (in-place-modified) original.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfoIterator">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-prename descclassname"><span class="pre">genomicranges.sequence_info.</span></span><span class="sig-name descname"><span class="pre">SeqInfoIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfoIterator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfoIterator" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>An iterator to a <a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a> object.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfoIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfoIterator.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfoIterator.__init__" title="Link to this definition">¶</a></dt>
<dd><p>Initialize the iterator.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>obj</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span>) – Source object to iterate.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfoIterator.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfoIterator.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfoIterator.__iter__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genomicranges.sequence_info.SeqInfoIterator.__next__">
<span class="sig-name descname"><span class="pre">__next__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#SeqInfoIterator.__next__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.SeqInfoIterator.__next__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genomicranges.sequence_info.merge_SeqInfo">
<span class="sig-prename descclassname"><span class="pre">genomicranges.sequence_info.</span></span><span class="sig-name descname"><span class="pre">merge_SeqInfo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">objects</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/sequence_info.html#merge_SeqInfo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.sequence_info.merge_SeqInfo" title="Link to this definition">¶</a></dt>
<dd><p>Merge multiple <a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a> objects, taking the union of all reference sequences. If the same reference
sequence is present with the same details across <code class="docutils literal notranslate"><span class="pre">objects</span></code>, only a single instance is present in the final object;
if details are contradictory, they are replaced with None.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>objects</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a>]</span>) – List of <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> objects.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo" title="genomicranges.sequence_info.SeqInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqInfo</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A single merged <code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code> object.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-genomicranges.utils">
<span id="genomicranges-utils-module"></span><h2>genomicranges.utils module<a class="headerlink" href="#module-genomicranges.utils" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="genomicranges.utils.compute_up_down">
<span class="sig-prename descclassname"><span class="pre">genomicranges.utils.</span></span><span class="sig-name descname"><span class="pre">compute_up_down</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">starts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strands</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upstream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downstream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'TSS'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/utils.html#compute_up_down"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.utils.compute_up_down" title="Link to this definition">¶</a></dt>
<dd><p>Compute promoter or terminator regions for genomic ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> – GenomicRanges object</p></li>
<li><p><strong>upstream</strong> – Number of bases upstream (scalar or array)</p></li>
<li><p><strong>downstream</strong> – Number of bases downstream (scalar or array)</p></li>
<li><p><strong>site</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a></span>) – “TSS” for promoters or “TES” for terminators</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>New starts and ends.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genomicranges.utils.extract_groups_from_granges">
<span class="sig-prename descclassname"><span class="pre">genomicranges.utils.</span></span><span class="sig-name descname"><span class="pre">extract_groups_from_granges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/utils.html#extract_groups_from_granges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.utils.extract_groups_from_granges" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genomicranges.utils.group_by_indices">
<span class="sig-prename descclassname"><span class="pre">genomicranges.utils.</span></span><span class="sig-name descname"><span class="pre">group_by_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">groups</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/utils.html#group_by_indices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.utils.group_by_indices" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genomicranges.utils.sanitize_strand_vector">
<span class="sig-prename descclassname"><span class="pre">genomicranges.utils.</span></span><span class="sig-name descname"><span class="pre">sanitize_strand_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/utils.html#sanitize_strand_vector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.utils.sanitize_strand_vector" title="Link to this definition">¶</a></dt>
<dd><p>Create a numpy representation for <code class="docutils literal notranslate"><span class="pre">strand</span></code>.</p>
<p>Mapping: 1 for “+” (forward strand), 0 for “*” (any strand) and -1 for “-” (reverse strand).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>strand</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>], <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>]</span>) – List of strand.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.12/library/exceptions.html#ValueError" title="(in Python v3.12)"><strong>ValueError</strong></a> – If strand is None.
    If strand contains values other than +,- and <a href="#id5"><span class="problematic" id="id6">*</span></a>.
    If strand is not a numpy vector, string of integers or strings.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A numpy vector.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genomicranges.utils.slide_intervals">
<span class="sig-prename descclassname"><span class="pre">genomicranges.utils.</span></span><span class="sig-name descname"><span class="pre">slide_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/utils.html#slide_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.utils.slide_intervals" title="Link to this definition">¶</a></dt>
<dd><p>Sliding intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Start interval.</p></li>
<li><p><strong>end</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – End interval.</p></li>
<li><p><strong>step</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Step of each interval.</p></li>
<li><p><strong>width</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Width of each interval.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of intervals split into bins.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genomicranges.utils.split_intervals">
<span class="sig-prename descclassname"><span class="pre">genomicranges.utils.</span></span><span class="sig-name descname"><span class="pre">split_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/utils.html#split_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.utils.split_intervals" title="Link to this definition">¶</a></dt>
<dd><p>Split an interval range into equal bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Start interval.</p></li>
<li><p><strong>end</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – End interval.</p></li>
<li><p><strong>step</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Width or step of each interval.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.List" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of intervals split into bins.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genomicranges.utils.wrapper_follow_precede">
<span class="sig-prename descclassname"><span class="pre">genomicranges.utils.</span></span><span class="sig-name descname"><span class="pre">wrapper_follow_precede</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genomicranges/utils.html#wrapper_follow_precede"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genomicranges.utils.wrapper_follow_precede" title="Link to this definition">¶</a></dt>
<dd><p>Processes a single group for precede/follow operations.
This function is designed to be called by a multiprocessing pool.</p>
</dd></dl>

</section>
<section id="module-genomicranges">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-genomicranges" title="Link to this heading">¶</a></h2>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="genomicranges.io.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">genomicranges.io package</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="modules.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">genomicranges</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, jkanche
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">genomicranges package</a><ul>
<li><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-genomicranges.GenomicRanges">genomicranges.GenomicRanges module</a><ul>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GRanges"><code class="docutils literal notranslate"><span class="pre">GRanges</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges"><code class="docutils literal notranslate"><span class="pre">GenomicRanges</span></code></a><ul>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__annotations__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__copy__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__copy__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__deepcopy__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__deepcopy__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__getitem__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__getitem__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__init__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__init__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__iter__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__iter__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__len__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__len__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__repr__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.__setitem__"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.__setitem__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.binned_average"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.binned_average()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.copy"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.copy()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.count_overlaps"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.count_overlaps()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.coverage"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.coverage()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.disjoin"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.disjoin()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.disjoint_bins"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.disjoint_bins()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.distance"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.distance()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.empty"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.empty()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.end"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.end</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.extract_groups_by_seqnames"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.extract_groups_by_seqnames()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.find_overlaps"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.find_overlaps()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.flank"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.flank()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.follow"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.follow()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.from_pandas"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.from_pandas()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.from_polars"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.from_polars()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.gaps"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.gaps()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_end"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_end()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_mcols"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_mcols()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_names"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_names()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_out_of_bound_index"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_out_of_bound_index()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_ranges"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_ranges()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_seqinfo"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_seqinfo()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_seqlengths"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_seqlengths()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_seqnames"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_seqnames()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_start"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_start()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_strand"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_strand()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_subset"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_subset()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.get_width"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.get_width()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.intersect"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.intersect()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.intersect_ncls"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.intersect_ncls()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.invert_strand"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.invert_strand()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.is_disjoint"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.is_disjoint()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.match"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.match()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.mcols"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.mcols</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.names"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.names</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.narrow"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.narrow()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.nearest"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.nearest()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.order"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.order()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.pintersect"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.pintersect()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.precede"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.precede()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.promoters"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.promoters()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.range"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.range()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.ranges"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.ranges</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.rank"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.rank()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.reduce"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.reduce()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.resize"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.resize()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.restrict"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.restrict()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.sample"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.sample()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.seqinfo"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.seqinfo</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.seqnames"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.seqnames</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.set_mcols"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.set_mcols()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.set_names"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.set_names()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.set_ranges"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.set_ranges()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.set_seqinfo"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.set_seqinfo()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.set_seqnames"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.set_seqnames()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.set_strand"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.set_strand()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.set_subset"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.set_subset()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.setdiff"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.setdiff()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.shift"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.shift()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.sliding_windows"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.sliding_windows()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.sort"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.sort()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.split"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.split()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.start"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.start</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.strand"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.strand</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.subset_by_overlaps"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.subset_by_overlaps()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.subtract"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.subtract()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.terminators"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.terminators()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.tile"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.tile()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.tile_genome"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.tile_genome()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.to_pandas"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.to_pandas()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.to_polars"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.to_polars()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.trim"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.trim()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.union"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.union()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRanges.width"><code class="docutils literal notranslate"><span class="pre">GenomicRanges.width</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRangesIter"><code class="docutils literal notranslate"><span class="pre">GenomicRangesIter</span></code></a><ul>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRangesIter.__init__"><code class="docutils literal notranslate"><span class="pre">GenomicRangesIter.__init__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRangesIter.__iter__"><code class="docutils literal notranslate"><span class="pre">GenomicRangesIter.__iter__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.GenomicRanges.GenomicRangesIter.__next__"><code class="docutils literal notranslate"><span class="pre">GenomicRangesIter.__next__()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-genomicranges.grangeslist">genomicranges.grangeslist module</a><ul>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGRangesList"><code class="docutils literal notranslate"><span class="pre">CompressedGRangesList</span></code></a><ul>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGRangesList.__annotations__"><code class="docutils literal notranslate"><span class="pre">CompressedGRangesList.__annotations__</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList</span></code></a><ul>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.__annotations__"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.__annotations__</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.__init__"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.__init__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.__repr__"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.empty"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.empty()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.end"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.end</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.extract_range"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.extract_range()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.from_list"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.from_list()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.ranges"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.ranges</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.seqinfo"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.seqinfo</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.seqnames"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.seqnames</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.stack"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.stack()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.start"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.start</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.strand"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.strand</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.grangeslist.CompressedGenomicRangesList.width"><code class="docutils literal notranslate"><span class="pre">CompressedGenomicRangesList.width</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-genomicranges.sequence_info">genomicranges.sequence_info module</a><ul>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo"><code class="docutils literal notranslate"><span class="pre">SeqInfo</span></code></a><ul>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.__copy__"><code class="docutils literal notranslate"><span class="pre">SeqInfo.__copy__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.__deepcopy__"><code class="docutils literal notranslate"><span class="pre">SeqInfo.__deepcopy__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.__getitem__"><code class="docutils literal notranslate"><span class="pre">SeqInfo.__getitem__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.__init__"><code class="docutils literal notranslate"><span class="pre">SeqInfo.__init__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.__iter__"><code class="docutils literal notranslate"><span class="pre">SeqInfo.__iter__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.__len__"><code class="docutils literal notranslate"><span class="pre">SeqInfo.__len__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.__repr__"><code class="docutils literal notranslate"><span class="pre">SeqInfo.__repr__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.copy"><code class="docutils literal notranslate"><span class="pre">SeqInfo.copy()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.empty"><code class="docutils literal notranslate"><span class="pre">SeqInfo.empty()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.genome"><code class="docutils literal notranslate"><span class="pre">SeqInfo.genome</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_genome"><code class="docutils literal notranslate"><span class="pre">SeqInfo.get_genome()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_is_circular"><code class="docutils literal notranslate"><span class="pre">SeqInfo.get_is_circular()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_seqlengths"><code class="docutils literal notranslate"><span class="pre">SeqInfo.get_seqlengths()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_seqnames"><code class="docutils literal notranslate"><span class="pre">SeqInfo.get_seqnames()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.get_subset"><code class="docutils literal notranslate"><span class="pre">SeqInfo.get_subset()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.is_circular"><code class="docutils literal notranslate"><span class="pre">SeqInfo.is_circular</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.seqlengths"><code class="docutils literal notranslate"><span class="pre">SeqInfo.seqlengths</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.seqnames"><code class="docutils literal notranslate"><span class="pre">SeqInfo.seqnames</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.set_genome"><code class="docutils literal notranslate"><span class="pre">SeqInfo.set_genome()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.set_is_circular"><code class="docutils literal notranslate"><span class="pre">SeqInfo.set_is_circular()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.set_seqlengths"><code class="docutils literal notranslate"><span class="pre">SeqInfo.set_seqlengths()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfo.set_seqnames"><code class="docutils literal notranslate"><span class="pre">SeqInfo.set_seqnames()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfoIterator"><code class="docutils literal notranslate"><span class="pre">SeqInfoIterator</span></code></a><ul>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfoIterator.__init__"><code class="docutils literal notranslate"><span class="pre">SeqInfoIterator.__init__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfoIterator.__iter__"><code class="docutils literal notranslate"><span class="pre">SeqInfoIterator.__iter__()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.sequence_info.SeqInfoIterator.__next__"><code class="docutils literal notranslate"><span class="pre">SeqInfoIterator.__next__()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#genomicranges.sequence_info.merge_SeqInfo"><code class="docutils literal notranslate"><span class="pre">merge_SeqInfo()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-genomicranges.utils">genomicranges.utils module</a><ul>
<li><a class="reference internal" href="#genomicranges.utils.compute_up_down"><code class="docutils literal notranslate"><span class="pre">compute_up_down()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.utils.extract_groups_from_granges"><code class="docutils literal notranslate"><span class="pre">extract_groups_from_granges()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.utils.group_by_indices"><code class="docutils literal notranslate"><span class="pre">group_by_indices()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.utils.sanitize_strand_vector"><code class="docutils literal notranslate"><span class="pre">sanitize_strand_vector()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.utils.slide_intervals"><code class="docutils literal notranslate"><span class="pre">slide_intervals()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.utils.split_intervals"><code class="docutils literal notranslate"><span class="pre">split_intervals()</span></code></a></li>
<li><a class="reference internal" href="#genomicranges.utils.wrapper_follow_precede"><code class="docutils literal notranslate"><span class="pre">wrapper_follow_precede()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-genomicranges">Module contents</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=b41f4c93"></script>
    <script src="../_static/doctools.js?v=fd6eb6e6"></script>
    <script src="../_static/sphinx_highlight.js?v=6ffebe34"></script>
    <script src="../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>